// @oclif/core/lib/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.memoizedWarn = exports.warn = exports.error = exports.exit = exports.config = exports.Logger = exports.CLIError = exports.ModuleLoadError = exports.ExitError = exports.handle = void 0;
const tslib_1 = require("tslib");
const cli_1 = require("./errors/cli");
const pretty_print_1 = tslib_1.__importStar(require("./errors/pretty-print"));
const exit_1 = require("./errors/exit");
const config_1 = require("./config");
var handle_1 = require("./handle");
Object.defineProperty(exports, "handle", { enumerable: true, get: function () { return handle_1.handle; } });
var exit_2 = require("./errors/exit");
Object.defineProperty(exports, "ExitError", { enumerable: true, get: function () { return exit_2.ExitError; } });
var module_load_1 = require("./errors/module-load");
Object.defineProperty(exports, "ModuleLoadError", { enumerable: true, get: function () { return module_load_1.ModuleLoadError; } });
var cli_2 = require("./errors/cli");
Object.defineProperty(exports, "CLIError", { enumerable: true, get: function () { return cli_2.CLIError; } });
var logger_1 = require("./logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return logger_1.Logger; } });
var config_2 = require("./config");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return config_2.config; } });
function exit(code = 0) {
    throw new exit_1.ExitError(code);
}
exports.exit = exit;
function error(input, options = {}) {
    let err;
    if (typeof input === 'string') {
        err = new cli_1.CLIError(input, options);
    }
    else if (input instanceof Error) {
        err = (0, cli_1.addOclifExitCode)(input, options);
    }
    else {
        throw new TypeError('first argument must be a string or instance of Error');
    }
    err = (0, pretty_print_1.applyPrettyPrintOptions)(err, options);
    if (options.exit === false) {
        const message = (0, pretty_print_1.default)(err);
        console.error(message);
        if (config_1.config.errorLogger)
            config_1.config.errorLogger.log(err?.stack ?? '');
    }
    else
        throw err;
}
exports.error = error;
function warn(input) {
    let err;
    if (typeof input === 'string') {
        err = new cli_1.CLIError.Warn(input);
    }
    else if (input instanceof Error) {
        err = (0, cli_1.addOclifExitCode)(input);
    }
    else {
        throw new TypeError('first argument must be a string or instance of Error');
    }
    const message = (0, pretty_print_1.default)(err);
    console.error(message);
    if (config_1.config.errorLogger)
        config_1.config.errorLogger.log(err?.stack ?? '');
}
exports.warn = warn;
const WARNINGS = new Set();
function memoizedWarn(input) {
    if (!WARNINGS.has(input))
        warn(input);
    WARNINGS.add(input);
}
exports.memoizedWarn = memoizedWarn;
