// @mitre/saf/lib/commands/supplement/passthrough/write.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const fs_1 = tslib_1.__importDefault(require("fs"));
const baseCommand_1 = require("../../../utils/oclif/baseCommand");
class WritePassthrough extends baseCommand_1.BaseCommand {
    static usage = '<%= command.id %> -i <input-hdf-json> (-f <input-passthrough-json> | -d <passthrough-json>) [-o <output-hdf-json>]';
    static summary = 'Overwrite the `passthrough` attribute in a given HDF file with the provided `passthrough` JSON data';
    static description = 'Passthrough data can be any context/structure. See sample ideas at https://github.com/mitre/saf/wiki/Supplement-HDF-files-with-additional-information-(ex.-%60passthrough%60,-%60target%60)';
    static examples = [
        {
            description: '\x1B[93mProviding passthrough-data\x1B[0m',
            command: '<%= config.bin %> <%= command.id %> -i hdf.json -d \'{"a": 5}\'',
        },
        {
            description: '\x1B[93mUsing passthrough-data file\x1B[0m',
            command: '<%= config.bin %> <%= command.id %> -i hdf.json -f passthrough.json -o new-hdf.json',
        },
    ];
    static flags = {
        input: core_1.Flags.string({ char: 'i', required: true, description: 'An input Heimdall Data Format file' }),
        passthroughFile: core_1.Flags.string({ char: 'f', exclusive: ['passthroughData'], description: 'An input passthrough-data file (can contain any valid JSON); this flag or `passthroughData` must be provided' }),
        passthroughData: core_1.Flags.string({ char: 'd', exclusive: ['passthroughFile'], description: 'Input passthrough-data (can be any valid JSON); this flag or `passthroughFile` must be provided' }),
        output: core_1.Flags.string({ char: 'o', description: 'An output Heimdall Data Format JSON file (otherwise the input file is overwritten)' }),
    };
    async run() {
        const { flags } = await this.parse(WritePassthrough);
        const input = JSON.parse(fs_1.default.readFileSync(flags.input, 'utf8'));
        const output = flags.output || flags.input;
        let passthrough;
        if (flags.passthroughFile) {
            try {
                passthrough = JSON.parse(fs_1.default.readFileSync(flags.passthroughFile, 'utf8'));
            }
            catch (error) {
                throw new Error(`Couldn't parse passthrough data: ${error}`);
            }
        }
        else if (flags.passthroughData) {
            try {
                passthrough = JSON.parse(flags.passthroughData);
            }
            catch {
                passthrough = flags.passthroughData;
            }
        }
        else {
            throw new Error('One out of passthroughFile or passthroughData must be passed');
        }
        input.passthrough = passthrough;
        fs_1.default.writeFileSync(output, JSON.stringify(input, null, 2));
    }
}
exports.default = WritePassthrough;
