// @mitre/saf/lib/commands/supplement/passthrough/write.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const fs_1 = tslib_1.__importDefault(require("fs"));
class WritePassthrough extends core_1.Command {
    async run() {
        const { flags } = await this.parse(WritePassthrough);
        const input = JSON.parse(fs_1.default.readFileSync(flags.input, 'utf8'));
        const output = flags.output || flags.input;
        let passthrough;
        if (flags.passthroughFile) {
            try {
                passthrough = JSON.parse(fs_1.default.readFileSync(flags.passthroughFile, 'utf8'));
            }
            catch (error) {
                throw new Error(`Couldn't parse passthrough data: ${error}`);
            }
        }
        else if (flags.passthroughData) {
            try {
                passthrough = JSON.parse(flags.passthroughData);
            }
            catch {
                passthrough = flags.passthroughData;
            }
        }
        else {
            throw new Error('One out of passthroughFile or passthroughData must be passed');
        }
        input.passthrough = passthrough;
        fs_1.default.writeFileSync(output, JSON.stringify(input, null, 2));
    }
}
WritePassthrough.usage = 'supplement passthrough write -i <input-hdf-json> (-f <input-passthrough-json> | -d <passthrough-json>) [-o <output-hdf-json>]';
WritePassthrough.summary = 'Overwrite the `passthrough` attribute in a given HDF file with the provided `passthrough` JSON data';
WritePassthrough.description = 'Passthrough data can be any context/structure. See sample ideas at https://github.com/mitre/saf/wiki/Supplement-HDF-files-with-additional-information-(ex.-%60passthrough%60,-%60target%60)';
WritePassthrough.examples = [
    'saf supplement passthrough write -i hdf.json -d \'{"a": 5}\'',
    'saf supplement passthrough write -i hdf.json -f passthrough.json -o new-hdf.json',
];
WritePassthrough.flags = {
    help: core_1.Flags.help({ char: 'h' }),
    input: core_1.Flags.string({ char: 'i', required: true, description: 'An input Heimdall Data Format file' }),
    passthroughFile: core_1.Flags.string({ char: 'f', exclusive: ['passthroughData'], description: 'An input passthrough-data file (can contain any valid JSON); this flag or `passthroughData` must be provided' }),
    passthroughData: core_1.Flags.string({ char: 'd', exclusive: ['passthroughFile'], description: 'Input passthrough-data (can be any valid JSON); this flag or `passthroughFile` must be provided' }),
    output: core_1.Flags.string({ char: 'o', description: 'An output Heimdall Data Format JSON file (otherwise the input file is overwritten)' }),
};
exports.default = WritePassthrough;
