// @oclif/core/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.settings = exports.Performance = exports.Parser = exports.run = exports.Interfaces = exports.toStandardizedId = exports.toConfiguredId = exports.loadHelpClass = exports.HelpBase = exports.Help = exports.CommandHelp = exports.Flags = exports.execute = exports.handle = exports.Errors = exports.Plugin = exports.Config = exports.Command = exports.stdout = exports.stderr = exports.flush = exports.ux = exports.Args = void 0;
const tslib_1 = require("tslib");
const stream_1 = require("./cli-ux/stream");
function checkCWD() {
    try {
        process.cwd();
    }
    catch (error) {
        if (error.code === 'ENOENT') {
            stream_1.stderr.write('WARNING: current directory does not exist\n');
        }
    }
}
checkCWD();
exports.Args = tslib_1.__importStar(require("./args"));
exports.ux = tslib_1.__importStar(require("./cli-ux"));
var flush_1 = require("./cli-ux/flush");
Object.defineProperty(exports, "flush", { enumerable: true, get: function () { return flush_1.flush; } });
var stream_2 = require("./cli-ux/stream");
Object.defineProperty(exports, "stderr", { enumerable: true, get: function () { return stream_2.stderr; } });
Object.defineProperty(exports, "stdout", { enumerable: true, get: function () { return stream_2.stdout; } });
var command_1 = require("./command");
Object.defineProperty(exports, "Command", { enumerable: true, get: function () { return command_1.Command; } });
var config_1 = require("./config");
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return config_1.Config; } });
Object.defineProperty(exports, "Plugin", { enumerable: true, get: function () { return config_1.Plugin; } });
exports.Errors = tslib_1.__importStar(require("./errors"));
var handle_1 = require("./errors/handle");
Object.defineProperty(exports, "handle", { enumerable: true, get: function () { return handle_1.handle; } });
var execute_1 = require("./execute");
Object.defineProperty(exports, "execute", { enumerable: true, get: function () { return execute_1.execute; } });
exports.Flags = tslib_1.__importStar(require("./flags"));
var help_1 = require("./help");
Object.defineProperty(exports, "CommandHelp", { enumerable: true, get: function () { return help_1.CommandHelp; } });
Object.defineProperty(exports, "Help", { enumerable: true, get: function () { return help_1.Help; } });
Object.defineProperty(exports, "HelpBase", { enumerable: true, get: function () { return help_1.HelpBase; } });
Object.defineProperty(exports, "loadHelpClass", { enumerable: true, get: function () { return help_1.loadHelpClass; } });
var util_1 = require("./help/util");
Object.defineProperty(exports, "toConfiguredId", { enumerable: true, get: function () { return util_1.toConfiguredId; } });
Object.defineProperty(exports, "toStandardizedId", { enumerable: true, get: function () { return util_1.toStandardizedId; } });
exports.Interfaces = tslib_1.__importStar(require("./interfaces"));
var main_1 = require("./main");
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return main_1.run; } });
exports.Parser = tslib_1.__importStar(require("./parser"));
var performance_1 = require("./performance");
Object.defineProperty(exports, "Performance", { enumerable: true, get: function () { return performance_1.Performance; } });
var settings_1 = require("./settings");
Object.defineProperty(exports, "settings", { enumerable: true, get: function () { return settings_1.settings; } });
