// @oclif/test/lib/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.command = exports.Config = exports.FancyTypes = exports.expect = exports.test = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const core_1 = require("@oclif/core");
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return core_1.Config; } });
const fancy_test_1 = require("fancy-test");
Object.defineProperty(exports, "expect", { enumerable: true, get: function () { return fancy_test_1.expect; } });
Object.defineProperty(exports, "FancyTypes", { enumerable: true, get: function () { return fancy_test_1.FancyTypes; } });
const command_1 = require("./command");
Object.defineProperty(exports, "command", { enumerable: true, get: function () { return command_1.command; } });
const exit_1 = tslib_1.__importDefault(require("./exit"));
const hook_1 = tslib_1.__importDefault(require("./hook"));
const load_config_1 = require("./load-config");
function traverseFilePathUntil(filename, predicate) {
    let current = filename;
    while (!predicate(current)) {
        current = path_1.default.dirname(current);
    }
    return current;
}
// Update to path.dirname(url.fileURLToPath(import.meta.url)) whenever we update tsconfig target to ES2020
// eslint-disable-next-line unicorn/prefer-module
load_config_1.loadConfig.root = traverseFilePathUntil((_b = (_a = require.main) === null || _a === void 0 ? void 0 : _a.path) !== null && _b !== void 0 ? _b : module.path, p => !p.includes('node_modules'));
exports.test = fancy_test_1.fancy
    .register('loadConfig', load_config_1.loadConfig)
    .register('command', command_1.command)
    .register('exit', exit_1.default)
    .register('hook', hook_1.default)
    .env({ NODE_ENV: 'test' });
exports.default = exports.test;
