// @mitre/saf/lib/commands/emasser/get/dashboards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const json_colorizer_1 = tslib_1.__importDefault(require("json-colorizer"));
const core_1 = require("@oclif/core");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const emass_client_1 = require("@mitre/emass_client");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const outputError_1 = require("../../../utils/emasser/outputError");
const utilities_1 = require("../../../utils/emasser/utilities");
const endpoint = 'dashboards';
class EmassergetDashboards extends core_1.Command {
    async run() {
        const { args, flags } = await this.parse(EmassergetDashboards);
        const apiCxn = new apiConnection_1.ApiConnection();
        const getDashboards = new emass_client_1.DashboardsApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        switch (args.name) {
            case 'status_details': {
                // Order is important here
                getDashboards.getSystemStatusDetails(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'control_compliance_summary': {
                // Order is important here
                getDashboards.getSystemControlComplianceSummary(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'security_control_details': {
                // Order is important here
                getDashboards.getSystemSecurityControlDetails(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'assessment_procedures_details': {
                // Order is important here
                getDashboards.getSystemAssessmentProceduresDetails(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'poam_summary': {
                // Order is important here
                getDashboards.getSystemPoamSummary(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'poam_details': {
                // Order is important here
                getDashboards.getSystemPoamDetails(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'hardware_summary': {
                // Order is important here
                getDashboards.getSystemHardwareSummary(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'hardware_details': {
                // Order is important here
                getDashboards.getSystemHardwareDetails(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'associations_details': {
                // Order is important here
                getDashboards.getSystemAssociationsDetails(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'assignments_details': {
                // Order is important here
                getDashboards.getUserSystemAssignmentsDetails(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'privacy_summary': {
                // Order is important here
                getDashboards.getSystemPrivacySummary(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'fisma_saop_summary': {
                // Order is important here
                getDashboards.getVaOmbFsmaSaopSummary(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'va_aa_summary': {
                // Order is important here
                getDashboards.getVaSystemAaSummary(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'va_a2_summary': {
                // Order is important here
                getDashboards.getVaSystemA2Summary(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'va_pl_109_summary': {
                // Order is important here
                getDashboards.getVaSystemPl109ReportingSummary(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            case 'fisma_inventory_summary': {
                // Order is important here
                getDashboards.getVaSystemFismaInvetorySummary(flags.orgId, flags.pageIndex, flags.pageSize).then((response) => {
                    console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
                }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
                break;
            }
            default: {
                throw this.error;
            }
        }
    }
    async catch(error) {
        if (error.message) {
            this.error(error);
        }
        else {
            const suggestions = 'get dashboards [-h or --help] for available arguments';
            this.warn('Invalid arguments\nTry this:\n\t' + suggestions);
        }
    }
}
EmassergetDashboards.usage = '<%= command.id %> [ARGUMENTS]';
EmassergetDashboards.description = (0, utilities_1.getDescriptionForEndpoint)(process.argv, endpoint);
EmassergetDashboards.examples = (0, utilities_1.getExamplesForEndpoint)(process.argv, endpoint);
EmassergetDashboards.flags = {
    help: core_1.Flags.help({ char: 'h', description: 'Show emasser CLI help for the GET Dashboards endpoint' }),
    ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
};
// NOTE: The way args are being implemented are mainly for the purposes of help clarity, there is, displays
//       the available arguments with associate description.
// Only args.name is used, there is, it contains the argument listed by the user.
// Example: If the user uses the command (saf emasser get dashboards status_details), args.name is set to status_details
EmassergetDashboards.args = {
    name: core_1.Args.string({ name: 'name', required: false, hidden: true }),
    status_details: core_1.Args.string({ name: 'status_details', description: 'Get systems status detail dashboard information', required: false }),
    control_compliance_summary: core_1.Args.string({ name: 'control_compliance_summary', description: 'Get systems control compliance summary dashboard information', required: false }),
    security_control_details: core_1.Args.string({ name: 'security_control_details', description: 'Get systems security control details dashboard information', required: false }),
    assessment_procedures_details: core_1.Args.string({ name: 'assessment_procedures_details', description: 'Get systems assessment procedures details dashboard information', required: false }),
    poam_summary: core_1.Args.string({ name: 'poam_summary', description: 'Get systems POA&Ms summary dashboard information', required: false }),
    poam_details: core_1.Args.string({ name: 'poam_details', description: 'Get system POA&Ms details dashboard information', required: false }),
    hardware_summary: core_1.Args.string({ name: 'hardware_summary', description: 'Get system hardware summary dashboard information', required: false }),
    hardware_details: core_1.Args.string({ name: 'hardware_details', description: 'Get system hardware details dashboard information', required: false }),
    associations_details: core_1.Args.string({ name: 'associations_details', description: 'Get system associations details dashboard information', required: false }),
    assignments_details: core_1.Args.string({ name: 'assignments_details', description: 'Get user system assignments details dashboard information', required: false }),
    privacy_summary: core_1.Args.string({ name: 'privacy_summary', description: 'Get user system privacy summary dashboard information', required: false }),
    fisma_saop_summary: core_1.Args.string({ name: 'fisma_saop_summary', description: 'Get VA OMB-FISMA SAOP summary dashboard information', required: false }),
    va_aa_summary: core_1.Args.string({ name: 'va_aa_summary', description: 'Get VA system A&A summary dashboard information', required: false }),
    va_a2_summary: core_1.Args.string({ name: 'va_a2_summary', description: 'Get VA system A2.0 summary dashboard information', required: false }),
    va_pl_109_summary: core_1.Args.string({ name: 'va_pl_109_summary', description: 'Get VA System P.L. 109 reporting summary dashboard information', required: false }),
    fisma_inventory_summary: core_1.Args.string({ name: 'fisma_inventory_summary', description: 'Get VA system FISMA inventory summary dashboard information', required: false }),
};
exports.default = EmassergetDashboards;
