// @mitre/saf/lib/commands/emasser/get/milestones.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const json_colorizer_1 = require("json-colorizer");
const core_1 = require("@oclif/core");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const emass_client_1 = require("@mitre/emass_client");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const outputError_1 = require("../../../utils/emasser/outputError");
const utilities_1 = require("../../../utils/emasser/utilities");
const endpoint = 'milestones';
class EmasserGetMilestones extends core_1.Command {
    static usage = '<%= command.id %> [ARGUMENT] \n \x1B[93m NOTE: see EXAMPLES for argument case format\x1B[0m';
    static description = (0, utilities_1.getDescriptionForEndpoint)(process.argv, endpoint);
    static examples = (0, utilities_1.getExamplesForEndpoint)(process.argv, endpoint);
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Show emasser CLI help for the GET Milestones endpoint' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    // NOTE: The way args are being implemented are mainly for the purposes of help clarity, there is, displays
    //       the available arguments with associate description.
    // Only args.name is used, there is, it contains the argument listed by the user.
    // Example: If the user uses the command (saf emasser get milestones byPoamId), args.name is set to byPoamId
    static args = {
        name: core_1.Args.string({ name: 'name', required: false, hidden: true }),
        byPoamId: core_1.Args.string({ name: 'byPoamId', description: 'Retrieves milestone(s) for specified system and poam Id', required: false }),
        byMilestoneId: core_1.Args.string({ name: 'byMilestoneId', description: 'Retrieves milestone(s) for specified system, poam, and milestone Id', required: false }),
    };
    async run() {
        const { args, flags } = await this.parse(EmasserGetMilestones);
        const apiCxn = new apiConnection_1.ApiConnection();
        const getMilestones = new emass_client_1.MilestonesApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        if (args.name === 'byPoamId') {
            // Order is important here
            getMilestones.getSystemMilestonesByPoamId(flags.systemId, flags.poamId, flags.scheduledCompletionDateStart, flags.scheduledCompletionDateEnd).then((response) => {
                console.log((0, json_colorizer_1.colorize)((0, outputFormatter_1.outputFormat)(response)));
            }).catch((error) => console.error((0, json_colorizer_1.colorize)((0, outputError_1.outputError)(error))));
        }
        else if (args.name === 'byMilestoneId') {
            // Order is important here
            getMilestones.getSystemMilestonesByPoamIdAndMilestoneId(flags.systemId, flags.poamId, flags.milestoneId).then((response) => {
                console.log((0, json_colorizer_1.colorize)((0, outputFormatter_1.outputFormat)(response)));
            }).catch((error) => console.error((0, json_colorizer_1.colorize)((0, outputError_1.outputError)(error))));
        }
        else {
            throw this.error;
        }
    }
    async catch(error) {
        if (error.message) {
            this.error(error);
        }
        else {
            const suggestions = 'get milestones [-h or --help]\n\tget milestones byPoamId\n\tget milestones byMilestoneId';
            this.warn('Invalid arguments\nTry this:\n\t' + suggestions);
        }
    }
}
exports.default = EmasserGetMilestones;
