// @mitre/saf/lib/commands/emasser/get/poams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const json_colorizer_1 = tslib_1.__importDefault(require("json-colorizer"));
const core_1 = require("@oclif/core");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const emass_client_1 = require("@mitre/emass_client");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const outputError_1 = require("../../../utils/emasser/outputError");
const utilities_1 = require("../../../utils/emasser/utilities");
const endpoint = 'poams';
class EmasserGetPoams extends core_1.Command {
    static usage = '<%= command.id %> [ARGUMENT] \n \x1B[93m NOTE: see EXAMPLES for argument case format\x1B[0m';
    static description = (0, utilities_1.getDescriptionForEndpoint)(process.argv, endpoint);
    static examples = (0, utilities_1.getExamplesForEndpoint)(process.argv, endpoint);
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Show emasser CLI help for the GET POA&Ms endpoint' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    // NOTE: The way args are being implemented are mainly for the purposes of help clarity, there is, displays
    //       the available arguments with associate description.
    // Only args.name is used, there is, it contains the argument listed by the user.
    // Example: If the user uses the command (saf emasser get poams byPoamId), args.name is set to byPoamId
    static args = {
        name: core_1.Args.string({ name: 'name', required: false, hidden: true }),
        forSystem: core_1.Args.string({ name: 'forSystem', description: 'Retrieves Poams for specified system ID', required: false }),
        byPoamId: core_1.Args.string({ name: 'byPoamId', description: 'Retrieves Poams for specified system and poam ID', required: false }),
    };
    async run() {
        const { args, flags } = await this.parse(EmasserGetPoams);
        const apiCxn = new apiConnection_1.ApiConnection();
        const getPoams = new emass_client_1.POAMApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        if (args.name === 'forSystem') {
            // Order is important here
            getPoams.getSystemPoams(flags.systemId, flags.scheduledCompletionDateStart, flags.scheduledCompletionDateEnd, flags.controlAcronyms, flags.ccis, flags.systemOnly).then((response) => {
                console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
            }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
        }
        else if (args.name === 'byPoamId') {
            // Order is important here
            getPoams.getSystemPoamsByPoamId(flags.systemId, flags.poamId).then((response) => {
                console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
            }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
        }
        else {
            throw this.error;
        }
    }
    async catch(error) {
        if (error.message) {
            this.error(error);
        }
        else {
            const suggestions = 'get poams [-h or --help]\n\tget poams forSystem\n\tget poams byPoamId';
            this.warn('Invalid arguments\nTry this:\n\t' + suggestions);
        }
    }
}
exports.default = EmasserGetPoams;
