// @mitre/saf/lib/commands/emasser/get/roles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const json_colorizer_1 = require("json-colorizer");
const core_1 = require("@oclif/core");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const emass_client_1 = require("@mitre/emass_client");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const utilities_1 = require("../../../utils/emasser/utilities");
const endpoint = 'roles';
class EmasserGetRoles extends core_1.Command {
    static usage = '<%= command.id %> [ARGUMENT] [FLAGS]\n \x1B[93m NOTE: see EXAMPLES for argument case format\x1B[0m';
    static description = (0, utilities_1.getDescriptionForEndpoint)(process.argv, endpoint);
    static examples = (0, utilities_1.getExamplesForEndpoint)(process.argv, endpoint);
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Show eMASSer CLI help for the GET Roles command' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    // NOTE: The way args are being implemented are mainly for the purposes of help clarity, there is, displays
    //       the available arguments with associate description.
    // Only args.name is used, there is, it contains the argument listed by the user.
    // Example: If the user uses the command (saf emasser get roles byCategory), args.name is set to byCategory
    static args = {
        name: core_1.Args.string({ name: 'name', required: false, hidden: true }),
        all: core_1.Args.string({ name: 'all', description: 'Retrieves all available system roles', required: false }),
        byCategory: core_1.Args.string({ name: 'byCategory', description: 'Retrieves role(s) - filtered by [options] params', required: false }),
    };
    async run() {
        const { args, flags } = await this.parse(EmasserGetRoles);
        const apiCxn = new apiConnection_1.ApiConnection();
        const getSystemRoles = new emass_client_1.SystemRolesApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        if (args.name === 'all') {
            // Order is important here
            getSystemRoles.getSystemRoles().then((response) => {
                console.log((0, json_colorizer_1.colorize)((0, outputFormatter_1.outputFormat)(response)));
            }).catch((error) => (0, utilities_1.displayError)(error, 'Roles'));
        }
        else if (args.name === 'byCategory') {
            // Order is important here
            getSystemRoles.getSystemRolesByCategoryId(flags.roleCategory, flags.role, flags.policy, flags.includeDecommissioned).then((response) => {
                console.log((0, json_colorizer_1.colorize)((0, outputFormatter_1.outputFormat)(response)));
            }).catch((error) => (0, utilities_1.displayError)(error, 'Roles'));
        }
        else {
            throw this.error;
        }
    }
    // skipcq: JS-0116 - Base class (CommandError) expects expected catch to be async
    async catch(error) {
        if (error instanceof Error) {
            this.warn(error.message);
        }
        else {
            const suggestions = 'get roles [-h or --help]\n\tget roles all\n\tget roles byCategory';
            this.warn('Invalid arguments\nTry this 👇:\n\t' + suggestions);
        }
    }
}
exports.default = EmasserGetRoles;
