// @mitre/saf/lib/commands/emasser/get/workflow_instances.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const json_colorizer_1 = tslib_1.__importDefault(require("json-colorizer"));
const core_1 = require("@oclif/core");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const emass_client_1 = require("@mitre/emass_client");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const outputError_1 = require("../../../utils/emasser/outputError");
const utilities_1 = require("../../../utils/emasser/utilities");
const endpoint = 'workflow_instances';
class EmasserGetWorkflowInstances extends core_1.Command {
    static usage = '<%= command.id %> [ARGUMENT] \n \x1B[93m NOTE: see EXAMPLES for argument case format\x1B[0m';
    static description = (0, utilities_1.getDescriptionForEndpoint)(process.argv, endpoint);
    static examples = (0, utilities_1.getExamplesForEndpoint)(process.argv);
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Show emasser CLI help for the GET Workflow Instances endpoint' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    // NOTE: The way args are being implemented are mainly for clarity purposes, there is, it displays
    //       the available arguments with associate description.
    // Only args.name is used, there is, it contains the argument listed by the user.
    // Example: If the user uses the command (saf emasser get workflow_instances byInstanceId), args.name is set to byInstanceId
    static args = {
        name: core_1.Args.string({ name: 'name', required: false, hidden: true }),
        all: core_1.Args.string({ name: 'all', description: 'Retrieves all workflow instances in a site', required: false }),
        byInstanceId: core_1.Args.string({ name: 'byInstanceId', description: 'Retrieves workflow(s) instance by ID', required: false }),
    };
    async run() {
        const { args, flags } = await this.parse(EmasserGetWorkflowInstances);
        const apiCxn = new apiConnection_1.ApiConnection();
        const getWorkflowInstances = new emass_client_1.WorkflowInstancesApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        if (args.name === 'all') {
            // Order is important here
            getWorkflowInstances.getSystemWorkflowInstances(flags.includeComments, flags.includeDecommissionSystems, flags.pageIndex, flags.sinceDate, flags.status).then((response) => {
                console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
            }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
        }
        else if (args.name === 'byInstanceId') {
            // Order is important here
            getWorkflowInstances.getSystemWorkflowInstancesByWorkflowInstanceId(flags.workflowInstanceId).then((response) => {
                console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
            }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
        }
        else {
            throw this.error;
        }
    }
    async catch(error) {
        if (error.message) {
            this.error(error);
        }
        else {
            const suggestions = 'get workflow_instances [-h or --help]\n\tget workflow_instances all\n\tget workflow_instances byInstanceId';
            this.warn('Invalid arguments\nTry this:\n\t' + suggestions);
        }
    }
}
exports.default = EmasserGetWorkflowInstances;
