// @oclif/core/lib/parser/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FailedFlagValidationError = exports.ArgInvalidOptionError = exports.FlagInvalidOptionError = exports.NonExistentFlagsError = exports.UnexpectedArgsError = exports.RequiredFlagError = exports.RequiredArgsError = exports.InvalidArgsSpecError = exports.CLIParseError = exports.CLIError = void 0;
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const list_1 = require("../cli-ux/list");
const errors_1 = require("../errors");
const util_1 = require("../util/util");
const help_1 = require("./help");
var errors_2 = require("../errors");
Object.defineProperty(exports, "CLIError", { enumerable: true, get: function () { return errors_2.CLIError; } });
class CLIParseError extends errors_1.CLIError {
    parse;
    constructor(options) {
        options.message += '\nSee more help with --help';
        super(options.message);
        this.parse = options.parse;
    }
}
exports.CLIParseError = CLIParseError;
class InvalidArgsSpecError extends CLIParseError {
    args;
    constructor({ args, parse }) {
        let message = 'Invalid argument spec';
        const namedArgs = Object.values(args).filter((a) => a.name);
        if (namedArgs.length > 0) {
            const list = (0, list_1.renderList)(namedArgs.map((a) => [`${a.name} (${a.required ? 'required' : 'optional'})`, a.description]));
            message += `:\n${list}`;
        }
        super({ message, parse });
        this.args = args;
    }
}
exports.InvalidArgsSpecError = InvalidArgsSpecError;
class RequiredArgsError extends CLIParseError {
    args;
    constructor({ args, flagsWithMultiple, parse, }) {
        let message = `Missing ${args.length} required arg${args.length === 1 ? '' : 's'}`;
        const namedArgs = args.filter((a) => a.name);
        if (namedArgs.length > 0) {
            const list = (0, list_1.renderList)(namedArgs.map((a) => [a.name, a.description]));
            message += `:\n${list}`;
        }
        if (flagsWithMultiple?.length) {
            const flags = flagsWithMultiple.map((f) => `--${f}`).join(', ');
            message += `\n\nNote: ${flags} allow${flagsWithMultiple.length === 1 ? 's' : ''} multiple values. Because of this you need to provide all arguments before providing ${flagsWithMultiple.length === 1 ? 'that flag' : 'those flags'}.`;
            message += '\nAlternatively, you can use "--" to signify the end of the flags and the beginning of arguments.';
        }
        super({ message, parse });
        this.args = args;
    }
}
exports.RequiredArgsError = RequiredArgsError;
class RequiredFlagError extends CLIParseError {
    flag;
    constructor({ flag, parse }) {
        const usage = (0, list_1.renderList)((0, help_1.flagUsages)([flag], { displayRequired: false }));
        const message = `Missing required flag:\n${usage}`;
        super({ message, parse });
        this.flag = flag;
    }
}
exports.RequiredFlagError = RequiredFlagError;
class UnexpectedArgsError extends CLIParseError {
    args;
    constructor({ args, parse }) {
        const message = `Unexpected argument${args.length === 1 ? '' : 's'}: ${args.join(', ')}`;
        super({ message, parse });
        this.args = args;
    }
}
exports.UnexpectedArgsError = UnexpectedArgsError;
class NonExistentFlagsError extends CLIParseError {
    flags;
    constructor({ flags, parse }) {
        const message = `Nonexistent flag${flags.length === 1 ? '' : 's'}: ${flags.join(', ')}`;
        super({ message, parse });
        this.flags = flags;
    }
}
exports.NonExistentFlagsError = NonExistentFlagsError;
class FlagInvalidOptionError extends CLIParseError {
    constructor(flag, input) {
        const message = `Expected --${flag.name}=${input} to be one of: ${flag.options.join(', ')}`;
        super({ message, parse: {} });
    }
}
exports.FlagInvalidOptionError = FlagInvalidOptionError;
class ArgInvalidOptionError extends CLIParseError {
    constructor(arg, input) {
        const message = `Expected ${input} to be one of: ${arg.options.join(', ')}`;
        super({ message, parse: {} });
    }
}
exports.ArgInvalidOptionError = ArgInvalidOptionError;
class FailedFlagValidationError extends CLIParseError {
    constructor({ failed, parse }) {
        const reasons = failed.map((r) => r.reason);
        const deduped = (0, util_1.uniq)(reasons);
        const errString = deduped.length === 1 ? 'error' : 'errors';
        const message = `The following ${errString} occurred:\n  ${chalk_1.default.dim(deduped.join('\n  '))}`;
        super({ message, parse });
    }
}
exports.FailedFlagValidationError = FailedFlagValidationError;
