// @mitre/saf/lib/commands/emasser/put/artifacts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const json_colorizer_1 = tslib_1.__importDefault(require("json-colorizer"));
const core_1 = require("@oclif/core");
const emass_client_1 = require("@mitre/emass_client");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const utilities_1 = require("../../../utils/emasser/utilities");
const outputError_1 = require("../../../utils/emasser/outputError");
class EmasserPutArtifacts extends core_1.Command {
    static usage = '<%= command.id %> [options]';
    static description = 'Updates artifacts for a system with provided entries';
    static examples = ['<%= config.bin %> <%= command.id %> [-s,--systemId] [-f,--filename] [--isTemplate,--no-isTemplate] [-t,--type] [-g--category] [options]'];
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Put (update) one or many artifacts in a system' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    async run() {
        const { flags } = await this.parse(EmasserPutArtifacts);
        const apiCxn = new apiConnection_1.ApiConnection();
        const artifactApi = new emass_client_1.ArtifactsApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        const requestBodyArray = [];
        requestBodyArray.push({
            filename: flags.filename,
            isTemplate: flags.isTemplate,
            type: flags.type,
            category: flags.category,
            // Optional arguments
            description: flags.description,
            referencePageNumber: flags.referencePageNumber,
            ccis: flags.ccis,
            controls: flags.controls,
            expirationDate: Number.parseFloat(flags.expirationDate),
            lastReviewedDate: Number.parseFloat(flags.lastReviewDate),
        });
        artifactApi.updateArtifactBySystemId(flags.systemId, requestBodyArray).then((response) => {
            console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response, false)));
        }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
    }
}
exports.default = EmasserPutArtifacts;
