// @mitre/saf/lib/commands/emasser/put/hardware_baseline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const promises_1 = require("fs/promises");
const json_colorizer_1 = require("json-colorizer");
const core_1 = require("@oclif/core");
const utilities_1 = require("../../../utils/emasser/utilities");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const emass_client_1 = require("@mitre/emass_client");
function getAllJsonExamples() {
    return {
        ...(0, utilities_1.getJsonExamples)('hardware-put-required'),
        ...(0, utilities_1.getJsonExamples)('hardware-post-put-conditional'),
        ...(0, utilities_1.getJsonExamples)('hardware-post-put-optional'),
    };
}
/**
 * Asserts that a required parameter exists.
 *
 * @param object - The name of the parameter or field being checked.
 * @param value - The value of the parameter or field. Can be a string, undefined, or null.
 * @throws Will throw an error if the value is undefined.
 */
function assertParamExists(object, value) {
    if (value === undefined) {
        (0, utilities_1.printRedMsg)(`Missing required parameter/field: ${object}`);
        throw new Error('Value not defined');
    }
}
/**
 * Adds required fields to the request body for a hardware object.
 *
 * This function ensures that the required fields `hardwareId` and `assetName`
 * are present in the provided `dataObj`. If any of these fields are missing,
 * an error is thrown and an example JSON structure is logged.
 *
 * @param dataObj - The hardware object containing the data to be validated and added to the request body.
 * @returns A new hardware object containing only the required fields.
 * @throws Will throw an error if `hardwareId` or `assetName` are missing from `dataObj`.
 */
function addRequiredFieldsToRequestBody(dataObj) {
    const bodyObj = {};
    try {
        assertParamExists('hardwareId', dataObj.hardwareId);
        assertParamExists('assetName', dataObj.assetName);
    }
    catch (error) {
        console.log('Required JSON fields are:');
        console.log((0, json_colorizer_1.colorize)(JSON.stringify((0, utilities_1.getJsonExamples)('hardware-put-required'), null, 2)));
        throw error;
    }
    // The required parameter "systemId" is validated by oclif
    bodyObj.hardwareId = dataObj.hardwareId;
    bodyObj.assetName = dataObj.assetName;
    return bodyObj;
}
/**
 * Adds conditional fields from the `dataObj` to the `bodyObject` if they exist.
 *
 * @param bodyObject - The target object to which fields will be added.
 * @param dataObj - The source object from which fields will be copied if they exist.
 */
function addConditionalFields(bodyObject, dataObj) {
    if (Object.prototype.hasOwnProperty.call(dataObj, 'publicFacingFqdn')) {
        bodyObject.publicFacingFqdn = dataObj.publicFacingFqdn;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'publicFacingIpAddress')) {
        bodyObject.publicFacingIpAddress = dataObj.publicFacingIpAddress;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'publicFacingUrls')) {
        bodyObject.publicFacingUrls = dataObj.publicFacingUrls;
    }
}
/**
 * Adds optional fields from the data object to the body object if they exist.
 *
 * @param bodyObject - The target object to which optional fields will be added.
 * @param dataObj - The source object containing optional fields.
 */
function addOptionalFields(bodyObject, dataObj) {
    if (Object.prototype.hasOwnProperty.call(dataObj, 'componentType')) {
        bodyObject.componentType = dataObj.componentType;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'nickname')) {
        bodyObject.nickname = dataObj.nickname;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'assetIpAddress')) {
        bodyObject.assetIpAddress = dataObj.assetIpAddress;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'publicFacing')) {
        bodyObject.publicFacing = dataObj.publicFacing;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'virtualAsset')) {
        bodyObject.virtualAsset = dataObj.virtualAsset;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'manufacturer')) {
        bodyObject.manufacturer = dataObj.manufacturer;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'modelNumber')) {
        bodyObject.modelNumber = dataObj.modelNumber;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'serialNumber')) {
        bodyObject.serialNumber = dataObj.serialNumber;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'osIosFwVersion')) {
        bodyObject.osIosFwVersion = dataObj.osIosFwVersion;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'memorySizeType')) {
        bodyObject.memorySizeType = dataObj.memorySizeType;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'location')) {
        bodyObject.location = dataObj.location;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'approvalStatus')) {
        bodyObject.approvalStatus = dataObj.approvalStatus;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'criticalAsset')) {
        bodyObject.criticalAsset = dataObj.criticalAsset;
    }
}
/**
 * Generates a body object for a hardware request by adding required, conditional, and optional fields.
 *
 * @param dataObject - The hardware data object to be processed.
 * @returns The generated body object with the necessary fields.
 * @throws Will exit the process with code 1 if an error occurs during the generation.
 */
function generateBodyObj(dataObject) {
    let bodyObj = {};
    try {
        bodyObj = addRequiredFieldsToRequestBody(dataObject);
        addConditionalFields(bodyObj, dataObject);
        addOptionalFields(bodyObj, dataObject);
    }
    catch {
        process.exit(1);
    }
    return bodyObj;
}
const CMD_HELP = 'saf emasser post hardware_baseline -h or --help';
class EmasserHardwareBaseline extends core_1.Command {
    static usage = '<%= command.id %> [FLAGS]\n\x1B[93m NOTE: see EXAMPLES for command usages\x1B[0m';
    static description = 'Update one or many hardware assets to a system.\n'
        + 'The CLI expects an input JSON file containing the required, conditional\n'
        + 'and optional fields for the hardware asset(s) being added to the system.';
    static examples = [
        '<%= config.bin %> <%= command.id %> [-s,--systemId] [-f,--dataFile]',
        'The input file should be a well formed JSON containing Hardware Assets.',
        '\x1B[1mRequired JSON parameter/field is:\x1B[0m',
        (0, json_colorizer_1.colorize)(JSON.stringify((0, utilities_1.getJsonExamples)('hardware-put-required'), null, 2)),
        '\x1B[1mConditional JSON parameters/fields are:\x1B[0m',
        (0, json_colorizer_1.colorize)(JSON.stringify((0, utilities_1.getJsonExamples)('hardware-post-put-conditional'), null, 2)),
        '\x1B[1mOptional JSON parameters/fields are:\x1B[0m',
        (0, json_colorizer_1.colorize)(JSON.stringify((0, utilities_1.getJsonExamples)('hardware-post-put-optional'), null, 2)),
        '\x1B[1m\x1B[32mAll accepted parameters/fields are:\x1B[0m',
        (0, json_colorizer_1.colorize)(getAllJsonExamples()),
    ];
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Show eMASSer CLI help for the PUT Hardware Baseline command' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    async run() {
        const { flags } = await this.parse(EmasserHardwareBaseline);
        const apiCxn = new apiConnection_1.ApiConnection();
        const hwBaseline = new emass_client_1.HardwareBaselineApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        const requestBodyArray = [];
        // Check if a Hardware json file was provided
        if (!fs_1.default.existsSync(flags.dataFile)) {
            console.error('\x1B[91m» Hardware data file (.json) not found or invalid:', flags.dataFile, '\x1B[0m');
            process.exit(1);
        }
        try {
            // Read and parse the JSON file
            const fileContent = await (0, promises_1.readFile)(flags.dataFile, 'utf8');
            const data = JSON.parse(fileContent);
            // Security Control information json file provided, check if we have multiple content to process
            if (Array.isArray(data)) {
                data.forEach((dataObject) => {
                    // Generate the put request object
                    requestBodyArray.push(generateBodyObj(dataObject));
                });
            }
            else if (typeof data === 'object' && data !== null) {
                const dataObject = data;
                // Generate the put request object
                requestBodyArray.push(generateBodyObj(dataObject));
            }
            else {
                console.error('\x1B[91m» Invalid data format in Hardware Baseline file\x1B[0m');
                process.exit(1);
            }
        }
        catch (error) {
            if (error instanceof Error) {
                console.error('\x1B[91m» Error reading Hardware Baseline data file, possible malformed json. Please use the -h flag for help.\x1B[0m');
                console.error('\x1B[93m→ Error message was:', error.message, '\x1B[0m');
            }
            else {
                console.error('\x1B[91m» Unknown error occurred while reading the file:', flags.dataFile, '\x1B[0m');
            }
            process.exit(1);
        }
        // Call the endpoint
        hwBaseline.updateHwBaselineAssets(flags.systemId, requestBodyArray).then((response) => {
            console.log((0, json_colorizer_1.colorize)((0, outputFormatter_1.outputFormat)(response, false)));
        }).catch((error) => (0, utilities_1.displayError)(error, 'Hardware Baseline'));
    }
    // skipcq: JS-0116 - Base class (CommandError) expects expected catch to return a Promise
    async catch(err) {
        // If error message is for missing flags, display
        // what fields are required, otherwise show the error
        if (err.message.includes('See more help with --help')) {
            this.warn(err.message.replace('with --help', `with: \x1B[93m${CMD_HELP}\x1B[0m`));
        }
        else {
            this.warn(err);
        }
    }
}
exports.default = EmasserHardwareBaseline;
