// @mitre/saf/lib/commands/emasser/put/milestones.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const json_colorizer_1 = tslib_1.__importDefault(require("json-colorizer"));
const core_1 = require("@oclif/core");
const outputError_1 = require("../../../utils/emasser/outputError");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const utilities_1 = require("../../../utils/emasser/utilities");
const emass_client_1 = require("@mitre/emass_client");
class EmasserPutMilestones extends core_1.Command {
    static usage = '<%= command.id %> [options]';
    static description = 'Update milestone(s) for specified system, poam, and milestone Id';
    static examples = ['<%= config.bin %> <%= command.id %> [-s,--systemId] [-p,--poamId] [-m,--milestoneId] [-d,--description] [-c,--scheduledCompletionDate]'];
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Show emasser CLI help for the PUT Milestones endpoint' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    async run() {
        const { flags } = await this.parse(EmasserPutMilestones);
        const apiCxn = new apiConnection_1.ApiConnection();
        const putMilestones = new emass_client_1.MilestonesApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        const requestBodyArray = [];
        requestBodyArray.push({
            milestoneId: flags.milestoneId,
            description: flags.description,
            scheduledCompletionDate: Number.parseFloat(flags.scheduledCompletionDate),
        });
        putMilestones.updateMilestoneBySystemIdAndPoamId(flags.systemId, flags.poamId, requestBodyArray).then((response) => {
            console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response, false)));
        }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
    }
}
exports.default = EmasserPutMilestones;
