// @mitre/saf/lib/commands/emasser/put/poams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const promises_1 = require("fs/promises");
const json_colorizer_1 = require("json-colorizer");
const core_1 = require("@oclif/core");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const utilities_1 = require("../../../utils/emasser/utilities");
const emass_client_1 = require("@mitre/emass_client");
function getAllJsonExamples() {
    return {
        ...(0, utilities_1.getJsonExamples)('poams-put-required'),
        ...(0, utilities_1.getJsonExamples)('poams-post-put-required-va'),
        ...(0, utilities_1.getJsonExamples)('poams-put-conditional'),
        ...(0, utilities_1.getJsonExamples)('poams-post-put-optional'),
    };
}
/**
 * Asserts that a required parameter exists and is not undefined.
 *
 * @param object - The name of the parameter or field being checked.
 * @param value - The value of the parameter or field to check. Can be a string, number, undefined, or null.
 * @throws {Error} Throws an error if the value is undefined.
 */
function assertParamExists(object, value) {
    if (value === undefined) {
        (0, utilities_1.printRedMsg)(`Missing required parameter/field: ${object}`);
        throw new Error('Value not defined');
    }
}
/**
 * Adds required fields to the request body for a POAMs (Plan of Action and Milestones) object.
 *
 * This function ensures that all required fields are present in the input `dataObj` and then
 * constructs a new `Poams` object with these fields. If any required field is missing, an error
 * is thrown, and an example of the required JSON structure is logged.
 *
 * @param {Poams} dataObj - The input POAMs object containing the data to be validated and added to the request body.
 * @returns {Poams} - A new POAMs object containing the required fields from the input `dataObj`.
 * @throws Will throw an error if any required field is missing in the input `dataObj`.
 */
function addRequiredFieldsToRequestBody(dataObj) {
    const bodyObj = {};
    try {
        assertParamExists('poamId', dataObj.poamId);
        assertParamExists('displayPoamId', dataObj.displayPoamId);
        assertParamExists('status', dataObj.status);
        assertParamExists('vulnerabilityDescription', dataObj.vulnerabilityDescription);
        assertParamExists('sourceIdentifyingVulnerability', dataObj.sourceIdentifyingVulnerability);
        assertParamExists('pocOrganization', dataObj.pocOrganization);
        assertParamExists('resources', dataObj.resources);
    }
    catch (error) {
        console.log('Required JSON fields are:');
        console.log((0, json_colorizer_1.colorize)(JSON.stringify((0, utilities_1.getJsonExamples)('poams-put-required'), null, 2)));
        throw error;
    }
    bodyObj.poamId = dataObj.poamId;
    bodyObj.displayPoamId = dataObj.displayPoamId;
    bodyObj.status = dataObj.status;
    bodyObj.vulnerabilityDescription = dataObj.vulnerabilityDescription;
    bodyObj.sourceIdentifyingVulnerability = dataObj.sourceIdentifyingVulnerability;
    bodyObj.pocOrganization = dataObj.pocOrganization;
    bodyObj.resources = dataObj.resources;
    return bodyObj;
}
/**
 * Adds conditional fields from the data object to the body object if they exist.
 *
 * @param bodyObject - The target object to which fields will be added.
 * @param dataObj - The source object from which fields will be copied.
 *
 * @remarks
 * This function checks if the specified properties ('pocFirstName', 'pocLastName',
 * 'pocEmail', 'pocPhoneNumber', 'severity') exist in the data object. If they do,
 * it copies their values to the corresponding properties in the body object.
 */
function addConditionalFields(bodyObject, dataObj) {
    if (Object.prototype.hasOwnProperty.call(dataObj, 'pocFirstName')) {
        bodyObject.pocFirstName = dataObj.pocFirstName;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'pocLastName')) {
        bodyObject.pocLastName = dataObj.pocLastName;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'pocEmail')) {
        bodyObject.pocEmail = dataObj.pocEmail;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'pocPhoneNumber')) {
        bodyObject.pocPhoneNumber = dataObj.pocPhoneNumber;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'severity')) {
        bodyObject.severity = dataObj.severity;
    }
}
/**
 * Adds optional fields from the data object to the body object if they exist.
 *
 * @param bodyObject - The target object to which optional fields will be added.
 * @param dataObj - The source object containing optional fields.
 */
// skipcq: JS-R1005 - Ignore Function cyclomatic complexity high threshold
function addOptionalFields(bodyObject, dataObj) {
    if (Object.prototype.hasOwnProperty.call(dataObj, 'externalUid')) {
        bodyObject.externalUid = dataObj.externalUid;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'controlAcronym')) {
        bodyObject.controlAcronym = dataObj.controlAcronym;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'assessmentProcedure')) {
        bodyObject.assessmentProcedure = dataObj.assessmentProcedure;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'securityChecks')) {
        bodyObject.securityChecks = dataObj.securityChecks;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'rawSeverity')) {
        bodyObject.rawSeverity = dataObj.rawSeverity;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'relevanceOfThreat')) {
        bodyObject.relevanceOfThreat = dataObj.relevanceOfThreat;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'likelihood')) {
        bodyObject.likelihood = dataObj.likelihood;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'impact')) {
        bodyObject.impact = dataObj.impact;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'impactDescription')) {
        bodyObject.impactDescription = dataObj.impactDescription;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'residualRiskLevel')) {
        bodyObject.residualRiskLevel = dataObj.residualRiskLevel;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'recommendations')) {
        bodyObject.recommendations = dataObj.recommendations;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'mitigations')) {
        bodyObject.mitigations = dataObj.mitigations;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'resultingResidualRiskLevelAfterProposedMitigations')) {
        bodyObject.resultingResidualRiskLevelAfterProposedMitigations = dataObj.resultingResidualRiskLevelAfterProposedMitigations;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'predisposingConditions')) {
        bodyObject.predisposingConditions = dataObj.predisposingConditions;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'threatDescription')) {
        bodyObject.threatDescription = dataObj.threatDescription;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'devicesAffected')) {
        bodyObject.devicesAffected = dataObj.devicesAffected;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'identifiedInCFOAuditOrOtherReview')) {
        bodyObject.identifiedInCFOAuditOrOtherReview = dataObj.identifiedInCFOAuditOrOtherReview;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'personnelResourcesFundedBaseHours')) {
        bodyObject.personnelResourcesFundedBaseHours = dataObj.personnelResourcesFundedBaseHours;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'personnelResourcesCostCode')) {
        bodyObject.personnelResourcesCostCode = dataObj.personnelResourcesCostCode;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'personnelResourcesUnfundedBaseHours')) {
        bodyObject.personnelResourcesUnfundedBaseHours = dataObj.personnelResourcesUnfundedBaseHours;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'personnelResourcesNonfundingObstacle')) {
        bodyObject.personnelResourcesNonfundingObstacle = dataObj.personnelResourcesNonfundingObstacle;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'personnelResourcesNonfundingObstacleOtherReason')) {
        bodyObject.personnelResourcesNonfundingObstacleOtherReason = dataObj.personnelResourcesNonfundingObstacleOtherReason;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'nonPersonnelResourcesFundedAmount')) {
        bodyObject.nonPersonnelResourcesFundedAmount = dataObj.nonPersonnelResourcesFundedAmount;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'nonPersonnelResourcesCostCode')) {
        bodyObject.nonPersonnelResourcesCostCode = dataObj.nonPersonnelResourcesCostCode;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'nonPersonnelResourcesUnfundedAmount')) {
        bodyObject.nonPersonnelResourcesUnfundedAmount = dataObj.nonPersonnelResourcesUnfundedAmount;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'nonPersonnelResourcesNonfundingObstacle')) {
        bodyObject.nonPersonnelResourcesNonfundingObstacle = dataObj.nonPersonnelResourcesNonfundingObstacle;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'nonPersonnelResourcesNonfundingObstacleOtherReason')) {
        bodyObject.nonPersonnelResourcesNonfundingObstacleOtherReason = dataObj.nonPersonnelResourcesNonfundingObstacleOtherReason;
    }
}
/**
 * Processes business logic for POA&M items based on their status and validates required fields.
 *
 * @param bodyObject - The object to populate with validated data.
 * @param dataObj - The object containing input data to validate and process.
 *
 * The function performs the following checks based on the `status` field in `dataObj`:
 *
 * - "Risk Accepted":
 *   - Requires `comments`.
 *   - Cannot have `scheduledCompletionDate` or `milestones`.
 *
 * - "Ongoing":
 *   - Requires at least one `milestones` object.
 *   - Each `milestone` must have `description` and `scheduledCompletionDate`.
 *   - Optionally includes `scheduledCompletionDate`.
 *
 * - "Completed":
 *   - Requires `comments`, `completionDate`, and at least one `milestones` object.
 *   - Optionally includes `scheduledCompletionDate`.
 *
 * - "Not Applicable":
 *   - Cannot be created.
 *
 * - "Archived":
 *   - Cannot be updated.
 *
 * Additionally, if any POC (Point of Contact) fields are provided, all POC fields are required:
 * - `pocFirstName`
 * - `pocLastName`
 * - `pocEmail`
 * - `pocPhoneNumber`
 *
 * The function uses `printRedMsg` to display error messages and `printHelpMsg` to display help messages.
 * It exits the process with a status code of 1 if any validation fails.
*/
// skipcq: JS-R1005 - Ignore Function cyclomatic complexity high threshold
function processBusinessLogic(bodyObject, dataObj) {
    const HELP_MSG = 'Invoke saf emasser post poams [-h, --help] for additional help';
    switch (dataObj.status) {
        case 'Risk Accepted': {
            if (dataObj.comments === undefined) {
                (0, utilities_1.printRedMsg)('When status is "Risk Accepted" the following parameter/field is required:');
                (0, utilities_1.printRedMsg)('    comments');
                (0, utilities_1.printHelpMsg)(HELP_MSG);
                process.exit(1);
                // Risk Accepted POA&M Item cannot be saved with a Scheduled Completion Date or Milestones.
            }
            else if (Object.prototype.hasOwnProperty.call(dataObj, 'scheduledCompletionDate') || Object.prototype.hasOwnProperty.call(dataObj, 'milestones')) {
                (0, utilities_1.printRedMsg)('When status is "Risk Accepted" POA&Ms CAN NOT be saved with the following parameters/fields:');
                (0, utilities_1.printRedMsg)('    scheduledCompletionDate, or milestone');
                (0, utilities_1.printHelpMsg)(HELP_MSG);
                process.exit(1);
            }
            else {
                bodyObject.comments = dataObj.comments;
            }
            break;
        }
        case 'Ongoing': {
            // Can not update scheduledCompletionDate if review status (reviewStatus is a readonly field) is Approved
            // API call will return and error if scheduledCompletionDate for Approved review status POA&Ms
            // POA&M Items that have a status of “Ongoing” cannot be saved without Milestones.
            if (!(Object.prototype.hasOwnProperty.call(dataObj, 'milestones'))) {
                (0, utilities_1.printRedMsg)('When status is "Ongoing" and updating a POA&M a Milestone is required');
                (0, utilities_1.printHelpMsg)(HELP_MSG);
                process.exit(1);
                // If we have a milestone, ensure the required fields are provided.
            }
            else if (!(lodash_1.default.some(dataObj.milestones, 'description')) || !(lodash_1.default.some(dataObj.milestones, 'scheduledCompletionDate'))) {
                (0, utilities_1.printRedMsg)('Milestone object requires the following fields:');
                (0, utilities_1.printRedMsg)('    "milestones": [{"description": "The milestone description", "scheduledCompletionDate": Unix date format }], ');
                process.exit(1);
            }
            else {
                // Add the POA&M schedule completion date if provided (backend may return an error)
                if (Object.prototype.hasOwnProperty.call(dataObj, 'scheduledCompletionDate')) {
                    bodyObject.scheduledCompletionDate = dataObj.scheduledCompletionDate;
                }
                // Add the milestone object
                const milestoneArray = [];
                dataObj.milestones?.forEach((milestone) => {
                    const milestoneObj = { description: '', scheduledCompletionDate: 0, isActive: false };
                    milestoneObj.description = milestone.description;
                    milestoneObj.scheduledCompletionDate = milestone.scheduledCompletionDate;
                    // isActive is used to prevent uploading duplicate/undesired milestones via the
                    // POA&M PUT call and must be set to false, otherwise a new milestone is created
                    if (Object.prototype.hasOwnProperty.call(milestone, 'isActive')) {
                        milestoneObj.isActive = milestone.isActive;
                    }
                    milestoneArray.push(milestoneObj);
                });
                bodyObject.milestones = [...milestoneArray];
            }
            break;
        }
        case 'Completed': {
            // Completed POA&M Item require the completionDate, comments, and Milestones.
            if (!(Object.prototype.hasOwnProperty.call(dataObj, 'comments'))
                || !(Object.prototype.hasOwnProperty.call(dataObj, 'completionDate'))
                || !(Object.prototype.hasOwnProperty.call(dataObj, 'milestones'))) {
                (0, utilities_1.printRedMsg)('When status is "Completed" the following parameters/fields are required:');
                (0, utilities_1.printRedMsg)('    comments, completionDate, and milestones');
                (0, utilities_1.printHelpMsg)(HELP_MSG);
                process.exit(1);
            }
            else {
                // Add the POA&M schedule and completion date, comments
                bodyObject.comments = dataObj.comments;
                bodyObject.completionDate = dataObj.completionDate;
                // Add the POA&M schedule completion date if provided (backend may return an error)
                if (Object.prototype.hasOwnProperty.call(dataObj, 'scheduledCompletionDate')) {
                    bodyObject.scheduledCompletionDate = dataObj.scheduledCompletionDate;
                }
                // Add the milestone object
                const milestoneArray = [];
                dataObj.milestones?.forEach((milestone) => {
                    const milestoneObj = { description: '', scheduledCompletionDate: 0, isActive: false };
                    milestoneObj.description = milestone.description;
                    milestoneObj.scheduledCompletionDate = milestone.scheduledCompletionDate;
                    // isActive is used to prevent uploading duplicate/undesired milestones via the
                    // POA&M PUT call and must be set to false, otherwise a new milestone is created
                    if (Object.prototype.hasOwnProperty.call(milestone, 'isActive')) {
                        milestoneObj.isActive = milestone.isActive;
                    }
                    milestoneArray.push(milestoneObj);
                });
                bodyObject.milestones = [...milestoneArray];
            }
            break;
        }
        case 'Not Applicable': {
            (0, utilities_1.printRedMsg)('POA&M Items with a status of "Not Applicable" will be updated through test result creation.');
            process.exit(0);
            break;
        }
        case 'Archived': {
            (0, utilities_1.printHelpMsg)('Archived POA&M Items cannot be updated');
            process.exit(0);
            break;
        }
        default: {
            (0, utilities_1.printRedMsg)('The "status" field must one of the following:');
            (0, utilities_1.printRedMsg)('    Risk Accepted, Ongoing, or Completed');
            (0, utilities_1.printRedMsg)(`Status provided was: ${dataObj.status}`);
            process.exit(1);
            break;
        }
    }
    // POC checks: If any poc information is provided all POC fields are required
    let missingFields = '';
    if ((lodash_1.default.get(dataObj, 'pocFirstName') === undefined))
        missingFields = 'pocFirstName';
    if ((lodash_1.default.get(dataObj, 'pocLastName') === undefined))
        missingFields += (missingFields === '') ? 'pocLastName' : ', pocLastName';
    if ((lodash_1.default.get(dataObj, 'pocEmail') === undefined))
        missingFields += (missingFields === '') ? 'pocEmail' : ', pocEmail';
    if ((lodash_1.default.get(dataObj, 'pocPhoneNumber') === undefined))
        missingFields += (missingFields === '') ? 'pocPhoneNumber' : ', pocPhoneNumber';
    const totalPocMissingFields = missingFields.split(',').length;
    if ((totalPocMissingFields >= 1 && totalPocMissingFields < 4) && missingFields !== '') {
        (0, utilities_1.printRedMsg)('If any POC fields are provided (pocFirstName, pocLastName, pocEmail, pocPhoneNumber) than all POC fields are required:');
        (0, utilities_1.printRedMsg)(`    Missing field(s): ${missingFields}`);
        (0, utilities_1.printHelpMsg)(HELP_MSG);
        process.exit(1);
    }
}
/**
 * Generates a new Poams object by processing the given dataObject.
 *
 * This function performs the following steps:
 * 1. Adds required fields to the request body.
 * 2. Processes business logic on the body object.
 * 3. Adds conditional fields based on the dataObject.
 * 4. Adds optional fields to the body object.
 *
 * If any error occurs during these steps, the process exits with code 1.
 *
 * @param dataObject - The input Poams object to be processed.
 * @returns The newly generated Poams object with the required, conditional, and optional fields.
 */
function generateBodyObj(dataObject) {
    let bodyObj = {};
    try {
        bodyObj = addRequiredFieldsToRequestBody(dataObject);
        processBusinessLogic(bodyObj, dataObject);
        addConditionalFields(bodyObj, dataObject);
        addOptionalFields(bodyObj, dataObject);
    }
    catch {
        process.exit(1);
    }
    return bodyObj;
}
const CMD_HELP = 'saf emasser put poams -h or --help';
class EmasserPutPoams extends core_1.Command {
    static usage = '<%= command.id %> [FLAGS]\n\x1B[93m NOTE: see EXAMPLES for command usages\x1B[0m';
    static description = 'Update a Plan of Action and Milestones (POA&M) into a systems.';
    static examples = [
        '<%= config.bin %> <%= command.id %> [-s,--systemId] [-f,--dataFile]',
        'The input file should be a well formed JSON containing the POA&M information based on defined business rules.',
        'Required JSON parameter/fields are: ',
        (0, json_colorizer_1.colorize)(JSON.stringify((0, utilities_1.getJsonExamples)('poams-put-required'), null, 2)),
        '\x1B[1mRequired for VA but Conditional for Army and USCG JSON parameters/fields are:\x1B[0m',
        (0, json_colorizer_1.colorize)(JSON.stringify((0, utilities_1.getJsonExamples)('poams-post-put-required-va'), null, 2)),
        'Conditional JSON parameters/fields are: ',
        (0, json_colorizer_1.colorize)(JSON.stringify((0, utilities_1.getJsonExamples)('poams-put-conditional'), null, 2)),
        'Optional JSON parameters/fields are:',
        (0, json_colorizer_1.colorize)(JSON.stringify((0, utilities_1.getJsonExamples)('poams-post-put-optional'), null, 2)),
        '\x1B[1m\x1B[32mAll accepted parameters/fields are:\x1B[0m',
        (0, json_colorizer_1.colorize)(getAllJsonExamples()),
    ];
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Show eMASSer CLI help for the PUT POA&Ms command' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    async run() {
        const { flags } = await this.parse(EmasserPutPoams);
        const apiCxn = new apiConnection_1.ApiConnection();
        const updatePoam = new emass_client_1.POAMApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        const requestBodyArray = [];
        // Check if a POA&Ms json file was provided
        if (!fs_1.default.existsSync(flags.dataFile)) {
            console.error('\x1B[91m» POA&M(s) data file (.json) not found or invalid:', flags.dataFile, '\x1B[0m');
            process.exit(1);
        }
        try {
            // Read and parse the JSON file
            const fileContent = await (0, promises_1.readFile)(flags.dataFile, 'utf8');
            const data = JSON.parse(fileContent);
            // POA&Ms json file provided, check if we have multiple POA&Ms to process
            if (Array.isArray(data)) {
                data.forEach((dataObject) => {
                    // Generate the put request object based on business logic
                    requestBodyArray.push(generateBodyObj(dataObject));
                });
            }
            else if (typeof data === 'object' && data !== null) {
                const dataObject = data;
                // Generate the put request object based on business logic
                requestBodyArray.push(generateBodyObj(dataObject));
            }
            else {
                console.error('\x1B[91m» Invalid data format in POA&Ms file\x1B[0m');
                process.exit(1);
            }
        }
        catch (error) {
            if (error instanceof Error) {
                console.error('\x1B[91m» Error reading POA&Ms data file, possible malformed json. Please use the -h flag for help.\x1B[0m');
                console.error('\x1B[93m→ Error message was:', error.message, '\x1B[0m');
            }
            else {
                console.error('\x1B[91m» Unknown error occurred while reading the file:', flags.dataFile, '\x1B[0m');
            }
            process.exit(1);
        }
        // Call the endpoint
        updatePoam.updatePoamBySystemId(flags.systemId, requestBodyArray).then((response) => {
            console.log((0, json_colorizer_1.colorize)((0, outputFormatter_1.outputFormat)(response)));
        }).catch((error) => (0, utilities_1.displayError)(error, 'POA&Ms'));
    }
    // skipcq: JS-0116 - Base class (CommandError) expects expected catch to return a Promise
    async catch(err) {
        // If error message is for missing flags, display
        // what fields are required, otherwise show the error
        if (err.message.includes('See more help with --help')) {
            this.warn(err.message.replace('with --help', `with: \x1B[93m${CMD_HELP}\x1B[0m`));
        }
        else {
            this.warn(err);
        }
    }
}
exports.default = EmasserPutPoams;
