// @mitre/saf/lib/commands/emasser/put/poams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const json_colorizer_1 = tslib_1.__importDefault(require("json-colorizer"));
const core_1 = require("@oclif/core");
const emass_client_1 = require("@mitre/emass_client");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const utilities_1 = require("../../../utils/emasser/utilities");
const outputError_1 = require("../../../utils/emasser/outputError");
const promises_1 = require("fs/promises");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const fs_1 = tslib_1.__importDefault(require("fs"));
function printRedMsg(msg) {
    console.log('\x1B[91m', msg, '\x1B[0m');
}
function printHelpMsg(msg) {
    console.log('\x1B[93m', msg, '\x1B[0m');
}
function assertParamExists(object, value) {
    if (value === undefined) {
        printRedMsg(`Missing required parameter/field: ${object}`);
        throw new Error('Value not defined');
    }
}
function addRequiredFieldsToRequestBody(dataObj) {
    const bodyObj = {};
    try {
        assertParamExists('poamId', dataObj.poamId);
        assertParamExists('displayPoamId', dataObj.displayPoamId);
        assertParamExists('status', dataObj.status);
        assertParamExists('vulnerabilityDescription', dataObj.vulnerabilityDescription);
        assertParamExists('sourceIdentVuln', dataObj.sourceIdentVuln);
        assertParamExists('pocOrganization', dataObj.pocOrganization);
        assertParamExists('resources', dataObj.resources);
        assertParamExists('mitigation', dataObj.mitigation);
    }
    catch (error) {
        console.log('Required JSON fields are:');
        console.log((0, json_colorizer_1.default)(JSON.stringify((0, utilities_1.getJsonExamples)('poams-put-required'), null, 2)));
        throw error;
    }
    bodyObj.poamId = dataObj.poamId;
    bodyObj.displayPoamId = dataObj.displayPoamId;
    bodyObj.status = dataObj.status;
    bodyObj.vulnerabilityDescription = dataObj.vulnerabilityDescription;
    bodyObj.sourceIdentVuln = dataObj.sourceIdentVuln;
    bodyObj.pocOrganization = dataObj.pocOrganization;
    bodyObj.resources = dataObj.resources;
    bodyObj.mitigation = dataObj.mitigation;
    return bodyObj;
}
function addConditionalFields(bodyObject, dataObj) {
    if (Object.prototype.hasOwnProperty.call(dataObj, 'pocFirstName')) {
        bodyObject.pocFirstName = dataObj.pocFirstName;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'pocLastName')) {
        bodyObject.pocLastName = dataObj.pocLastName;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'pocEmail')) {
        bodyObject.pocEmail = dataObj.pocEmail;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'pocPhoneNumber')) {
        bodyObject.pocPhoneNumber = dataObj.pocPhoneNumber;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'severity')) {
        bodyObject.severity = dataObj.severity;
    }
}
function addOptionalFields(bodyObject, dataObj) {
    if (Object.prototype.hasOwnProperty.call(dataObj, 'externalUid')) {
        bodyObject.externalUid = dataObj.externalUid;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'controlAcronym')) {
        bodyObject.controlAcronym = dataObj.controlAcronym;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'cci')) {
        bodyObject.cci = dataObj.cci;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'securityChecks')) {
        bodyObject.securityChecks = dataObj.securityChecks;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'rawSeverity')) {
        bodyObject.rawSeverity = dataObj.rawSeverity;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'relevanceOfThreat')) {
        bodyObject.relevanceOfThreat = dataObj.relevanceOfThreat;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'likelihood')) {
        bodyObject.likelihood = dataObj.likelihood;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'impact')) {
        bodyObject.impact = dataObj.impact;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'impactDescription')) {
        bodyObject.impactDescription = dataObj.impactDescription;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'residualRiskLevel')) {
        bodyObject.residualRiskLevel = dataObj.residualRiskLevel;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'recommendations')) {
        bodyObject.recommendations = dataObj.recommendations;
    }
    // if (Object.prototype.hasOwnProperty.call(dataObj,'mitigation')) {bodyObject.mitigation = dataObj.mitigation; }
}
function processBusinessLogic(bodyObject, dataObj) {
    //-----------------------------------------------------------------------------
    // Conditional fields that are required based on the "status" field value
    // "Risk Accepted"   comments, resources
    // "Ongoing"         scheduledCompletionDate, resources, milestones (at least 1)
    // "Completed"       scheduledCompletionDate, comments, resources,
    //                   completionDate, milestones (at least 1)
    // "Not Applicable"  POAM can not be created
    //-----------------------------------------------------------------------------
    const HELP_MSG = '\nInvoke saf emasser post poams [-h, --help] for additional help';
    switch (dataObj.status) {
        case 'Risk Accepted': {
            if (dataObj.comments === undefined) {
                printRedMsg('When status is "Risk Accepted" the following parameters/fields are required:');
                printRedMsg('    comments');
                printHelpMsg(HELP_MSG);
                process.exit(1);
            }
            else if (Object.prototype.hasOwnProperty.call(dataObj, 'scheduledCompletionDate') || Object.prototype.hasOwnProperty.call(dataObj, 'milestones')) {
                printRedMsg('When status is "Risk Accepted" POA&Ms CAN NOT be saved with the following parameters/field:');
                printRedMsg('    scheduledCompletionDate, or milestone');
                printHelpMsg(HELP_MSG);
                process.exit(1);
            }
            else {
                bodyObject.comments = dataObj.comments;
            }
            break;
        }
        case 'Ongoing': {
            if (!(Object.prototype.hasOwnProperty.call(dataObj, 'scheduledCompletionDate') && Object.prototype.hasOwnProperty.call(dataObj, 'milestones'))) {
                printRedMsg('When status is "Ongoing" the following parameters/fields are required:');
                printRedMsg('    scheduledCompletionDate, milestones');
                printHelpMsg(HELP_MSG);
                process.exit(1);
            }
            else if (((lodash_1.default.some(dataObj.milestones, function (milestone) {
                return milestone.description;
            }))) ||
                ((lodash_1.default.some(dataObj.milestones, function (milestone) {
                    return milestone.scheduledCompletionDate;
                })))) {
                printRedMsg('When milestones are define the parameters "description" and "scheduledCompletionDate" are required.');
                printRedMsg('Missing one of the required milestones parameters:');
                printRedMsg('    description, scheduledCompletionDate');
                process.exit(1);
            }
            else {
                // Add the POA&M completion date
                bodyObject.scheduledCompletionDate = dataObj.scheduledCompletionDate;
                // Add the milestone object
                const milestoneArray = [];
                dataObj.milestones?.forEach((milestone) => {
                    const milestoneObj = {};
                    milestoneObj.milestoneId = milestone.milestoneId;
                    milestoneObj.description = milestone.description;
                    milestoneObj.scheduledCompletionDate = milestone.scheduledCompletionDate;
                    milestoneArray.push(milestoneObj);
                });
                bodyObject.milestones = [...milestoneArray];
            }
            break;
        }
        case 'Completed': {
            if (!(Object.prototype.hasOwnProperty.call(dataObj, 'scheduledCompletionDate')) || !(Object.prototype.hasOwnProperty.call(dataObj, 'comments')) ||
                !(Object.prototype.hasOwnProperty.call(dataObj, 'completionDate')) || !(Object.prototype.hasOwnProperty.call(dataObj, 'milestones'))) {
                printRedMsg('When status is "Completed" the following parameters/fields are required:');
                printRedMsg('    scheduledCompletionDate, comments, completionDate, or milestone');
                printHelpMsg(HELP_MSG);
                process.exit(1);
            }
            else {
                // Add the POA&M schedule and completion date, comments
                bodyObject.comments = dataObj.comments;
                bodyObject.completionDate = dataObj.completionDate;
                bodyObject.scheduledCompletionDate = dataObj.scheduledCompletionDate;
                // Add the milestone object
                const milestoneArray = [];
                dataObj.milestones?.forEach((milestone) => {
                    const milestoneObj = {};
                    milestoneObj.milestoneId = milestone.milestoneId;
                    milestoneObj.description = milestone.description;
                    milestoneObj.scheduledCompletionDate = milestone.scheduledCompletionDate;
                    milestoneArray.push(milestoneObj);
                });
                bodyObject.milestones = [...milestoneArray];
            }
            break;
        }
        case 'Archived': {
            printHelpMsg('Archived POA&M Items cannot be updated');
            process.exit(0);
            break;
        }
        default: {
            printRedMsg('The "status" field must one of the following:');
            printRedMsg('    Risk Accepted, Ongoing, or Completed');
            printRedMsg(`Status provided was: ${dataObj.status}`);
            process.exit(1);
            break;
        }
    }
    // POC checks: If any poc information is provided all POC fields are required
    if ((Object.prototype.hasOwnProperty.call(dataObj, 'pocFirstName') || Object.prototype.hasOwnProperty.call(dataObj, 'pocLastName') ||
        Object.prototype.hasOwnProperty.call(dataObj, 'pocEmail') || Object.prototype.hasOwnProperty.call(dataObj, 'pocPhoneNumber')) && (!(Object.prototype.hasOwnProperty.call(dataObj, 'pocFirstName')) || !(Object.prototype.hasOwnProperty.call(dataObj, 'pocLastName')) ||
        !(Object.prototype.hasOwnProperty.call(dataObj, 'pocEmail')) || !(Object.prototype.hasOwnProperty.call(dataObj, 'pocPhoneNumber')))) {
        printRedMsg('If any POC content is provided (pocFirstName, pocLastName, pocEmail, pocPhoneNumber) than all POC parameters are required:');
        printRedMsg('    pocFirstName, pocLastName, pocEmail, pocPhoneNumber');
        printHelpMsg(HELP_MSG);
        process.exit(1);
    }
}
function generateBodyObj(dataObject) {
    let bodyObj = {};
    try {
        bodyObj = addRequiredFieldsToRequestBody(dataObject);
        processBusinessLogic(bodyObj, dataObject);
        addConditionalFields(bodyObj, dataObject);
        addOptionalFields(bodyObj, dataObject);
    }
    catch {
        process.exit(1);
    }
    return bodyObj;
}
class EmasserPutPoams extends core_1.Command {
    static usage = '<%= command.id %> [ARGUMENTS]';
    static description = 'Update a Plan of Action and Milestones (POA&M) into a systems.';
    static examples = ['<%= config.bin %> <%= command.id %> [-s,--systemId] [-f,--poamFile]',
        'The input file should be a well formed JSON containing the POA&M information based on defined business rules.',
        'Required JSON parameter/fields are: ',
        (0, json_colorizer_1.default)(JSON.stringify((0, utilities_1.getJsonExamples)('poams-put-required'), null, 2)),
        'Conditional JSON parameters/fields are: ',
        (0, json_colorizer_1.default)(JSON.stringify((0, utilities_1.getJsonExamples)('poams-put-conditional'), null, 2)),
        'Optional JSON parameters/fields are:',
        (0, json_colorizer_1.default)(JSON.stringify((0, utilities_1.getJsonExamples)('poams-post-put-optional'), null, 2))];
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Put (update) a Plan of Action and Milestones (POA&M) item(s) in a system. See emasser Features (emasserFeatures.md) for additional information.' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    async run() {
        const { flags } = await this.parse(EmasserPutPoams);
        const apiCxn = new apiConnection_1.ApiConnection();
        const updatePoam = new emass_client_1.POAMApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        const requestBodyArray = [];
        // Check if a POA&Ms json file was provided
        if (fs_1.default.existsSync(flags.poamFile)) {
            let data;
            try {
                data = JSON.parse(await (0, promises_1.readFile)(flags.poamFile, 'utf8'));
            }
            catch (error) {
                if (error.code === 'ENOENT') {
                    console.log('POA&Ms JSON file not found!');
                    process.exit(1);
                }
                else {
                    console.log('Error reading POA&Ms file, possible malformed json. Please use the -h flag for help.');
                    console.log('Error message was:', error.message);
                    process.exit(1);
                }
            }
            // POA&Ms json file provided, check if we have multiple POA&Ms to process
            if (Array.isArray(data)) {
                data.forEach((dataObject) => {
                    // Generate the put request object based on business logic
                    requestBodyArray.push(generateBodyObj(dataObject));
                });
            }
            else if (typeof data === 'object') {
                const dataObject = data;
                // Generate the put request object based on business logic
                requestBodyArray.push(generateBodyObj(dataObject));
            }
        }
        else {
            console.error('Invalid or POA&M JSON file not found on the provided directory:', flags.poamFile);
            process.exit(1);
        }
        updatePoam.updatePoamBySystemId(flags.systemId, requestBodyArray).then((response) => {
            console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response)));
        }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
    }
}
exports.default = EmasserPutPoams;
