// @mitre/saf/lib/commands/convert/aws_config2hdf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const fs_1 = tslib_1.__importDefault(require("fs"));
const hdf_converters_1 = require("@mitre/hdf-converters");
const global_1 = require("../../utils/global");
class AWSConfig2HDF extends core_1.Command {
    // Refs may not be defined if no resources were found
    ensureRefs(output) {
        return {
            ...output,
            profiles: output.profiles.map(profile => {
                return { ...profile,
                    controls: profile.controls.map(control => {
                        if (!control.refs || !control.results) {
                            return {
                                ...control,
                                refs: [],
                                results: [],
                            };
                        }
                        return control;
                    }),
                };
            }),
        };
    }
    async run() {
        const { flags } = await this.parse(AWSConfig2HDF);
        const converter = flags.accessKeyId && flags.secretAccessKey ? new hdf_converters_1.AwsConfigMapper({
            credentials: {
                accessKeyId: flags.accessKeyId || '',
                secretAccessKey: flags.secretAccessKey || '',
                sessionToken: flags.sessionToken,
            },
            region: flags.region,
        }, !flags.insecure) : new hdf_converters_1.AwsConfigMapper({ region: flags.region }, !flags.insecure);
        fs_1.default.writeFileSync((0, global_1.checkSuffix)(flags.output), JSON.stringify(this.ensureRefs(await converter.toHdf())));
    }
}
exports.default = AWSConfig2HDF;
AWSConfig2HDF.usage = 'convert aws_config2hdf -a, --accessKeyId=accessKeyId -r, --region=region -s, --secretAccessKey=secretAccessKey -t, --sessionToken=sessionToken -o, --output=OUTPUT';
AWSConfig2HDF.description = 'Pull Configuration findings from AWS Config and convert into a Heimdall Data Format JSON file';
AWSConfig2HDF.examples = ['saf convert aws_config2hdf -a ABCDEFGHIJKLMNOPQRSTUV -s +4NOT39A48REAL93SECRET934 -r us-east-1 -o output-hdf-name.json'];
AWSConfig2HDF.flags = {
    help: core_1.Flags.help({ char: 'h' }),
    accessKeyId: core_1.Flags.string({ char: 'a', required: false }),
    secretAccessKey: core_1.Flags.string({ char: 's', required: false }),
    sessionToken: core_1.Flags.string({ char: 't', required: false }),
    region: core_1.Flags.string({ char: 'r', required: true }),
    insecure: core_1.Flags.boolean({ char: 'i', required: false, default: false, description: 'Disable SSL verification, this is insecure.' }),
    output: core_1.Flags.string({ char: 'o', required: true }),
};
