// @mitre/saf/lib/commands/convert/fortify2hdf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const fs_1 = tslib_1.__importDefault(require("fs"));
const hdf_converters_1 = require("@mitre/hdf-converters");
const global_1 = require("../../utils/global");
const baseCommand_1 = require("../../utils/oclif/baseCommand");
class Fortify2HDF extends baseCommand_1.BaseCommand {
    static usage = '<%= command.id %> -i <fortify-fvdl> -o <hdf-scan-results-json> [-h] [-w]';
    static description = 'Translate a Fortify results FVDL file into a Heimdall Data Format JSON file; '
        + 'the FVDL file is an XML that can be extracted from the Fortify FPR project file '
        + 'using standard file compression tools';
    static examples = ['<%= config.bin %> <%= command.id %> -i audit.fvdl -o output-hdf-name.json'];
    static flags = {
        input: core_1.Flags.string({
            char: 'i',
            required: true,
            description: 'Input FVDL File',
        }),
        output: core_1.Flags.string({
            char: 'o',
            required: true,
            description: 'Output HDF JSON File',
        }),
        includeRaw: core_1.Flags.boolean({
            char: 'w',
            required: false,
            description: 'Include raw input file in HDF JSON file',
        }),
    };
    async run() {
        const { flags } = await this.parse(Fortify2HDF);
        // Check for correct input type
        const data = fs_1.default.readFileSync(flags.input, 'utf8');
        (0, global_1.checkInput)({ data, filename: flags.input }, 'fortify', 'Fortify results FVDL file');
        const converter = new hdf_converters_1.FortifyMapper(data, flags.includeRaw);
        fs_1.default.writeFileSync((0, global_1.checkSuffix)(flags.output), JSON.stringify(converter.toHdf(), null, 2));
    }
}
exports.default = Fortify2HDF;
