// @mitre/saf/lib/commands/convert/prowler2hdf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const fs_1 = tslib_1.__importDefault(require("fs"));
const hdf_converters_1 = require("@mitre/hdf-converters");
const global_1 = require("../../utils/global");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const path_1 = tslib_1.__importDefault(require("path"));
const baseCommand_1 = require("../../utils/oclif/baseCommand");
class Prowler2HDF extends baseCommand_1.BaseCommand {
    static usage = '<%= command.id %> -i <prowler-finding-json> -o <hdf-output-folder> [-h]';
    static description = 'Translate a Prowler-derived AWS Security Finding Format results from JSONL into a Heimdall Data Format JSON file';
    static examples = ['<%= config.bin %> <%= command.id %> -i prowler-asff.json -o output-folder'];
    static flags = {
        input: core_1.Flags.string({
            char: 'i',
            required: true,
            description: 'Input Prowler ASFF JSON File',
        }),
        output: core_1.Flags.string({
            char: 'o',
            required: true,
            description: 'Output HDF JSON Folder',
        }),
    };
    async run() {
        const { flags } = await this.parse(Prowler2HDF);
        const data = fs_1.default.readFileSync(flags.input, 'utf8');
        (0, global_1.checkInput)({ data: data, filename: flags.input }, // skipcq: JS-0240
        'asff', 'Prowler-derived AWS Security Finding Format results');
        const converter = new hdf_converters_1.ASFFResults(data);
        const results = converter.toHdf();
        // Create output folder if not exists
        if (!fs_1.default.existsSync(flags.output)) {
            fs_1.default.mkdirSync(flags.output);
        }
        lodash_1.default.forOwn(results, (result, filename) => {
            fs_1.default.writeFileSync(path_1.default.join(flags.output, (0, global_1.checkSuffix)((0, global_1.basename)(filename))), JSON.stringify(result, null, 2));
        });
    }
}
exports.default = Prowler2HDF;
