// @mitre/saf/lib/commands/convert/sarif2hdf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const fs_1 = tslib_1.__importDefault(require("fs"));
const hdf_converters_1 = require("@mitre/hdf-converters");
const global_1 = require("../../utils/global");
const baseCommand_1 = require("../../utils/oclif/baseCommand");
class Sarif2HDF extends baseCommand_1.BaseCommand {
    static usage = '<%= command.id %> -i <sarif-json> -o <hdf-scan-results-json> [-h] [-w]';
    static description = 'Translate a SARIF JSON file into a Heimdall Data Format JSON file\n'
        + 'SARIF levels to HDF impact mapping are:\n  '
        + 'SARIF level error -> HDF impact 0.7\n'
        + 'SARIF level warning -> HDF impact 0.5\n'
        + 'SARIF level note -> HDF impact 0.3\n'
        + 'SARIF level none -> HDF impact 0.1\n'
        + 'SARIF level not provided -> HDF impact 0.1 as default';
    static examples = ['<%= config.bin %> <%= command.id %> -i sarif-results.json -o output-hdf-name.json'];
    static flags = {
        input: core_1.Flags.string({
            char: 'i',
            required: true,
            description: 'Input SARIF JSON File',
        }),
        output: core_1.Flags.string({
            char: 'o',
            required: true,
            description: 'Output HDF JSON File',
        }),
        includeRaw: core_1.Flags.boolean({
            char: 'w',
            required: false,
            description: 'Include raw input file in HDF JSON file',
        }),
    };
    async run() {
        const { flags } = await this.parse(Sarif2HDF);
        // Check for correct input type
        const data = fs_1.default.readFileSync(flags.input, 'utf8');
        (0, global_1.checkInput)({ data, filename: flags.input }, 'sarif', 'SARIF JSON');
        const converter = new hdf_converters_1.SarifMapper(data, flags.includeRaw);
        fs_1.default.writeFileSync((0, global_1.checkSuffix)(flags.output), JSON.stringify(converter.toHdf(), null, 2));
    }
}
exports.default = Sarif2HDF;
