// @mitre/saf/lib/commands/convert/zap2hdf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const fs_1 = tslib_1.__importDefault(require("fs"));
const hdf_converters_1 = require("@mitre/hdf-converters");
const global_1 = require("../../utils/global");
class Zap2HDF extends core_1.Command {
    async run() {
        const { flags } = await this.parse(Zap2HDF);
        // Check for correct input type
        const data = fs_1.default.readFileSync(flags.input, 'utf8');
        (0, global_1.checkInput)({ data: data, filename: flags.input }, 'zap', 'OWASP ZAP results JSON');
        const converter = new hdf_converters_1.ZapMapper(fs_1.default.readFileSync(flags.input, 'utf8'), flags.name);
        fs_1.default.writeFileSync((0, global_1.checkSuffix)(flags.output), JSON.stringify(converter.toHdf()));
    }
}
Zap2HDF.usage = 'convert zap2hdf -i <zap-json> -n <target-site-name> -o <hdf-scan-results-json> [-h]';
Zap2HDF.description = 'Translate a OWASP ZAP results JSON to a Heimdall Data Format JSON file';
Zap2HDF.examples = ['saf convert zap2hdf -i zap_results.json -n mitre.org -o scan_results.json'];
Zap2HDF.flags = {
    help: core_1.Flags.help({ char: 'h' }),
    input: core_1.Flags.string({ char: 'i', required: true, description: 'Input OWASP Zap Results JSON File' }),
    name: core_1.Flags.string({ char: 'n', required: true, description: 'Target Site Name' }),
    output: core_1.Flags.string({ char: 'o', required: true, description: 'Output HDF JSON File' }),
};
exports.default = Zap2HDF;
