// @mitre/hdf-converters/lib/src/asff-mapper/case-trivy.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTrivy = void 0;
const html_entities_1 = require("html-entities");
const inspecjs_1 = require("inspecjs");
const lodash_1 = __importDefault(require("lodash"));
const global_1 = require("../utils/global");
function findingId(finding) {
    const generatorId = lodash_1.default.get(finding, 'GeneratorId');
    const cveId = lodash_1.default.get(finding, 'Resources[0].Details.Other.CVE ID');
    if (typeof cveId === 'string') {
        return (0, html_entities_1.encode)(`${generatorId}/${cveId}`);
    }
    else {
        const id = lodash_1.default.get(finding, 'Id');
        return (0, html_entities_1.encode)(`${generatorId}/${id}`);
    }
}
function findingNistTag(finding) {
    const cveId = lodash_1.default.get(finding, 'Resources[0].Details.Other.CVE ID');
    if (typeof cveId === 'string') {
        return global_1.DEFAULT_UPDATE_REMEDIATION_NIST_TAGS;
    }
    else {
        return [];
    }
}
function subfindingsStatus() {
    return inspecjs_1.ExecJSON.ControlResultStatus.Failed;
}
function subfindingsMessage(finding) {
    const cveId = lodash_1.default.get(finding, 'Resources[0].Details.Other.CVE ID');
    if (typeof cveId === 'string') {
        const patchedPackage = lodash_1.default.get(finding, 'Resources[0].Details.Other.Patched Package');
        const patchedVersionMessage = patchedPackage.length === 0
            ? 'There is no patched version of the package.'
            : `The package has been patched since version(s): ${patchedPackage}.`;
        return `For package ${lodash_1.default.get(finding, 'Resources[0].Details.Other.PkgName')}, the current version that is installed is ${lodash_1.default.get(finding, 'Resources[0].Details.Other.Installed Package')}.  ${patchedVersionMessage}`;
    }
    else {
        return undefined;
    }
}
function productName() {
    return 'Aqua Security - Trivy';
}
function doesNotHaveFindingTitlePrefix() {
    return true;
}
function filename() {
    return `${productName()}.json`;
}
function meta() {
    return { name: 'Trivy', title: 'Trivy Findings' };
}
function getTrivy() {
    return {
        findingId,
        findingNistTag,
        subfindingsStatus,
        subfindingsMessage,
        doesNotHaveFindingTitlePrefix,
        productName,
        filename,
        meta
    };
}
exports.getTrivy = getTrivy;
