// @mitre/hdf-converters/lib/src/asff-mapper/case-cms-inspec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCMSInSpec = void 0;
const html_entities_1 = require("html-entities");
const lodash_1 = __importDefault(require("lodash"));
function findingId(finding) {
    return (0, html_entities_1.encode)(lodash_1.default.get(finding, 'ProductFields.aws/securityhub/FindingId')
        .split('/')
        .slice(-1)[0]
        .split('-')
        .slice(0, -1)
        .join('-'));
}
function findingTitle(finding) {
    return (0, html_entities_1.encode)(lodash_1.default.get(finding, 'Description')
        .slice(`${lodash_1.default.get(finding, 'Title')} titled `.length)
        .split(' : ')[0]);
}
function findingDescription(finding) {
    return findingTitle(finding);
}
function subfindingsCodeDesc(finding) {
    return (0, html_entities_1.encode)(lodash_1.default.get(finding, 'Description')
        .slice(`${lodash_1.default.get(finding, 'Title')} titled `.length)
        .split(' : ')[1]);
}
function titlePrefix() {
    return '';
}
function productName(findings) {
    const finding = Array.isArray(findings) ? findings[0] : findings;
    return `cms_chef_inspec-${(0, html_entities_1.encode)(lodash_1.default.get(finding, 'ProductFields.aws/securityhub/FindingId')
        .split('/')
        .slice(-2)[0])}`;
}
function filename(findingInfo) {
    return `${productName(findingInfo[0])}.json`;
}
function getCMSInSpec() {
    return {
        findingId,
        findingTitle,
        findingDescription,
        subfindingsCodeDesc,
        titlePrefix,
        productName,
        filename
    };
}
exports.getCMSInSpec = getCMSInSpec;
