// @mitre/hdf-converters/lib/src/asff-mapper/case-prowler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProwler = void 0;
const html_entities_1 = require("html-entities");
const lodash_1 = __importDefault(require("lodash"));
const desc = () => ' ';
function subfindingsCodeDesc(finding) {
    return (0, html_entities_1.encode)(lodash_1.default.get(finding, 'Description'));
}
function findingId(finding) {
    const generatorId = lodash_1.default.get(finding, 'GeneratorId');
    const hyphenIndex = generatorId.indexOf('-');
    return (0, html_entities_1.encode)(generatorId.slice(hyphenIndex + 1));
}
function productName(findings) {
    const finding = Array.isArray(findings) ? findings[0] : findings;
    return (0, html_entities_1.encode)(lodash_1.default.get(finding, 'ProductFields.ProviderName'));
}
function filename(findingInfo) {
    return `${productName(findingInfo[1])}.json`;
}
function meta() {
    return { name: 'Prowler', title: 'Prowler Findings' };
}
function getProwler() {
    return {
        subfindingsCodeDesc,
        findingId,
        productName,
        desc,
        filename,
        meta
    };
}
exports.getProwler = getProwler;
