// @aws-sdk/client-config-service/dist-es/models/models_1.js
import { ConfigServiceServiceException as __BaseException } from "./ConfigServiceServiceException";
export class ResourceConcurrentModificationException extends __BaseException {
    constructor(opts) {
        super({
            name: "ResourceConcurrentModificationException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceConcurrentModificationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceConcurrentModificationException.prototype);
    }
}
export class TooManyTagsException extends __BaseException {
    constructor(opts) {
        super({
            name: "TooManyTagsException",
            $fault: "client",
            ...opts,
        });
        this.name = "TooManyTagsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TooManyTagsException.prototype);
    }
}
export const PutConformancePackRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutConformancePackResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutDeliveryChannelRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutEvaluationsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutEvaluationsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutExternalEvaluationRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutExternalEvaluationResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutOrganizationConfigRuleRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutOrganizationConfigRuleResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutOrganizationConformancePackRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutOrganizationConformancePackResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutRemediationConfigurationsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutRemediationConfigurationsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutRemediationExceptionsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutRemediationExceptionsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutResourceConfigRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutRetentionConfigurationRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutRetentionConfigurationResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutStoredQueryRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutStoredQueryResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const QueryInfoFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const SelectAggregateResourceConfigRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const SelectAggregateResourceConfigResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const SelectResourceConfigRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const SelectResourceConfigResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const StartConfigRulesEvaluationRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const StartConfigRulesEvaluationResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const StartConfigurationRecorderRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const StartRemediationExecutionRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const StartRemediationExecutionResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const StopConfigurationRecorderRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const TagResourceRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const UntagResourceRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
