// @aws-sdk/client-config-service/dist-es/models/models_0.js
import { __assign } from "tslib";
export var AccountAggregationSource;
(function (AccountAggregationSource) {
    AccountAggregationSource.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AccountAggregationSource || (AccountAggregationSource = {}));
export var ComplianceContributorCount;
(function (ComplianceContributorCount) {
    ComplianceContributorCount.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ComplianceContributorCount || (ComplianceContributorCount = {}));
export var ComplianceType;
(function (ComplianceType) {
    ComplianceType["Compliant"] = "COMPLIANT";
    ComplianceType["Insufficient_Data"] = "INSUFFICIENT_DATA";
    ComplianceType["Non_Compliant"] = "NON_COMPLIANT";
    ComplianceType["Not_Applicable"] = "NOT_APPLICABLE";
})(ComplianceType || (ComplianceType = {}));
export var Compliance;
(function (Compliance) {
    Compliance.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Compliance || (Compliance = {}));
export var AggregateComplianceByConfigRule;
(function (AggregateComplianceByConfigRule) {
    AggregateComplianceByConfigRule.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregateComplianceByConfigRule || (AggregateComplianceByConfigRule = {}));
export var ConformancePackComplianceType;
(function (ConformancePackComplianceType) {
    ConformancePackComplianceType["COMPLIANT"] = "COMPLIANT";
    ConformancePackComplianceType["INSUFFICIENT_DATA"] = "INSUFFICIENT_DATA";
    ConformancePackComplianceType["NON_COMPLIANT"] = "NON_COMPLIANT";
})(ConformancePackComplianceType || (ConformancePackComplianceType = {}));
export var AggregateConformancePackCompliance;
(function (AggregateConformancePackCompliance) {
    AggregateConformancePackCompliance.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregateConformancePackCompliance || (AggregateConformancePackCompliance = {}));
export var AggregateComplianceByConformancePack;
(function (AggregateComplianceByConformancePack) {
    AggregateComplianceByConformancePack.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregateComplianceByConformancePack || (AggregateComplianceByConformancePack = {}));
export var ComplianceSummary;
(function (ComplianceSummary) {
    ComplianceSummary.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ComplianceSummary || (ComplianceSummary = {}));
export var AggregateComplianceCount;
(function (AggregateComplianceCount) {
    AggregateComplianceCount.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregateComplianceCount || (AggregateComplianceCount = {}));
export var AggregateConformancePackComplianceCount;
(function (AggregateConformancePackComplianceCount) {
    AggregateConformancePackComplianceCount.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregateConformancePackComplianceCount || (AggregateConformancePackComplianceCount = {}));
export var AggregateConformancePackComplianceFilters;
(function (AggregateConformancePackComplianceFilters) {
    AggregateConformancePackComplianceFilters.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregateConformancePackComplianceFilters || (AggregateConformancePackComplianceFilters = {}));
export var AggregateConformancePackComplianceSummary;
(function (AggregateConformancePackComplianceSummary) {
    AggregateConformancePackComplianceSummary.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregateConformancePackComplianceSummary || (AggregateConformancePackComplianceSummary = {}));
export var AggregateConformancePackComplianceSummaryFilters;
(function (AggregateConformancePackComplianceSummaryFilters) {
    AggregateConformancePackComplianceSummaryFilters.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregateConformancePackComplianceSummaryFilters || (AggregateConformancePackComplianceSummaryFilters = {}));
export var AggregateConformancePackComplianceSummaryGroupKey;
(function (AggregateConformancePackComplianceSummaryGroupKey) {
    AggregateConformancePackComplianceSummaryGroupKey["ACCOUNT_ID"] = "ACCOUNT_ID";
    AggregateConformancePackComplianceSummaryGroupKey["AWS_REGION"] = "AWS_REGION";
})(AggregateConformancePackComplianceSummaryGroupKey || (AggregateConformancePackComplianceSummaryGroupKey = {}));
export var AggregatedSourceStatusType;
(function (AggregatedSourceStatusType) {
    AggregatedSourceStatusType["FAILED"] = "FAILED";
    AggregatedSourceStatusType["OUTDATED"] = "OUTDATED";
    AggregatedSourceStatusType["SUCCEEDED"] = "SUCCEEDED";
})(AggregatedSourceStatusType || (AggregatedSourceStatusType = {}));
export var AggregatedSourceType;
(function (AggregatedSourceType) {
    AggregatedSourceType["ACCOUNT"] = "ACCOUNT";
    AggregatedSourceType["ORGANIZATION"] = "ORGANIZATION";
})(AggregatedSourceType || (AggregatedSourceType = {}));
export var AggregatedSourceStatus;
(function (AggregatedSourceStatus) {
    AggregatedSourceStatus.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregatedSourceStatus || (AggregatedSourceStatus = {}));
export var EvaluationResultQualifier;
(function (EvaluationResultQualifier) {
    EvaluationResultQualifier.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(EvaluationResultQualifier || (EvaluationResultQualifier = {}));
export var EvaluationResultIdentifier;
(function (EvaluationResultIdentifier) {
    EvaluationResultIdentifier.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(EvaluationResultIdentifier || (EvaluationResultIdentifier = {}));
export var AggregateEvaluationResult;
(function (AggregateEvaluationResult) {
    AggregateEvaluationResult.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregateEvaluationResult || (AggregateEvaluationResult = {}));
export var ResourceType;
(function (ResourceType) {
    ResourceType["AccountPublicAccessBlock"] = "AWS::S3::AccountPublicAccessBlock";
    ResourceType["Alarm"] = "AWS::CloudWatch::Alarm";
    ResourceType["Api"] = "AWS::ApiGatewayV2::Api";
    ResourceType["Application"] = "AWS::ElasticBeanstalk::Application";
    ResourceType["ApplicationVersion"] = "AWS::ElasticBeanstalk::ApplicationVersion";
    ResourceType["AssociationCompliance"] = "AWS::SSM::AssociationCompliance";
    ResourceType["AutoScalingGroup"] = "AWS::AutoScaling::AutoScalingGroup";
    ResourceType["BackupPlan"] = "AWS::Backup::BackupPlan";
    ResourceType["BackupRecoveryPoint"] = "AWS::Backup::RecoveryPoint";
    ResourceType["BackupSelection"] = "AWS::Backup::BackupSelection";
    ResourceType["BackupVault"] = "AWS::Backup::BackupVault";
    ResourceType["Bucket"] = "AWS::S3::Bucket";
    ResourceType["Certificate"] = "AWS::ACM::Certificate";
    ResourceType["CloudFormationProduct"] = "AWS::ServiceCatalog::CloudFormationProduct";
    ResourceType["CloudFormationProvisionedProduct"] = "AWS::ServiceCatalog::CloudFormationProvisionedProduct";
    ResourceType["Cluster"] = "AWS::Redshift::Cluster";
    ResourceType["ClusterParameterGroup"] = "AWS::Redshift::ClusterParameterGroup";
    ResourceType["ClusterSecurityGroup"] = "AWS::Redshift::ClusterSecurityGroup";
    ResourceType["ClusterSnapshot"] = "AWS::Redshift::ClusterSnapshot";
    ResourceType["ClusterSubnetGroup"] = "AWS::Redshift::ClusterSubnetGroup";
    ResourceType["CodeDeployApplication"] = "AWS::CodeDeploy::Application";
    ResourceType["CodeDeployDeploymentConfig"] = "AWS::CodeDeploy::DeploymentConfig";
    ResourceType["CodeDeployDeploymentGroup"] = "AWS::CodeDeploy::DeploymentGroup";
    ResourceType["ConformancePackCompliance"] = "AWS::Config::ConformancePackCompliance";
    ResourceType["CustomerGateway"] = "AWS::EC2::CustomerGateway";
    ResourceType["DBCluster"] = "AWS::RDS::DBCluster";
    ResourceType["DBClusterSnapshot"] = "AWS::RDS::DBClusterSnapshot";
    ResourceType["DBInstance"] = "AWS::RDS::DBInstance";
    ResourceType["DBSecurityGroup"] = "AWS::RDS::DBSecurityGroup";
    ResourceType["DBSnapshot"] = "AWS::RDS::DBSnapshot";
    ResourceType["DBSubnetGroup"] = "AWS::RDS::DBSubnetGroup";
    ResourceType["Distribution"] = "AWS::CloudFront::Distribution";
    ResourceType["Domain"] = "AWS::Elasticsearch::Domain";
    ResourceType["ECRRepository"] = "AWS::ECR::Repository";
    ResourceType["ECSCluster"] = "AWS::ECS::Cluster";
    ResourceType["ECSService"] = "AWS::ECS::Service";
    ResourceType["ECSTaskDefinition"] = "AWS::ECS::TaskDefinition";
    ResourceType["EFSAccessPoint"] = "AWS::EFS::AccessPoint";
    ResourceType["EFSFileSystem"] = "AWS::EFS::FileSystem";
    ResourceType["EIP"] = "AWS::EC2::EIP";
    ResourceType["EKSCluster"] = "AWS::EKS::Cluster";
    ResourceType["EgressOnlyInternetGateway"] = "AWS::EC2::EgressOnlyInternetGateway";
    ResourceType["EncryptionConfig"] = "AWS::XRay::EncryptionConfig";
    ResourceType["Environment"] = "AWS::ElasticBeanstalk::Environment";
    ResourceType["EventSubscription"] = "AWS::RDS::EventSubscription";
    ResourceType["FileData"] = "AWS::SSM::FileData";
    ResourceType["FlowLog"] = "AWS::EC2::FlowLog";
    ResourceType["Function"] = "AWS::Lambda::Function";
    ResourceType["Group"] = "AWS::IAM::Group";
    ResourceType["Host"] = "AWS::EC2::Host";
    ResourceType["IPSetV2"] = "AWS::WAFv2::IPSet";
    ResourceType["Instance"] = "AWS::EC2::Instance";
    ResourceType["InternetGateway"] = "AWS::EC2::InternetGateway";
    ResourceType["Key"] = "AWS::KMS::Key";
    ResourceType["KinesisStream"] = "AWS::Kinesis::Stream";
    ResourceType["KinesisStreamConsumer"] = "AWS::Kinesis::StreamConsumer";
    ResourceType["LaunchConfiguration"] = "AWS::AutoScaling::LaunchConfiguration";
    ResourceType["LoadBalancer"] = "AWS::ElasticLoadBalancing::LoadBalancer";
    ResourceType["LoadBalancerV2"] = "AWS::ElasticLoadBalancingV2::LoadBalancer";
    ResourceType["ManagedInstanceInventory"] = "AWS::SSM::ManagedInstanceInventory";
    ResourceType["ManagedRuleSetV2"] = "AWS::WAFv2::ManagedRuleSet";
    ResourceType["NatGateway"] = "AWS::EC2::NatGateway";
    ResourceType["NetworkAcl"] = "AWS::EC2::NetworkAcl";
    ResourceType["NetworkFirewallFirewall"] = "AWS::NetworkFirewall::Firewall";
    ResourceType["NetworkFirewallFirewallPolicy"] = "AWS::NetworkFirewall::FirewallPolicy";
    ResourceType["NetworkFirewallRuleGroup"] = "AWS::NetworkFirewall::RuleGroup";
    ResourceType["NetworkInterface"] = "AWS::EC2::NetworkInterface";
    ResourceType["OpenSearchDomain"] = "AWS::OpenSearch::Domain";
    ResourceType["PatchCompliance"] = "AWS::SSM::PatchCompliance";
    ResourceType["Pipeline"] = "AWS::CodePipeline::Pipeline";
    ResourceType["Policy"] = "AWS::IAM::Policy";
    ResourceType["Portfolio"] = "AWS::ServiceCatalog::Portfolio";
    ResourceType["Project"] = "AWS::CodeBuild::Project";
    ResourceType["Protection"] = "AWS::Shield::Protection";
    ResourceType["QLDBLedger"] = "AWS::QLDB::Ledger";
    ResourceType["Queue"] = "AWS::SQS::Queue";
    ResourceType["RateBasedRule"] = "AWS::WAF::RateBasedRule";
    ResourceType["RedshiftEventSubscription"] = "AWS::Redshift::EventSubscription";
    ResourceType["RegexPatternSetV2"] = "AWS::WAFv2::RegexPatternSet";
    ResourceType["RegionalProtection"] = "AWS::ShieldRegional::Protection";
    ResourceType["RegionalRateBasedRule"] = "AWS::WAFRegional::RateBasedRule";
    ResourceType["RegionalRule"] = "AWS::WAFRegional::Rule";
    ResourceType["RegionalRuleGroup"] = "AWS::WAFRegional::RuleGroup";
    ResourceType["RegionalWebACL"] = "AWS::WAFRegional::WebACL";
    ResourceType["RegisteredHAInstance"] = "AWS::EC2::RegisteredHAInstance";
    ResourceType["ResourceCompliance"] = "AWS::Config::ResourceCompliance";
    ResourceType["RestApi"] = "AWS::ApiGateway::RestApi";
    ResourceType["Role"] = "AWS::IAM::Role";
    ResourceType["RouteTable"] = "AWS::EC2::RouteTable";
    ResourceType["Rule"] = "AWS::WAF::Rule";
    ResourceType["RuleGroup"] = "AWS::WAF::RuleGroup";
    ResourceType["RuleGroupV2"] = "AWS::WAFv2::RuleGroup";
    ResourceType["ScalingPolicy"] = "AWS::AutoScaling::ScalingPolicy";
    ResourceType["ScheduledAction"] = "AWS::AutoScaling::ScheduledAction";
    ResourceType["Secret"] = "AWS::SecretsManager::Secret";
    ResourceType["SecurityGroup"] = "AWS::EC2::SecurityGroup";
    ResourceType["Stack"] = "AWS::CloudFormation::Stack";
    ResourceType["Stage"] = "AWS::ApiGateway::Stage";
    ResourceType["StageV2"] = "AWS::ApiGatewayV2::Stage";
    ResourceType["StreamingDistribution"] = "AWS::CloudFront::StreamingDistribution";
    ResourceType["Subnet"] = "AWS::EC2::Subnet";
    ResourceType["Table"] = "AWS::DynamoDB::Table";
    ResourceType["Topic"] = "AWS::SNS::Topic";
    ResourceType["Trail"] = "AWS::CloudTrail::Trail";
    ResourceType["TransitGateway"] = "AWS::EC2::TransitGateway";
    ResourceType["User"] = "AWS::IAM::User";
    ResourceType["VPC"] = "AWS::EC2::VPC";
    ResourceType["VPCEndpoint"] = "AWS::EC2::VPCEndpoint";
    ResourceType["VPCEndpointService"] = "AWS::EC2::VPCEndpointService";
    ResourceType["VPCPeeringConnection"] = "AWS::EC2::VPCPeeringConnection";
    ResourceType["VPNConnection"] = "AWS::EC2::VPNConnection";
    ResourceType["VPNGateway"] = "AWS::EC2::VPNGateway";
    ResourceType["Volume"] = "AWS::EC2::Volume";
    ResourceType["WebACL"] = "AWS::WAF::WebACL";
    ResourceType["WebACLV2"] = "AWS::WAFv2::WebACL";
})(ResourceType || (ResourceType = {}));
export var AggregateResourceIdentifier;
(function (AggregateResourceIdentifier) {
    AggregateResourceIdentifier.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregateResourceIdentifier || (AggregateResourceIdentifier = {}));
export var AggregationAuthorization;
(function (AggregationAuthorization) {
    AggregationAuthorization.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AggregationAuthorization || (AggregationAuthorization = {}));
export var ConfigurationItemStatus;
(function (ConfigurationItemStatus) {
    ConfigurationItemStatus["OK"] = "OK";
    ConfigurationItemStatus["ResourceDeleted"] = "ResourceDeleted";
    ConfigurationItemStatus["ResourceDeletedNotRecorded"] = "ResourceDeletedNotRecorded";
    ConfigurationItemStatus["ResourceDiscovered"] = "ResourceDiscovered";
    ConfigurationItemStatus["ResourceNotRecorded"] = "ResourceNotRecorded";
})(ConfigurationItemStatus || (ConfigurationItemStatus = {}));
export var BaseConfigurationItem;
(function (BaseConfigurationItem) {
    BaseConfigurationItem.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(BaseConfigurationItem || (BaseConfigurationItem = {}));
export var BatchGetAggregateResourceConfigRequest;
(function (BatchGetAggregateResourceConfigRequest) {
    BatchGetAggregateResourceConfigRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(BatchGetAggregateResourceConfigRequest || (BatchGetAggregateResourceConfigRequest = {}));
export var BatchGetAggregateResourceConfigResponse;
(function (BatchGetAggregateResourceConfigResponse) {
    BatchGetAggregateResourceConfigResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(BatchGetAggregateResourceConfigResponse || (BatchGetAggregateResourceConfigResponse = {}));
export var ResourceKey;
(function (ResourceKey) {
    ResourceKey.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ResourceKey || (ResourceKey = {}));
export var BatchGetResourceConfigRequest;
(function (BatchGetResourceConfigRequest) {
    BatchGetResourceConfigRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(BatchGetResourceConfigRequest || (BatchGetResourceConfigRequest = {}));
export var BatchGetResourceConfigResponse;
(function (BatchGetResourceConfigResponse) {
    BatchGetResourceConfigResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(BatchGetResourceConfigResponse || (BatchGetResourceConfigResponse = {}));
export var ChronologicalOrder;
(function (ChronologicalOrder) {
    ChronologicalOrder["Forward"] = "Forward";
    ChronologicalOrder["Reverse"] = "Reverse";
})(ChronologicalOrder || (ChronologicalOrder = {}));
export var ComplianceByConfigRule;
(function (ComplianceByConfigRule) {
    ComplianceByConfigRule.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ComplianceByConfigRule || (ComplianceByConfigRule = {}));
export var ComplianceByResource;
(function (ComplianceByResource) {
    ComplianceByResource.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ComplianceByResource || (ComplianceByResource = {}));
export var ComplianceSummaryByResourceType;
(function (ComplianceSummaryByResourceType) {
    ComplianceSummaryByResourceType.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ComplianceSummaryByResourceType || (ComplianceSummaryByResourceType = {}));
export var DeliveryStatus;
(function (DeliveryStatus) {
    DeliveryStatus["Failure"] = "Failure";
    DeliveryStatus["Not_Applicable"] = "Not_Applicable";
    DeliveryStatus["Success"] = "Success";
})(DeliveryStatus || (DeliveryStatus = {}));
export var ConfigExportDeliveryInfo;
(function (ConfigExportDeliveryInfo) {
    ConfigExportDeliveryInfo.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigExportDeliveryInfo || (ConfigExportDeliveryInfo = {}));
export var ConfigRuleState;
(function (ConfigRuleState) {
    ConfigRuleState["ACTIVE"] = "ACTIVE";
    ConfigRuleState["DELETING"] = "DELETING";
    ConfigRuleState["DELETING_RESULTS"] = "DELETING_RESULTS";
    ConfigRuleState["EVALUATING"] = "EVALUATING";
})(ConfigRuleState || (ConfigRuleState = {}));
export var MaximumExecutionFrequency;
(function (MaximumExecutionFrequency) {
    MaximumExecutionFrequency["One_Hour"] = "One_Hour";
    MaximumExecutionFrequency["Six_Hours"] = "Six_Hours";
    MaximumExecutionFrequency["Three_Hours"] = "Three_Hours";
    MaximumExecutionFrequency["Twelve_Hours"] = "Twelve_Hours";
    MaximumExecutionFrequency["TwentyFour_Hours"] = "TwentyFour_Hours";
})(MaximumExecutionFrequency || (MaximumExecutionFrequency = {}));
export var Scope;
(function (Scope) {
    Scope.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Scope || (Scope = {}));
export var Owner;
(function (Owner) {
    Owner["Aws"] = "AWS";
    Owner["Custom_Lambda"] = "CUSTOM_LAMBDA";
})(Owner || (Owner = {}));
export var EventSource;
(function (EventSource) {
    EventSource["Aws_Config"] = "aws.config";
})(EventSource || (EventSource = {}));
export var MessageType;
(function (MessageType) {
    MessageType["ConfigurationItemChangeNotification"] = "ConfigurationItemChangeNotification";
    MessageType["ConfigurationSnapshotDeliveryCompleted"] = "ConfigurationSnapshotDeliveryCompleted";
    MessageType["OversizedConfigurationItemChangeNotification"] = "OversizedConfigurationItemChangeNotification";
    MessageType["ScheduledNotification"] = "ScheduledNotification";
})(MessageType || (MessageType = {}));
export var SourceDetail;
(function (SourceDetail) {
    SourceDetail.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SourceDetail || (SourceDetail = {}));
export var Source;
(function (Source) {
    Source.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Source || (Source = {}));
export var ConfigRule;
(function (ConfigRule) {
    ConfigRule.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigRule || (ConfigRule = {}));
export var ConfigRuleComplianceFilters;
(function (ConfigRuleComplianceFilters) {
    ConfigRuleComplianceFilters.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigRuleComplianceFilters || (ConfigRuleComplianceFilters = {}));
export var ConfigRuleComplianceSummaryFilters;
(function (ConfigRuleComplianceSummaryFilters) {
    ConfigRuleComplianceSummaryFilters.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigRuleComplianceSummaryFilters || (ConfigRuleComplianceSummaryFilters = {}));
export var ConfigRuleComplianceSummaryGroupKey;
(function (ConfigRuleComplianceSummaryGroupKey) {
    ConfigRuleComplianceSummaryGroupKey["ACCOUNT_ID"] = "ACCOUNT_ID";
    ConfigRuleComplianceSummaryGroupKey["AWS_REGION"] = "AWS_REGION";
})(ConfigRuleComplianceSummaryGroupKey || (ConfigRuleComplianceSummaryGroupKey = {}));
export var ConfigRuleEvaluationStatus;
(function (ConfigRuleEvaluationStatus) {
    ConfigRuleEvaluationStatus.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigRuleEvaluationStatus || (ConfigRuleEvaluationStatus = {}));
export var ConfigSnapshotDeliveryProperties;
(function (ConfigSnapshotDeliveryProperties) {
    ConfigSnapshotDeliveryProperties.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigSnapshotDeliveryProperties || (ConfigSnapshotDeliveryProperties = {}));
export var ConfigStreamDeliveryInfo;
(function (ConfigStreamDeliveryInfo) {
    ConfigStreamDeliveryInfo.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigStreamDeliveryInfo || (ConfigStreamDeliveryInfo = {}));
export var OrganizationAggregationSource;
(function (OrganizationAggregationSource) {
    OrganizationAggregationSource.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OrganizationAggregationSource || (OrganizationAggregationSource = {}));
export var ConfigurationAggregator;
(function (ConfigurationAggregator) {
    ConfigurationAggregator.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigurationAggregator || (ConfigurationAggregator = {}));
export var Relationship;
(function (Relationship) {
    Relationship.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Relationship || (Relationship = {}));
export var ConfigurationItem;
(function (ConfigurationItem) {
    ConfigurationItem.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigurationItem || (ConfigurationItem = {}));
export var RecordingGroup;
(function (RecordingGroup) {
    RecordingGroup.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RecordingGroup || (RecordingGroup = {}));
export var ConfigurationRecorder;
(function (ConfigurationRecorder) {
    ConfigurationRecorder.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigurationRecorder || (ConfigurationRecorder = {}));
export var RecorderStatus;
(function (RecorderStatus) {
    RecorderStatus["Failure"] = "Failure";
    RecorderStatus["Pending"] = "Pending";
    RecorderStatus["Success"] = "Success";
})(RecorderStatus || (RecorderStatus = {}));
export var ConfigurationRecorderStatus;
(function (ConfigurationRecorderStatus) {
    ConfigurationRecorderStatus.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigurationRecorderStatus || (ConfigurationRecorderStatus = {}));
export var ConformancePackComplianceFilters;
(function (ConformancePackComplianceFilters) {
    ConformancePackComplianceFilters.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConformancePackComplianceFilters || (ConformancePackComplianceFilters = {}));
export var ConformancePackComplianceSummary;
(function (ConformancePackComplianceSummary) {
    ConformancePackComplianceSummary.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConformancePackComplianceSummary || (ConformancePackComplianceSummary = {}));
export var ConformancePackInputParameter;
(function (ConformancePackInputParameter) {
    ConformancePackInputParameter.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConformancePackInputParameter || (ConformancePackInputParameter = {}));
export var ConformancePackDetail;
(function (ConformancePackDetail) {
    ConformancePackDetail.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConformancePackDetail || (ConformancePackDetail = {}));
export var ConformancePackEvaluationFilters;
(function (ConformancePackEvaluationFilters) {
    ConformancePackEvaluationFilters.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConformancePackEvaluationFilters || (ConformancePackEvaluationFilters = {}));
export var ConformancePackEvaluationResult;
(function (ConformancePackEvaluationResult) {
    ConformancePackEvaluationResult.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConformancePackEvaluationResult || (ConformancePackEvaluationResult = {}));
export var ConformancePackRuleCompliance;
(function (ConformancePackRuleCompliance) {
    ConformancePackRuleCompliance.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConformancePackRuleCompliance || (ConformancePackRuleCompliance = {}));
export var ConformancePackState;
(function (ConformancePackState) {
    ConformancePackState["CREATE_COMPLETE"] = "CREATE_COMPLETE";
    ConformancePackState["CREATE_FAILED"] = "CREATE_FAILED";
    ConformancePackState["CREATE_IN_PROGRESS"] = "CREATE_IN_PROGRESS";
    ConformancePackState["DELETE_FAILED"] = "DELETE_FAILED";
    ConformancePackState["DELETE_IN_PROGRESS"] = "DELETE_IN_PROGRESS";
})(ConformancePackState || (ConformancePackState = {}));
export var ConformancePackStatusDetail;
(function (ConformancePackStatusDetail) {
    ConformancePackStatusDetail.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConformancePackStatusDetail || (ConformancePackStatusDetail = {}));
export var DeleteAggregationAuthorizationRequest;
(function (DeleteAggregationAuthorizationRequest) {
    DeleteAggregationAuthorizationRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteAggregationAuthorizationRequest || (DeleteAggregationAuthorizationRequest = {}));
export var DeleteConfigRuleRequest;
(function (DeleteConfigRuleRequest) {
    DeleteConfigRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConfigRuleRequest || (DeleteConfigRuleRequest = {}));
export var DeleteConfigurationAggregatorRequest;
(function (DeleteConfigurationAggregatorRequest) {
    DeleteConfigurationAggregatorRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConfigurationAggregatorRequest || (DeleteConfigurationAggregatorRequest = {}));
export var DeleteConfigurationRecorderRequest;
(function (DeleteConfigurationRecorderRequest) {
    DeleteConfigurationRecorderRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConfigurationRecorderRequest || (DeleteConfigurationRecorderRequest = {}));
export var DeleteConformancePackRequest;
(function (DeleteConformancePackRequest) {
    DeleteConformancePackRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConformancePackRequest || (DeleteConformancePackRequest = {}));
export var DeleteDeliveryChannelRequest;
(function (DeleteDeliveryChannelRequest) {
    DeleteDeliveryChannelRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteDeliveryChannelRequest || (DeleteDeliveryChannelRequest = {}));
export var DeleteEvaluationResultsRequest;
(function (DeleteEvaluationResultsRequest) {
    DeleteEvaluationResultsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteEvaluationResultsRequest || (DeleteEvaluationResultsRequest = {}));
export var DeleteEvaluationResultsResponse;
(function (DeleteEvaluationResultsResponse) {
    DeleteEvaluationResultsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteEvaluationResultsResponse || (DeleteEvaluationResultsResponse = {}));
export var DeleteOrganizationConfigRuleRequest;
(function (DeleteOrganizationConfigRuleRequest) {
    DeleteOrganizationConfigRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteOrganizationConfigRuleRequest || (DeleteOrganizationConfigRuleRequest = {}));
export var DeleteOrganizationConformancePackRequest;
(function (DeleteOrganizationConformancePackRequest) {
    DeleteOrganizationConformancePackRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteOrganizationConformancePackRequest || (DeleteOrganizationConformancePackRequest = {}));
export var DeletePendingAggregationRequestRequest;
(function (DeletePendingAggregationRequestRequest) {
    DeletePendingAggregationRequestRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeletePendingAggregationRequestRequest || (DeletePendingAggregationRequestRequest = {}));
export var DeleteRemediationConfigurationRequest;
(function (DeleteRemediationConfigurationRequest) {
    DeleteRemediationConfigurationRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteRemediationConfigurationRequest || (DeleteRemediationConfigurationRequest = {}));
export var DeleteRemediationConfigurationResponse;
(function (DeleteRemediationConfigurationResponse) {
    DeleteRemediationConfigurationResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteRemediationConfigurationResponse || (DeleteRemediationConfigurationResponse = {}));
export var RemediationExceptionResourceKey;
(function (RemediationExceptionResourceKey) {
    RemediationExceptionResourceKey.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RemediationExceptionResourceKey || (RemediationExceptionResourceKey = {}));
export var DeleteRemediationExceptionsRequest;
(function (DeleteRemediationExceptionsRequest) {
    DeleteRemediationExceptionsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteRemediationExceptionsRequest || (DeleteRemediationExceptionsRequest = {}));
export var FailedDeleteRemediationExceptionsBatch;
(function (FailedDeleteRemediationExceptionsBatch) {
    FailedDeleteRemediationExceptionsBatch.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(FailedDeleteRemediationExceptionsBatch || (FailedDeleteRemediationExceptionsBatch = {}));
export var DeleteRemediationExceptionsResponse;
(function (DeleteRemediationExceptionsResponse) {
    DeleteRemediationExceptionsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteRemediationExceptionsResponse || (DeleteRemediationExceptionsResponse = {}));
export var DeleteResourceConfigRequest;
(function (DeleteResourceConfigRequest) {
    DeleteResourceConfigRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteResourceConfigRequest || (DeleteResourceConfigRequest = {}));
export var DeleteRetentionConfigurationRequest;
(function (DeleteRetentionConfigurationRequest) {
    DeleteRetentionConfigurationRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteRetentionConfigurationRequest || (DeleteRetentionConfigurationRequest = {}));
export var DeleteStoredQueryRequest;
(function (DeleteStoredQueryRequest) {
    DeleteStoredQueryRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteStoredQueryRequest || (DeleteStoredQueryRequest = {}));
export var DeleteStoredQueryResponse;
(function (DeleteStoredQueryResponse) {
    DeleteStoredQueryResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteStoredQueryResponse || (DeleteStoredQueryResponse = {}));
export var DeliverConfigSnapshotRequest;
(function (DeliverConfigSnapshotRequest) {
    DeliverConfigSnapshotRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeliverConfigSnapshotRequest || (DeliverConfigSnapshotRequest = {}));
export var DeliverConfigSnapshotResponse;
(function (DeliverConfigSnapshotResponse) {
    DeliverConfigSnapshotResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeliverConfigSnapshotResponse || (DeliverConfigSnapshotResponse = {}));
export var DeliveryChannel;
(function (DeliveryChannel) {
    DeliveryChannel.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeliveryChannel || (DeliveryChannel = {}));
export var DeliveryChannelStatus;
(function (DeliveryChannelStatus) {
    DeliveryChannelStatus.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeliveryChannelStatus || (DeliveryChannelStatus = {}));
export var DescribeAggregateComplianceByConfigRulesRequest;
(function (DescribeAggregateComplianceByConfigRulesRequest) {
    DescribeAggregateComplianceByConfigRulesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeAggregateComplianceByConfigRulesRequest || (DescribeAggregateComplianceByConfigRulesRequest = {}));
export var DescribeAggregateComplianceByConfigRulesResponse;
(function (DescribeAggregateComplianceByConfigRulesResponse) {
    DescribeAggregateComplianceByConfigRulesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeAggregateComplianceByConfigRulesResponse || (DescribeAggregateComplianceByConfigRulesResponse = {}));
export var DescribeAggregateComplianceByConformancePacksRequest;
(function (DescribeAggregateComplianceByConformancePacksRequest) {
    DescribeAggregateComplianceByConformancePacksRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeAggregateComplianceByConformancePacksRequest || (DescribeAggregateComplianceByConformancePacksRequest = {}));
export var DescribeAggregateComplianceByConformancePacksResponse;
(function (DescribeAggregateComplianceByConformancePacksResponse) {
    DescribeAggregateComplianceByConformancePacksResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeAggregateComplianceByConformancePacksResponse || (DescribeAggregateComplianceByConformancePacksResponse = {}));
export var DescribeAggregationAuthorizationsRequest;
(function (DescribeAggregationAuthorizationsRequest) {
    DescribeAggregationAuthorizationsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeAggregationAuthorizationsRequest || (DescribeAggregationAuthorizationsRequest = {}));
export var DescribeAggregationAuthorizationsResponse;
(function (DescribeAggregationAuthorizationsResponse) {
    DescribeAggregationAuthorizationsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeAggregationAuthorizationsResponse || (DescribeAggregationAuthorizationsResponse = {}));
export var DescribeComplianceByConfigRuleRequest;
(function (DescribeComplianceByConfigRuleRequest) {
    DescribeComplianceByConfigRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeComplianceByConfigRuleRequest || (DescribeComplianceByConfigRuleRequest = {}));
export var DescribeComplianceByConfigRuleResponse;
(function (DescribeComplianceByConfigRuleResponse) {
    DescribeComplianceByConfigRuleResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeComplianceByConfigRuleResponse || (DescribeComplianceByConfigRuleResponse = {}));
export var DescribeComplianceByResourceRequest;
(function (DescribeComplianceByResourceRequest) {
    DescribeComplianceByResourceRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeComplianceByResourceRequest || (DescribeComplianceByResourceRequest = {}));
export var DescribeComplianceByResourceResponse;
(function (DescribeComplianceByResourceResponse) {
    DescribeComplianceByResourceResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeComplianceByResourceResponse || (DescribeComplianceByResourceResponse = {}));
export var DescribeConfigRuleEvaluationStatusRequest;
(function (DescribeConfigRuleEvaluationStatusRequest) {
    DescribeConfigRuleEvaluationStatusRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigRuleEvaluationStatusRequest || (DescribeConfigRuleEvaluationStatusRequest = {}));
export var DescribeConfigRuleEvaluationStatusResponse;
(function (DescribeConfigRuleEvaluationStatusResponse) {
    DescribeConfigRuleEvaluationStatusResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigRuleEvaluationStatusResponse || (DescribeConfigRuleEvaluationStatusResponse = {}));
export var DescribeConfigRulesRequest;
(function (DescribeConfigRulesRequest) {
    DescribeConfigRulesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigRulesRequest || (DescribeConfigRulesRequest = {}));
export var DescribeConfigRulesResponse;
(function (DescribeConfigRulesResponse) {
    DescribeConfigRulesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigRulesResponse || (DescribeConfigRulesResponse = {}));
export var DescribeConfigurationAggregatorsRequest;
(function (DescribeConfigurationAggregatorsRequest) {
    DescribeConfigurationAggregatorsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigurationAggregatorsRequest || (DescribeConfigurationAggregatorsRequest = {}));
export var DescribeConfigurationAggregatorsResponse;
(function (DescribeConfigurationAggregatorsResponse) {
    DescribeConfigurationAggregatorsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigurationAggregatorsResponse || (DescribeConfigurationAggregatorsResponse = {}));
export var DescribeConfigurationAggregatorSourcesStatusRequest;
(function (DescribeConfigurationAggregatorSourcesStatusRequest) {
    DescribeConfigurationAggregatorSourcesStatusRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigurationAggregatorSourcesStatusRequest || (DescribeConfigurationAggregatorSourcesStatusRequest = {}));
export var DescribeConfigurationAggregatorSourcesStatusResponse;
(function (DescribeConfigurationAggregatorSourcesStatusResponse) {
    DescribeConfigurationAggregatorSourcesStatusResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigurationAggregatorSourcesStatusResponse || (DescribeConfigurationAggregatorSourcesStatusResponse = {}));
export var DescribeConfigurationRecordersRequest;
(function (DescribeConfigurationRecordersRequest) {
    DescribeConfigurationRecordersRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigurationRecordersRequest || (DescribeConfigurationRecordersRequest = {}));
export var DescribeConfigurationRecordersResponse;
(function (DescribeConfigurationRecordersResponse) {
    DescribeConfigurationRecordersResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigurationRecordersResponse || (DescribeConfigurationRecordersResponse = {}));
export var DescribeConfigurationRecorderStatusRequest;
(function (DescribeConfigurationRecorderStatusRequest) {
    DescribeConfigurationRecorderStatusRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigurationRecorderStatusRequest || (DescribeConfigurationRecorderStatusRequest = {}));
export var DescribeConfigurationRecorderStatusResponse;
(function (DescribeConfigurationRecorderStatusResponse) {
    DescribeConfigurationRecorderStatusResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigurationRecorderStatusResponse || (DescribeConfigurationRecorderStatusResponse = {}));
export var DescribeConformancePackComplianceRequest;
(function (DescribeConformancePackComplianceRequest) {
    DescribeConformancePackComplianceRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConformancePackComplianceRequest || (DescribeConformancePackComplianceRequest = {}));
export var DescribeConformancePackComplianceResponse;
(function (DescribeConformancePackComplianceResponse) {
    DescribeConformancePackComplianceResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConformancePackComplianceResponse || (DescribeConformancePackComplianceResponse = {}));
export var DescribeConformancePacksRequest;
(function (DescribeConformancePacksRequest) {
    DescribeConformancePacksRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConformancePacksRequest || (DescribeConformancePacksRequest = {}));
export var DescribeConformancePacksResponse;
(function (DescribeConformancePacksResponse) {
    DescribeConformancePacksResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConformancePacksResponse || (DescribeConformancePacksResponse = {}));
export var DescribeConformancePackStatusRequest;
(function (DescribeConformancePackStatusRequest) {
    DescribeConformancePackStatusRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConformancePackStatusRequest || (DescribeConformancePackStatusRequest = {}));
export var DescribeConformancePackStatusResponse;
(function (DescribeConformancePackStatusResponse) {
    DescribeConformancePackStatusResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConformancePackStatusResponse || (DescribeConformancePackStatusResponse = {}));
export var DescribeDeliveryChannelsRequest;
(function (DescribeDeliveryChannelsRequest) {
    DescribeDeliveryChannelsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeDeliveryChannelsRequest || (DescribeDeliveryChannelsRequest = {}));
export var DescribeDeliveryChannelsResponse;
(function (DescribeDeliveryChannelsResponse) {
    DescribeDeliveryChannelsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeDeliveryChannelsResponse || (DescribeDeliveryChannelsResponse = {}));
export var DescribeDeliveryChannelStatusRequest;
(function (DescribeDeliveryChannelStatusRequest) {
    DescribeDeliveryChannelStatusRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeDeliveryChannelStatusRequest || (DescribeDeliveryChannelStatusRequest = {}));
export var DescribeDeliveryChannelStatusResponse;
(function (DescribeDeliveryChannelStatusResponse) {
    DescribeDeliveryChannelStatusResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeDeliveryChannelStatusResponse || (DescribeDeliveryChannelStatusResponse = {}));
export var DescribeOrganizationConfigRulesRequest;
(function (DescribeOrganizationConfigRulesRequest) {
    DescribeOrganizationConfigRulesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeOrganizationConfigRulesRequest || (DescribeOrganizationConfigRulesRequest = {}));
export var OrganizationConfigRuleTriggerType;
(function (OrganizationConfigRuleTriggerType) {
    OrganizationConfigRuleTriggerType["CONFIGURATION_ITEM_CHANGE_NOTIFICATION"] = "ConfigurationItemChangeNotification";
    OrganizationConfigRuleTriggerType["OVERSIZED_CONFIGURATION_ITEM_CHANGE_NOTIFCATION"] = "OversizedConfigurationItemChangeNotification";
    OrganizationConfigRuleTriggerType["SCHEDULED_NOTIFICATION"] = "ScheduledNotification";
})(OrganizationConfigRuleTriggerType || (OrganizationConfigRuleTriggerType = {}));
export var OrganizationCustomRuleMetadata;
(function (OrganizationCustomRuleMetadata) {
    OrganizationCustomRuleMetadata.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OrganizationCustomRuleMetadata || (OrganizationCustomRuleMetadata = {}));
export var OrganizationManagedRuleMetadata;
(function (OrganizationManagedRuleMetadata) {
    OrganizationManagedRuleMetadata.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OrganizationManagedRuleMetadata || (OrganizationManagedRuleMetadata = {}));
export var OrganizationConfigRule;
(function (OrganizationConfigRule) {
    OrganizationConfigRule.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OrganizationConfigRule || (OrganizationConfigRule = {}));
export var DescribeOrganizationConfigRulesResponse;
(function (DescribeOrganizationConfigRulesResponse) {
    DescribeOrganizationConfigRulesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeOrganizationConfigRulesResponse || (DescribeOrganizationConfigRulesResponse = {}));
export var DescribeOrganizationConfigRuleStatusesRequest;
(function (DescribeOrganizationConfigRuleStatusesRequest) {
    DescribeOrganizationConfigRuleStatusesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeOrganizationConfigRuleStatusesRequest || (DescribeOrganizationConfigRuleStatusesRequest = {}));
export var OrganizationRuleStatus;
(function (OrganizationRuleStatus) {
    OrganizationRuleStatus["CREATE_FAILED"] = "CREATE_FAILED";
    OrganizationRuleStatus["CREATE_IN_PROGRESS"] = "CREATE_IN_PROGRESS";
    OrganizationRuleStatus["CREATE_SUCCESSFUL"] = "CREATE_SUCCESSFUL";
    OrganizationRuleStatus["DELETE_FAILED"] = "DELETE_FAILED";
    OrganizationRuleStatus["DELETE_IN_PROGRESS"] = "DELETE_IN_PROGRESS";
    OrganizationRuleStatus["DELETE_SUCCESSFUL"] = "DELETE_SUCCESSFUL";
    OrganizationRuleStatus["UPDATE_FAILED"] = "UPDATE_FAILED";
    OrganizationRuleStatus["UPDATE_IN_PROGRESS"] = "UPDATE_IN_PROGRESS";
    OrganizationRuleStatus["UPDATE_SUCCESSFUL"] = "UPDATE_SUCCESSFUL";
})(OrganizationRuleStatus || (OrganizationRuleStatus = {}));
export var OrganizationConfigRuleStatus;
(function (OrganizationConfigRuleStatus) {
    OrganizationConfigRuleStatus.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OrganizationConfigRuleStatus || (OrganizationConfigRuleStatus = {}));
export var DescribeOrganizationConfigRuleStatusesResponse;
(function (DescribeOrganizationConfigRuleStatusesResponse) {
    DescribeOrganizationConfigRuleStatusesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeOrganizationConfigRuleStatusesResponse || (DescribeOrganizationConfigRuleStatusesResponse = {}));
export var DescribeOrganizationConformancePacksRequest;
(function (DescribeOrganizationConformancePacksRequest) {
    DescribeOrganizationConformancePacksRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeOrganizationConformancePacksRequest || (DescribeOrganizationConformancePacksRequest = {}));
export var OrganizationConformancePack;
(function (OrganizationConformancePack) {
    OrganizationConformancePack.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OrganizationConformancePack || (OrganizationConformancePack = {}));
export var DescribeOrganizationConformancePacksResponse;
(function (DescribeOrganizationConformancePacksResponse) {
    DescribeOrganizationConformancePacksResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeOrganizationConformancePacksResponse || (DescribeOrganizationConformancePacksResponse = {}));
export var DescribeOrganizationConformancePackStatusesRequest;
(function (DescribeOrganizationConformancePackStatusesRequest) {
    DescribeOrganizationConformancePackStatusesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeOrganizationConformancePackStatusesRequest || (DescribeOrganizationConformancePackStatusesRequest = {}));
export var OrganizationResourceStatus;
(function (OrganizationResourceStatus) {
    OrganizationResourceStatus["CREATE_FAILED"] = "CREATE_FAILED";
    OrganizationResourceStatus["CREATE_IN_PROGRESS"] = "CREATE_IN_PROGRESS";
    OrganizationResourceStatus["CREATE_SUCCESSFUL"] = "CREATE_SUCCESSFUL";
    OrganizationResourceStatus["DELETE_FAILED"] = "DELETE_FAILED";
    OrganizationResourceStatus["DELETE_IN_PROGRESS"] = "DELETE_IN_PROGRESS";
    OrganizationResourceStatus["DELETE_SUCCESSFUL"] = "DELETE_SUCCESSFUL";
    OrganizationResourceStatus["UPDATE_FAILED"] = "UPDATE_FAILED";
    OrganizationResourceStatus["UPDATE_IN_PROGRESS"] = "UPDATE_IN_PROGRESS";
    OrganizationResourceStatus["UPDATE_SUCCESSFUL"] = "UPDATE_SUCCESSFUL";
})(OrganizationResourceStatus || (OrganizationResourceStatus = {}));
export var OrganizationConformancePackStatus;
(function (OrganizationConformancePackStatus) {
    OrganizationConformancePackStatus.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OrganizationConformancePackStatus || (OrganizationConformancePackStatus = {}));
export var DescribeOrganizationConformancePackStatusesResponse;
(function (DescribeOrganizationConformancePackStatusesResponse) {
    DescribeOrganizationConformancePackStatusesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeOrganizationConformancePackStatusesResponse || (DescribeOrganizationConformancePackStatusesResponse = {}));
export var DescribePendingAggregationRequestsRequest;
(function (DescribePendingAggregationRequestsRequest) {
    DescribePendingAggregationRequestsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribePendingAggregationRequestsRequest || (DescribePendingAggregationRequestsRequest = {}));
export var PendingAggregationRequest;
(function (PendingAggregationRequest) {
    PendingAggregationRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PendingAggregationRequest || (PendingAggregationRequest = {}));
export var DescribePendingAggregationRequestsResponse;
(function (DescribePendingAggregationRequestsResponse) {
    DescribePendingAggregationRequestsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribePendingAggregationRequestsResponse || (DescribePendingAggregationRequestsResponse = {}));
export var DescribeRemediationConfigurationsRequest;
(function (DescribeRemediationConfigurationsRequest) {
    DescribeRemediationConfigurationsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeRemediationConfigurationsRequest || (DescribeRemediationConfigurationsRequest = {}));
export var SsmControls;
(function (SsmControls) {
    SsmControls.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SsmControls || (SsmControls = {}));
export var ExecutionControls;
(function (ExecutionControls) {
    ExecutionControls.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ExecutionControls || (ExecutionControls = {}));
export var ResourceValueType;
(function (ResourceValueType) {
    ResourceValueType["RESOURCE_ID"] = "RESOURCE_ID";
})(ResourceValueType || (ResourceValueType = {}));
export var ResourceValue;
(function (ResourceValue) {
    ResourceValue.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ResourceValue || (ResourceValue = {}));
export var StaticValue;
(function (StaticValue) {
    StaticValue.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(StaticValue || (StaticValue = {}));
export var RemediationParameterValue;
(function (RemediationParameterValue) {
    RemediationParameterValue.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RemediationParameterValue || (RemediationParameterValue = {}));
export var RemediationTargetType;
(function (RemediationTargetType) {
    RemediationTargetType["SSM_DOCUMENT"] = "SSM_DOCUMENT";
})(RemediationTargetType || (RemediationTargetType = {}));
export var RemediationConfiguration;
(function (RemediationConfiguration) {
    RemediationConfiguration.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RemediationConfiguration || (RemediationConfiguration = {}));
export var DescribeRemediationConfigurationsResponse;
(function (DescribeRemediationConfigurationsResponse) {
    DescribeRemediationConfigurationsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeRemediationConfigurationsResponse || (DescribeRemediationConfigurationsResponse = {}));
export var DescribeRemediationExceptionsRequest;
(function (DescribeRemediationExceptionsRequest) {
    DescribeRemediationExceptionsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeRemediationExceptionsRequest || (DescribeRemediationExceptionsRequest = {}));
export var RemediationException;
(function (RemediationException) {
    RemediationException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RemediationException || (RemediationException = {}));
export var DescribeRemediationExceptionsResponse;
(function (DescribeRemediationExceptionsResponse) {
    DescribeRemediationExceptionsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeRemediationExceptionsResponse || (DescribeRemediationExceptionsResponse = {}));
export var DescribeRemediationExecutionStatusRequest;
(function (DescribeRemediationExecutionStatusRequest) {
    DescribeRemediationExecutionStatusRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeRemediationExecutionStatusRequest || (DescribeRemediationExecutionStatusRequest = {}));
export var RemediationExecutionState;
(function (RemediationExecutionState) {
    RemediationExecutionState["FAILED"] = "FAILED";
    RemediationExecutionState["IN_PROGRESS"] = "IN_PROGRESS";
    RemediationExecutionState["QUEUED"] = "QUEUED";
    RemediationExecutionState["SUCCEEDED"] = "SUCCEEDED";
})(RemediationExecutionState || (RemediationExecutionState = {}));
export var RemediationExecutionStepState;
(function (RemediationExecutionStepState) {
    RemediationExecutionStepState["FAILED"] = "FAILED";
    RemediationExecutionStepState["PENDING"] = "PENDING";
    RemediationExecutionStepState["SUCCEEDED"] = "SUCCEEDED";
})(RemediationExecutionStepState || (RemediationExecutionStepState = {}));
export var RemediationExecutionStep;
(function (RemediationExecutionStep) {
    RemediationExecutionStep.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RemediationExecutionStep || (RemediationExecutionStep = {}));
export var RemediationExecutionStatus;
(function (RemediationExecutionStatus) {
    RemediationExecutionStatus.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RemediationExecutionStatus || (RemediationExecutionStatus = {}));
export var DescribeRemediationExecutionStatusResponse;
(function (DescribeRemediationExecutionStatusResponse) {
    DescribeRemediationExecutionStatusResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeRemediationExecutionStatusResponse || (DescribeRemediationExecutionStatusResponse = {}));
export var DescribeRetentionConfigurationsRequest;
(function (DescribeRetentionConfigurationsRequest) {
    DescribeRetentionConfigurationsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeRetentionConfigurationsRequest || (DescribeRetentionConfigurationsRequest = {}));
export var RetentionConfiguration;
(function (RetentionConfiguration) {
    RetentionConfiguration.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RetentionConfiguration || (RetentionConfiguration = {}));
export var DescribeRetentionConfigurationsResponse;
(function (DescribeRetentionConfigurationsResponse) {
    DescribeRetentionConfigurationsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeRetentionConfigurationsResponse || (DescribeRetentionConfigurationsResponse = {}));
export var Evaluation;
(function (Evaluation) {
    Evaluation.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Evaluation || (Evaluation = {}));
export var EvaluationResult;
(function (EvaluationResult) {
    EvaluationResult.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(EvaluationResult || (EvaluationResult = {}));
export var ExternalEvaluation;
(function (ExternalEvaluation) {
    ExternalEvaluation.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ExternalEvaluation || (ExternalEvaluation = {}));
export var FailedRemediationBatch;
(function (FailedRemediationBatch) {
    FailedRemediationBatch.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(FailedRemediationBatch || (FailedRemediationBatch = {}));
export var FailedRemediationExceptionBatch;
(function (FailedRemediationExceptionBatch) {
    FailedRemediationExceptionBatch.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(FailedRemediationExceptionBatch || (FailedRemediationExceptionBatch = {}));
export var FieldInfo;
(function (FieldInfo) {
    FieldInfo.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(FieldInfo || (FieldInfo = {}));
export var GetAggregateComplianceDetailsByConfigRuleRequest;
(function (GetAggregateComplianceDetailsByConfigRuleRequest) {
    GetAggregateComplianceDetailsByConfigRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAggregateComplianceDetailsByConfigRuleRequest || (GetAggregateComplianceDetailsByConfigRuleRequest = {}));
export var GetAggregateComplianceDetailsByConfigRuleResponse;
(function (GetAggregateComplianceDetailsByConfigRuleResponse) {
    GetAggregateComplianceDetailsByConfigRuleResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAggregateComplianceDetailsByConfigRuleResponse || (GetAggregateComplianceDetailsByConfigRuleResponse = {}));
export var GetAggregateConfigRuleComplianceSummaryRequest;
(function (GetAggregateConfigRuleComplianceSummaryRequest) {
    GetAggregateConfigRuleComplianceSummaryRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAggregateConfigRuleComplianceSummaryRequest || (GetAggregateConfigRuleComplianceSummaryRequest = {}));
export var GetAggregateConfigRuleComplianceSummaryResponse;
(function (GetAggregateConfigRuleComplianceSummaryResponse) {
    GetAggregateConfigRuleComplianceSummaryResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAggregateConfigRuleComplianceSummaryResponse || (GetAggregateConfigRuleComplianceSummaryResponse = {}));
export var GetAggregateConformancePackComplianceSummaryRequest;
(function (GetAggregateConformancePackComplianceSummaryRequest) {
    GetAggregateConformancePackComplianceSummaryRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAggregateConformancePackComplianceSummaryRequest || (GetAggregateConformancePackComplianceSummaryRequest = {}));
export var GetAggregateConformancePackComplianceSummaryResponse;
(function (GetAggregateConformancePackComplianceSummaryResponse) {
    GetAggregateConformancePackComplianceSummaryResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAggregateConformancePackComplianceSummaryResponse || (GetAggregateConformancePackComplianceSummaryResponse = {}));
export var ResourceCountFilters;
(function (ResourceCountFilters) {
    ResourceCountFilters.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ResourceCountFilters || (ResourceCountFilters = {}));
export var ResourceCountGroupKey;
(function (ResourceCountGroupKey) {
    ResourceCountGroupKey["ACCOUNT_ID"] = "ACCOUNT_ID";
    ResourceCountGroupKey["AWS_REGION"] = "AWS_REGION";
    ResourceCountGroupKey["RESOURCE_TYPE"] = "RESOURCE_TYPE";
})(ResourceCountGroupKey || (ResourceCountGroupKey = {}));
export var GetAggregateDiscoveredResourceCountsRequest;
(function (GetAggregateDiscoveredResourceCountsRequest) {
    GetAggregateDiscoveredResourceCountsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAggregateDiscoveredResourceCountsRequest || (GetAggregateDiscoveredResourceCountsRequest = {}));
export var GroupedResourceCount;
(function (GroupedResourceCount) {
    GroupedResourceCount.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GroupedResourceCount || (GroupedResourceCount = {}));
export var GetAggregateDiscoveredResourceCountsResponse;
(function (GetAggregateDiscoveredResourceCountsResponse) {
    GetAggregateDiscoveredResourceCountsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAggregateDiscoveredResourceCountsResponse || (GetAggregateDiscoveredResourceCountsResponse = {}));
export var GetAggregateResourceConfigRequest;
(function (GetAggregateResourceConfigRequest) {
    GetAggregateResourceConfigRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAggregateResourceConfigRequest || (GetAggregateResourceConfigRequest = {}));
export var GetAggregateResourceConfigResponse;
(function (GetAggregateResourceConfigResponse) {
    GetAggregateResourceConfigResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAggregateResourceConfigResponse || (GetAggregateResourceConfigResponse = {}));
export var GetComplianceDetailsByConfigRuleRequest;
(function (GetComplianceDetailsByConfigRuleRequest) {
    GetComplianceDetailsByConfigRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetComplianceDetailsByConfigRuleRequest || (GetComplianceDetailsByConfigRuleRequest = {}));
export var GetComplianceDetailsByConfigRuleResponse;
(function (GetComplianceDetailsByConfigRuleResponse) {
    GetComplianceDetailsByConfigRuleResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetComplianceDetailsByConfigRuleResponse || (GetComplianceDetailsByConfigRuleResponse = {}));
export var GetComplianceDetailsByResourceRequest;
(function (GetComplianceDetailsByResourceRequest) {
    GetComplianceDetailsByResourceRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetComplianceDetailsByResourceRequest || (GetComplianceDetailsByResourceRequest = {}));
export var GetComplianceDetailsByResourceResponse;
(function (GetComplianceDetailsByResourceResponse) {
    GetComplianceDetailsByResourceResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetComplianceDetailsByResourceResponse || (GetComplianceDetailsByResourceResponse = {}));
export var GetComplianceSummaryByConfigRuleResponse;
(function (GetComplianceSummaryByConfigRuleResponse) {
    GetComplianceSummaryByConfigRuleResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetComplianceSummaryByConfigRuleResponse || (GetComplianceSummaryByConfigRuleResponse = {}));
export var GetComplianceSummaryByResourceTypeRequest;
(function (GetComplianceSummaryByResourceTypeRequest) {
    GetComplianceSummaryByResourceTypeRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetComplianceSummaryByResourceTypeRequest || (GetComplianceSummaryByResourceTypeRequest = {}));
export var GetComplianceSummaryByResourceTypeResponse;
(function (GetComplianceSummaryByResourceTypeResponse) {
    GetComplianceSummaryByResourceTypeResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetComplianceSummaryByResourceTypeResponse || (GetComplianceSummaryByResourceTypeResponse = {}));
export var GetConformancePackComplianceDetailsRequest;
(function (GetConformancePackComplianceDetailsRequest) {
    GetConformancePackComplianceDetailsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetConformancePackComplianceDetailsRequest || (GetConformancePackComplianceDetailsRequest = {}));
export var GetConformancePackComplianceDetailsResponse;
(function (GetConformancePackComplianceDetailsResponse) {
    GetConformancePackComplianceDetailsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetConformancePackComplianceDetailsResponse || (GetConformancePackComplianceDetailsResponse = {}));
export var GetConformancePackComplianceSummaryRequest;
(function (GetConformancePackComplianceSummaryRequest) {
    GetConformancePackComplianceSummaryRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetConformancePackComplianceSummaryRequest || (GetConformancePackComplianceSummaryRequest = {}));
export var GetConformancePackComplianceSummaryResponse;
(function (GetConformancePackComplianceSummaryResponse) {
    GetConformancePackComplianceSummaryResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetConformancePackComplianceSummaryResponse || (GetConformancePackComplianceSummaryResponse = {}));
export var GetDiscoveredResourceCountsRequest;
(function (GetDiscoveredResourceCountsRequest) {
    GetDiscoveredResourceCountsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetDiscoveredResourceCountsRequest || (GetDiscoveredResourceCountsRequest = {}));
export var ResourceCount;
(function (ResourceCount) {
    ResourceCount.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ResourceCount || (ResourceCount = {}));
export var GetDiscoveredResourceCountsResponse;
(function (GetDiscoveredResourceCountsResponse) {
    GetDiscoveredResourceCountsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetDiscoveredResourceCountsResponse || (GetDiscoveredResourceCountsResponse = {}));
export var MemberAccountRuleStatus;
(function (MemberAccountRuleStatus) {
    MemberAccountRuleStatus["CREATE_FAILED"] = "CREATE_FAILED";
    MemberAccountRuleStatus["CREATE_IN_PROGRESS"] = "CREATE_IN_PROGRESS";
    MemberAccountRuleStatus["CREATE_SUCCESSFUL"] = "CREATE_SUCCESSFUL";
    MemberAccountRuleStatus["DELETE_FAILED"] = "DELETE_FAILED";
    MemberAccountRuleStatus["DELETE_IN_PROGRESS"] = "DELETE_IN_PROGRESS";
    MemberAccountRuleStatus["DELETE_SUCCESSFUL"] = "DELETE_SUCCESSFUL";
    MemberAccountRuleStatus["UPDATE_FAILED"] = "UPDATE_FAILED";
    MemberAccountRuleStatus["UPDATE_IN_PROGRESS"] = "UPDATE_IN_PROGRESS";
    MemberAccountRuleStatus["UPDATE_SUCCESSFUL"] = "UPDATE_SUCCESSFUL";
})(MemberAccountRuleStatus || (MemberAccountRuleStatus = {}));
export var StatusDetailFilters;
(function (StatusDetailFilters) {
    StatusDetailFilters.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(StatusDetailFilters || (StatusDetailFilters = {}));
export var GetOrganizationConfigRuleDetailedStatusRequest;
(function (GetOrganizationConfigRuleDetailedStatusRequest) {
    GetOrganizationConfigRuleDetailedStatusRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetOrganizationConfigRuleDetailedStatusRequest || (GetOrganizationConfigRuleDetailedStatusRequest = {}));
export var MemberAccountStatus;
(function (MemberAccountStatus) {
    MemberAccountStatus.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(MemberAccountStatus || (MemberAccountStatus = {}));
export var GetOrganizationConfigRuleDetailedStatusResponse;
(function (GetOrganizationConfigRuleDetailedStatusResponse) {
    GetOrganizationConfigRuleDetailedStatusResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetOrganizationConfigRuleDetailedStatusResponse || (GetOrganizationConfigRuleDetailedStatusResponse = {}));
export var OrganizationResourceDetailedStatus;
(function (OrganizationResourceDetailedStatus) {
    OrganizationResourceDetailedStatus["CREATE_FAILED"] = "CREATE_FAILED";
    OrganizationResourceDetailedStatus["CREATE_IN_PROGRESS"] = "CREATE_IN_PROGRESS";
    OrganizationResourceDetailedStatus["CREATE_SUCCESSFUL"] = "CREATE_SUCCESSFUL";
    OrganizationResourceDetailedStatus["DELETE_FAILED"] = "DELETE_FAILED";
    OrganizationResourceDetailedStatus["DELETE_IN_PROGRESS"] = "DELETE_IN_PROGRESS";
    OrganizationResourceDetailedStatus["DELETE_SUCCESSFUL"] = "DELETE_SUCCESSFUL";
    OrganizationResourceDetailedStatus["UPDATE_FAILED"] = "UPDATE_FAILED";
    OrganizationResourceDetailedStatus["UPDATE_IN_PROGRESS"] = "UPDATE_IN_PROGRESS";
    OrganizationResourceDetailedStatus["UPDATE_SUCCESSFUL"] = "UPDATE_SUCCESSFUL";
})(OrganizationResourceDetailedStatus || (OrganizationResourceDetailedStatus = {}));
export var OrganizationResourceDetailedStatusFilters;
(function (OrganizationResourceDetailedStatusFilters) {
    OrganizationResourceDetailedStatusFilters.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OrganizationResourceDetailedStatusFilters || (OrganizationResourceDetailedStatusFilters = {}));
export var GetOrganizationConformancePackDetailedStatusRequest;
(function (GetOrganizationConformancePackDetailedStatusRequest) {
    GetOrganizationConformancePackDetailedStatusRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetOrganizationConformancePackDetailedStatusRequest || (GetOrganizationConformancePackDetailedStatusRequest = {}));
export var OrganizationConformancePackDetailedStatus;
(function (OrganizationConformancePackDetailedStatus) {
    OrganizationConformancePackDetailedStatus.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OrganizationConformancePackDetailedStatus || (OrganizationConformancePackDetailedStatus = {}));
export var GetOrganizationConformancePackDetailedStatusResponse;
(function (GetOrganizationConformancePackDetailedStatusResponse) {
    GetOrganizationConformancePackDetailedStatusResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetOrganizationConformancePackDetailedStatusResponse || (GetOrganizationConformancePackDetailedStatusResponse = {}));
export var GetResourceConfigHistoryRequest;
(function (GetResourceConfigHistoryRequest) {
    GetResourceConfigHistoryRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetResourceConfigHistoryRequest || (GetResourceConfigHistoryRequest = {}));
export var GetResourceConfigHistoryResponse;
(function (GetResourceConfigHistoryResponse) {
    GetResourceConfigHistoryResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetResourceConfigHistoryResponse || (GetResourceConfigHistoryResponse = {}));
export var GetStoredQueryRequest;
(function (GetStoredQueryRequest) {
    GetStoredQueryRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetStoredQueryRequest || (GetStoredQueryRequest = {}));
export var StoredQuery;
(function (StoredQuery) {
    StoredQuery.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(StoredQuery || (StoredQuery = {}));
export var GetStoredQueryResponse;
(function (GetStoredQueryResponse) {
    GetStoredQueryResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetStoredQueryResponse || (GetStoredQueryResponse = {}));
export var ResourceFilters;
(function (ResourceFilters) {
    ResourceFilters.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ResourceFilters || (ResourceFilters = {}));
export var ListAggregateDiscoveredResourcesRequest;
(function (ListAggregateDiscoveredResourcesRequest) {
    ListAggregateDiscoveredResourcesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListAggregateDiscoveredResourcesRequest || (ListAggregateDiscoveredResourcesRequest = {}));
export var ListAggregateDiscoveredResourcesResponse;
(function (ListAggregateDiscoveredResourcesResponse) {
    ListAggregateDiscoveredResourcesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListAggregateDiscoveredResourcesResponse || (ListAggregateDiscoveredResourcesResponse = {}));
export var ListDiscoveredResourcesRequest;
(function (ListDiscoveredResourcesRequest) {
    ListDiscoveredResourcesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListDiscoveredResourcesRequest || (ListDiscoveredResourcesRequest = {}));
export var ResourceIdentifier;
(function (ResourceIdentifier) {
    ResourceIdentifier.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ResourceIdentifier || (ResourceIdentifier = {}));
export var ListDiscoveredResourcesResponse;
(function (ListDiscoveredResourcesResponse) {
    ListDiscoveredResourcesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListDiscoveredResourcesResponse || (ListDiscoveredResourcesResponse = {}));
export var ListStoredQueriesRequest;
(function (ListStoredQueriesRequest) {
    ListStoredQueriesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListStoredQueriesRequest || (ListStoredQueriesRequest = {}));
export var StoredQueryMetadata;
(function (StoredQueryMetadata) {
    StoredQueryMetadata.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(StoredQueryMetadata || (StoredQueryMetadata = {}));
export var ListStoredQueriesResponse;
(function (ListStoredQueriesResponse) {
    ListStoredQueriesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListStoredQueriesResponse || (ListStoredQueriesResponse = {}));
export var ListTagsForResourceRequest;
(function (ListTagsForResourceRequest) {
    ListTagsForResourceRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListTagsForResourceRequest || (ListTagsForResourceRequest = {}));
export var Tag;
(function (Tag) {
    Tag.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Tag || (Tag = {}));
export var ListTagsForResourceResponse;
(function (ListTagsForResourceResponse) {
    ListTagsForResourceResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListTagsForResourceResponse || (ListTagsForResourceResponse = {}));
export var PutAggregationAuthorizationRequest;
(function (PutAggregationAuthorizationRequest) {
    PutAggregationAuthorizationRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutAggregationAuthorizationRequest || (PutAggregationAuthorizationRequest = {}));
export var PutAggregationAuthorizationResponse;
(function (PutAggregationAuthorizationResponse) {
    PutAggregationAuthorizationResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutAggregationAuthorizationResponse || (PutAggregationAuthorizationResponse = {}));
export var PutConfigRuleRequest;
(function (PutConfigRuleRequest) {
    PutConfigRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutConfigRuleRequest || (PutConfigRuleRequest = {}));
export var PutConfigurationAggregatorRequest;
(function (PutConfigurationAggregatorRequest) {
    PutConfigurationAggregatorRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutConfigurationAggregatorRequest || (PutConfigurationAggregatorRequest = {}));
export var PutConfigurationAggregatorResponse;
(function (PutConfigurationAggregatorResponse) {
    PutConfigurationAggregatorResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutConfigurationAggregatorResponse || (PutConfigurationAggregatorResponse = {}));
export var PutConfigurationRecorderRequest;
(function (PutConfigurationRecorderRequest) {
    PutConfigurationRecorderRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutConfigurationRecorderRequest || (PutConfigurationRecorderRequest = {}));
export var PutConformancePackRequest;
(function (PutConformancePackRequest) {
    PutConformancePackRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutConformancePackRequest || (PutConformancePackRequest = {}));
export var PutConformancePackResponse;
(function (PutConformancePackResponse) {
    PutConformancePackResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutConformancePackResponse || (PutConformancePackResponse = {}));
export var PutDeliveryChannelRequest;
(function (PutDeliveryChannelRequest) {
    PutDeliveryChannelRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutDeliveryChannelRequest || (PutDeliveryChannelRequest = {}));
export var PutEvaluationsRequest;
(function (PutEvaluationsRequest) {
    PutEvaluationsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutEvaluationsRequest || (PutEvaluationsRequest = {}));
export var PutEvaluationsResponse;
(function (PutEvaluationsResponse) {
    PutEvaluationsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutEvaluationsResponse || (PutEvaluationsResponse = {}));
export var PutExternalEvaluationRequest;
(function (PutExternalEvaluationRequest) {
    PutExternalEvaluationRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutExternalEvaluationRequest || (PutExternalEvaluationRequest = {}));
export var PutExternalEvaluationResponse;
(function (PutExternalEvaluationResponse) {
    PutExternalEvaluationResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutExternalEvaluationResponse || (PutExternalEvaluationResponse = {}));
export var PutOrganizationConfigRuleRequest;
(function (PutOrganizationConfigRuleRequest) {
    PutOrganizationConfigRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutOrganizationConfigRuleRequest || (PutOrganizationConfigRuleRequest = {}));
export var PutOrganizationConfigRuleResponse;
(function (PutOrganizationConfigRuleResponse) {
    PutOrganizationConfigRuleResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutOrganizationConfigRuleResponse || (PutOrganizationConfigRuleResponse = {}));
export var PutOrganizationConformancePackRequest;
(function (PutOrganizationConformancePackRequest) {
    PutOrganizationConformancePackRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutOrganizationConformancePackRequest || (PutOrganizationConformancePackRequest = {}));
export var PutOrganizationConformancePackResponse;
(function (PutOrganizationConformancePackResponse) {
    PutOrganizationConformancePackResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutOrganizationConformancePackResponse || (PutOrganizationConformancePackResponse = {}));
export var PutRemediationConfigurationsRequest;
(function (PutRemediationConfigurationsRequest) {
    PutRemediationConfigurationsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutRemediationConfigurationsRequest || (PutRemediationConfigurationsRequest = {}));
export var PutRemediationConfigurationsResponse;
(function (PutRemediationConfigurationsResponse) {
    PutRemediationConfigurationsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutRemediationConfigurationsResponse || (PutRemediationConfigurationsResponse = {}));
export var PutRemediationExceptionsRequest;
(function (PutRemediationExceptionsRequest) {
    PutRemediationExceptionsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutRemediationExceptionsRequest || (PutRemediationExceptionsRequest = {}));
export var PutRemediationExceptionsResponse;
(function (PutRemediationExceptionsResponse) {
    PutRemediationExceptionsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutRemediationExceptionsResponse || (PutRemediationExceptionsResponse = {}));
