// @aws-sdk/client-config-service/dist-es/models/models_0.js
import { ConfigServiceServiceException as __BaseException } from "./ConfigServiceServiceException";
export var ComplianceType;
(function (ComplianceType) {
    ComplianceType["Compliant"] = "COMPLIANT";
    ComplianceType["Insufficient_Data"] = "INSUFFICIENT_DATA";
    ComplianceType["Non_Compliant"] = "NON_COMPLIANT";
    ComplianceType["Not_Applicable"] = "NOT_APPLICABLE";
})(ComplianceType || (ComplianceType = {}));
export var ConformancePackComplianceType;
(function (ConformancePackComplianceType) {
    ConformancePackComplianceType["COMPLIANT"] = "COMPLIANT";
    ConformancePackComplianceType["INSUFFICIENT_DATA"] = "INSUFFICIENT_DATA";
    ConformancePackComplianceType["NON_COMPLIANT"] = "NON_COMPLIANT";
})(ConformancePackComplianceType || (ConformancePackComplianceType = {}));
export var AggregateConformancePackComplianceSummaryGroupKey;
(function (AggregateConformancePackComplianceSummaryGroupKey) {
    AggregateConformancePackComplianceSummaryGroupKey["ACCOUNT_ID"] = "ACCOUNT_ID";
    AggregateConformancePackComplianceSummaryGroupKey["AWS_REGION"] = "AWS_REGION";
})(AggregateConformancePackComplianceSummaryGroupKey || (AggregateConformancePackComplianceSummaryGroupKey = {}));
export var AggregatedSourceStatusType;
(function (AggregatedSourceStatusType) {
    AggregatedSourceStatusType["FAILED"] = "FAILED";
    AggregatedSourceStatusType["OUTDATED"] = "OUTDATED";
    AggregatedSourceStatusType["SUCCEEDED"] = "SUCCEEDED";
})(AggregatedSourceStatusType || (AggregatedSourceStatusType = {}));
export var AggregatedSourceType;
(function (AggregatedSourceType) {
    AggregatedSourceType["ACCOUNT"] = "ACCOUNT";
    AggregatedSourceType["ORGANIZATION"] = "ORGANIZATION";
})(AggregatedSourceType || (AggregatedSourceType = {}));
export var ResourceType;
(function (ResourceType) {
    ResourceType["AccessAnalyzerAnalyzer"] = "AWS::AccessAnalyzer::Analyzer";
    ResourceType["AccountPublicAccessBlock"] = "AWS::S3::AccountPublicAccessBlock";
    ResourceType["Alarm"] = "AWS::CloudWatch::Alarm";
    ResourceType["Api"] = "AWS::ApiGatewayV2::Api";
    ResourceType["AppConfigApplication"] = "AWS::AppConfig::Application";
    ResourceType["AppSyncGraphQLApi"] = "AWS::AppSync::GraphQLApi";
    ResourceType["Application"] = "AWS::ElasticBeanstalk::Application";
    ResourceType["ApplicationVersion"] = "AWS::ElasticBeanstalk::ApplicationVersion";
    ResourceType["AssociationCompliance"] = "AWS::SSM::AssociationCompliance";
    ResourceType["AthenaDataCatalog"] = "AWS::Athena::DataCatalog";
    ResourceType["AthenaWorkGroup"] = "AWS::Athena::WorkGroup";
    ResourceType["AutoScalingGroup"] = "AWS::AutoScaling::AutoScalingGroup";
    ResourceType["BackupPlan"] = "AWS::Backup::BackupPlan";
    ResourceType["BackupRecoveryPoint"] = "AWS::Backup::RecoveryPoint";
    ResourceType["BackupSelection"] = "AWS::Backup::BackupSelection";
    ResourceType["BackupVault"] = "AWS::Backup::BackupVault";
    ResourceType["BatchComputeEnvironment"] = "AWS::Batch::ComputeEnvironment";
    ResourceType["BatchJobQueue"] = "AWS::Batch::JobQueue";
    ResourceType["Bucket"] = "AWS::S3::Bucket";
    ResourceType["Certificate"] = "AWS::ACM::Certificate";
    ResourceType["CloudFormationProduct"] = "AWS::ServiceCatalog::CloudFormationProduct";
    ResourceType["CloudFormationProvisionedProduct"] = "AWS::ServiceCatalog::CloudFormationProvisionedProduct";
    ResourceType["Cluster"] = "AWS::Redshift::Cluster";
    ResourceType["ClusterParameterGroup"] = "AWS::Redshift::ClusterParameterGroup";
    ResourceType["ClusterSecurityGroup"] = "AWS::Redshift::ClusterSecurityGroup";
    ResourceType["ClusterSnapshot"] = "AWS::Redshift::ClusterSnapshot";
    ResourceType["ClusterSubnetGroup"] = "AWS::Redshift::ClusterSubnetGroup";
    ResourceType["CodeDeployApplication"] = "AWS::CodeDeploy::Application";
    ResourceType["CodeDeployDeploymentConfig"] = "AWS::CodeDeploy::DeploymentConfig";
    ResourceType["CodeDeployDeploymentGroup"] = "AWS::CodeDeploy::DeploymentGroup";
    ResourceType["ConformancePackCompliance"] = "AWS::Config::ConformancePackCompliance";
    ResourceType["CustomerGateway"] = "AWS::EC2::CustomerGateway";
    ResourceType["DBCluster"] = "AWS::RDS::DBCluster";
    ResourceType["DBClusterSnapshot"] = "AWS::RDS::DBClusterSnapshot";
    ResourceType["DBInstance"] = "AWS::RDS::DBInstance";
    ResourceType["DBSecurityGroup"] = "AWS::RDS::DBSecurityGroup";
    ResourceType["DBSnapshot"] = "AWS::RDS::DBSnapshot";
    ResourceType["DBSubnetGroup"] = "AWS::RDS::DBSubnetGroup";
    ResourceType["DMSCertificate"] = "AWS::DMS::Certificate";
    ResourceType["DMSEventSubscription"] = "AWS::DMS::EventSubscription";
    ResourceType["DMSReplicationSubnetGroup"] = "AWS::DMS::ReplicationSubnetGroup";
    ResourceType["DataSyncLocationEFS"] = "AWS::DataSync::LocationEFS";
    ResourceType["DataSyncLocationFSxLustre"] = "AWS::DataSync::LocationFSxLustre";
    ResourceType["DataSyncLocationNFS"] = "AWS::DataSync::LocationNFS";
    ResourceType["DataSyncLocationS3"] = "AWS::DataSync::LocationS3";
    ResourceType["DataSyncLocationSMB"] = "AWS::DataSync::LocationSMB";
    ResourceType["DataSyncTask"] = "AWS::DataSync::Task";
    ResourceType["DetectiveGraph"] = "AWS::Detective::Graph";
    ResourceType["Distribution"] = "AWS::CloudFront::Distribution";
    ResourceType["Domain"] = "AWS::Elasticsearch::Domain";
    ResourceType["ECRPublicRepository"] = "AWS::ECR::PublicRepository";
    ResourceType["ECRRepository"] = "AWS::ECR::Repository";
    ResourceType["ECSCluster"] = "AWS::ECS::Cluster";
    ResourceType["ECSService"] = "AWS::ECS::Service";
    ResourceType["ECSTaskDefinition"] = "AWS::ECS::TaskDefinition";
    ResourceType["EFSAccessPoint"] = "AWS::EFS::AccessPoint";
    ResourceType["EFSFileSystem"] = "AWS::EFS::FileSystem";
    ResourceType["EIP"] = "AWS::EC2::EIP";
    ResourceType["EKSCluster"] = "AWS::EKS::Cluster";
    ResourceType["EKSFargateProfile"] = "AWS::EKS::FargateProfile";
    ResourceType["EMRSecurityConfiguration"] = "AWS::EMR::SecurityConfiguration";
    ResourceType["EgressOnlyInternetGateway"] = "AWS::EC2::EgressOnlyInternetGateway";
    ResourceType["EncryptionConfig"] = "AWS::XRay::EncryptionConfig";
    ResourceType["Environment"] = "AWS::ElasticBeanstalk::Environment";
    ResourceType["EventSubscription"] = "AWS::RDS::EventSubscription";
    ResourceType["FileData"] = "AWS::SSM::FileData";
    ResourceType["FlowLog"] = "AWS::EC2::FlowLog";
    ResourceType["Function"] = "AWS::Lambda::Function";
    ResourceType["GlobalAcceleratorAccelerator"] = "AWS::GlobalAccelerator::Accelerator";
    ResourceType["GlobalAcceleratorEndpointGroup"] = "AWS::GlobalAccelerator::EndpointGroup";
    ResourceType["GlobalAcceleratorListener"] = "AWS::GlobalAccelerator::Listener";
    ResourceType["GlueJob"] = "AWS::Glue::Job";
    ResourceType["Group"] = "AWS::IAM::Group";
    ResourceType["GuardDutyDetector"] = "AWS::GuardDuty::Detector";
    ResourceType["GuardDutyIPSet"] = "AWS::GuardDuty::IPSet";
    ResourceType["GuardDutyThreatIntelSet"] = "AWS::GuardDuty::ThreatIntelSet";
    ResourceType["Host"] = "AWS::EC2::Host";
    ResourceType["IPSetV2"] = "AWS::WAFv2::IPSet";
    ResourceType["Instance"] = "AWS::EC2::Instance";
    ResourceType["InternetGateway"] = "AWS::EC2::InternetGateway";
    ResourceType["Key"] = "AWS::KMS::Key";
    ResourceType["KinesisStream"] = "AWS::Kinesis::Stream";
    ResourceType["KinesisStreamConsumer"] = "AWS::Kinesis::StreamConsumer";
    ResourceType["LaunchConfiguration"] = "AWS::AutoScaling::LaunchConfiguration";
    ResourceType["LaunchTemplate"] = "AWS::EC2::LaunchTemplate";
    ResourceType["ListenerV2"] = "AWS::ElasticLoadBalancingV2::Listener";
    ResourceType["LoadBalancer"] = "AWS::ElasticLoadBalancing::LoadBalancer";
    ResourceType["LoadBalancerV2"] = "AWS::ElasticLoadBalancingV2::LoadBalancer";
    ResourceType["MSKCluster"] = "AWS::MSK::Cluster";
    ResourceType["ManagedInstanceInventory"] = "AWS::SSM::ManagedInstanceInventory";
    ResourceType["ManagedRuleSetV2"] = "AWS::WAFv2::ManagedRuleSet";
    ResourceType["NatGateway"] = "AWS::EC2::NatGateway";
    ResourceType["NetworkAcl"] = "AWS::EC2::NetworkAcl";
    ResourceType["NetworkFirewallFirewall"] = "AWS::NetworkFirewall::Firewall";
    ResourceType["NetworkFirewallFirewallPolicy"] = "AWS::NetworkFirewall::FirewallPolicy";
    ResourceType["NetworkFirewallRuleGroup"] = "AWS::NetworkFirewall::RuleGroup";
    ResourceType["NetworkInsightsAccessScopeAnalysis"] = "AWS::EC2::NetworkInsightsAccessScopeAnalysis";
    ResourceType["NetworkInterface"] = "AWS::EC2::NetworkInterface";
    ResourceType["OpenSearchDomain"] = "AWS::OpenSearch::Domain";
    ResourceType["PatchCompliance"] = "AWS::SSM::PatchCompliance";
    ResourceType["Pipeline"] = "AWS::CodePipeline::Pipeline";
    ResourceType["Policy"] = "AWS::IAM::Policy";
    ResourceType["Portfolio"] = "AWS::ServiceCatalog::Portfolio";
    ResourceType["Project"] = "AWS::CodeBuild::Project";
    ResourceType["Protection"] = "AWS::Shield::Protection";
    ResourceType["QLDBLedger"] = "AWS::QLDB::Ledger";
    ResourceType["Queue"] = "AWS::SQS::Queue";
    ResourceType["RateBasedRule"] = "AWS::WAF::RateBasedRule";
    ResourceType["RedshiftEventSubscription"] = "AWS::Redshift::EventSubscription";
    ResourceType["RegexPatternSetV2"] = "AWS::WAFv2::RegexPatternSet";
    ResourceType["RegionalProtection"] = "AWS::ShieldRegional::Protection";
    ResourceType["RegionalRateBasedRule"] = "AWS::WAFRegional::RateBasedRule";
    ResourceType["RegionalRule"] = "AWS::WAFRegional::Rule";
    ResourceType["RegionalRuleGroup"] = "AWS::WAFRegional::RuleGroup";
    ResourceType["RegionalWebACL"] = "AWS::WAFRegional::WebACL";
    ResourceType["RegisteredHAInstance"] = "AWS::EC2::RegisteredHAInstance";
    ResourceType["ResourceCompliance"] = "AWS::Config::ResourceCompliance";
    ResourceType["RestApi"] = "AWS::ApiGateway::RestApi";
    ResourceType["Role"] = "AWS::IAM::Role";
    ResourceType["Route53HostedZone"] = "AWS::Route53::HostedZone";
    ResourceType["Route53ResolverResolverEndpoint"] = "AWS::Route53Resolver::ResolverEndpoint";
    ResourceType["Route53ResolverResolverRule"] = "AWS::Route53Resolver::ResolverRule";
    ResourceType["Route53ResolverResolverRuleAssociation"] = "AWS::Route53Resolver::ResolverRuleAssociation";
    ResourceType["RouteTable"] = "AWS::EC2::RouteTable";
    ResourceType["Rule"] = "AWS::WAF::Rule";
    ResourceType["RuleGroup"] = "AWS::WAF::RuleGroup";
    ResourceType["RuleGroupV2"] = "AWS::WAFv2::RuleGroup";
    ResourceType["SESConfigurationSet"] = "AWS::SES::ConfigurationSet";
    ResourceType["SESContactList"] = "AWS::SES::ContactList";
    ResourceType["SageMakerCodeRepository"] = "AWS::SageMaker::CodeRepository";
    ResourceType["SageMakerModel"] = "AWS::SageMaker::Model";
    ResourceType["SageMakerNotebookInstanceLifecycleConfig"] = "AWS::SageMaker::NotebookInstanceLifecycleConfig";
    ResourceType["SageMakerWorkteam"] = "AWS::SageMaker::Workteam";
    ResourceType["ScalingPolicy"] = "AWS::AutoScaling::ScalingPolicy";
    ResourceType["ScheduledAction"] = "AWS::AutoScaling::ScheduledAction";
    ResourceType["Secret"] = "AWS::SecretsManager::Secret";
    ResourceType["SecurityGroup"] = "AWS::EC2::SecurityGroup";
    ResourceType["ServiceDiscoveryPublicDnsNamespace"] = "AWS::ServiceDiscovery::PublicDnsNamespace";
    ResourceType["ServiceDiscoveryService"] = "AWS::ServiceDiscovery::Service";
    ResourceType["Stack"] = "AWS::CloudFormation::Stack";
    ResourceType["Stage"] = "AWS::ApiGateway::Stage";
    ResourceType["StageV2"] = "AWS::ApiGatewayV2::Stage";
    ResourceType["StepFunctionsActivity"] = "AWS::StepFunctions::Activity";
    ResourceType["StepFunctionsStateMachine"] = "AWS::StepFunctions::StateMachine";
    ResourceType["StreamingDistribution"] = "AWS::CloudFront::StreamingDistribution";
    ResourceType["Subnet"] = "AWS::EC2::Subnet";
    ResourceType["Table"] = "AWS::DynamoDB::Table";
    ResourceType["Topic"] = "AWS::SNS::Topic";
    ResourceType["Trail"] = "AWS::CloudTrail::Trail";
    ResourceType["TransitGateway"] = "AWS::EC2::TransitGateway";
    ResourceType["TransitGatewayAttachment"] = "AWS::EC2::TransitGatewayAttachment";
    ResourceType["TransitGatewayRouteTable"] = "AWS::EC2::TransitGatewayRouteTable";
    ResourceType["User"] = "AWS::IAM::User";
    ResourceType["VPC"] = "AWS::EC2::VPC";
    ResourceType["VPCEndpoint"] = "AWS::EC2::VPCEndpoint";
    ResourceType["VPCEndpointService"] = "AWS::EC2::VPCEndpointService";
    ResourceType["VPCPeeringConnection"] = "AWS::EC2::VPCPeeringConnection";
    ResourceType["VPNConnection"] = "AWS::EC2::VPNConnection";
    ResourceType["VPNGateway"] = "AWS::EC2::VPNGateway";
    ResourceType["Volume"] = "AWS::EC2::Volume";
    ResourceType["WebACL"] = "AWS::WAF::WebACL";
    ResourceType["WebACLV2"] = "AWS::WAFv2::WebACL";
    ResourceType["WorkSpacesConnectionAlias"] = "AWS::WorkSpaces::ConnectionAlias";
    ResourceType["WorkSpacesWorkspace"] = "AWS::WorkSpaces::Workspace";
})(ResourceType || (ResourceType = {}));
export var ConfigurationItemStatus;
(function (ConfigurationItemStatus) {
    ConfigurationItemStatus["OK"] = "OK";
    ConfigurationItemStatus["ResourceDeleted"] = "ResourceDeleted";
    ConfigurationItemStatus["ResourceDeletedNotRecorded"] = "ResourceDeletedNotRecorded";
    ConfigurationItemStatus["ResourceDiscovered"] = "ResourceDiscovered";
    ConfigurationItemStatus["ResourceNotRecorded"] = "ResourceNotRecorded";
})(ConfigurationItemStatus || (ConfigurationItemStatus = {}));
export class NoSuchConfigurationAggregatorException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchConfigurationAggregatorException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchConfigurationAggregatorException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchConfigurationAggregatorException.prototype);
    }
}
export class ValidationException extends __BaseException {
    constructor(opts) {
        super({
            name: "ValidationException",
            $fault: "client",
            ...opts,
        });
        this.name = "ValidationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ValidationException.prototype);
    }
}
export class NoAvailableConfigurationRecorderException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoAvailableConfigurationRecorderException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoAvailableConfigurationRecorderException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoAvailableConfigurationRecorderException.prototype);
    }
}
export var ChronologicalOrder;
(function (ChronologicalOrder) {
    ChronologicalOrder["Forward"] = "Forward";
    ChronologicalOrder["Reverse"] = "Reverse";
})(ChronologicalOrder || (ChronologicalOrder = {}));
export var DeliveryStatus;
(function (DeliveryStatus) {
    DeliveryStatus["Failure"] = "Failure";
    DeliveryStatus["Not_Applicable"] = "Not_Applicable";
    DeliveryStatus["Success"] = "Success";
})(DeliveryStatus || (DeliveryStatus = {}));
export var ConfigRuleState;
(function (ConfigRuleState) {
    ConfigRuleState["ACTIVE"] = "ACTIVE";
    ConfigRuleState["DELETING"] = "DELETING";
    ConfigRuleState["DELETING_RESULTS"] = "DELETING_RESULTS";
    ConfigRuleState["EVALUATING"] = "EVALUATING";
})(ConfigRuleState || (ConfigRuleState = {}));
export var MaximumExecutionFrequency;
(function (MaximumExecutionFrequency) {
    MaximumExecutionFrequency["One_Hour"] = "One_Hour";
    MaximumExecutionFrequency["Six_Hours"] = "Six_Hours";
    MaximumExecutionFrequency["Three_Hours"] = "Three_Hours";
    MaximumExecutionFrequency["Twelve_Hours"] = "Twelve_Hours";
    MaximumExecutionFrequency["TwentyFour_Hours"] = "TwentyFour_Hours";
})(MaximumExecutionFrequency || (MaximumExecutionFrequency = {}));
export var Owner;
(function (Owner) {
    Owner["Aws"] = "AWS";
    Owner["Custom_Lambda"] = "CUSTOM_LAMBDA";
    Owner["Custom_Policy"] = "CUSTOM_POLICY";
})(Owner || (Owner = {}));
export var EventSource;
(function (EventSource) {
    EventSource["Aws_Config"] = "aws.config";
})(EventSource || (EventSource = {}));
export var MessageType;
(function (MessageType) {
    MessageType["ConfigurationItemChangeNotification"] = "ConfigurationItemChangeNotification";
    MessageType["ConfigurationSnapshotDeliveryCompleted"] = "ConfigurationSnapshotDeliveryCompleted";
    MessageType["OversizedConfigurationItemChangeNotification"] = "OversizedConfigurationItemChangeNotification";
    MessageType["ScheduledNotification"] = "ScheduledNotification";
})(MessageType || (MessageType = {}));
export var ConfigRuleComplianceSummaryGroupKey;
(function (ConfigRuleComplianceSummaryGroupKey) {
    ConfigRuleComplianceSummaryGroupKey["ACCOUNT_ID"] = "ACCOUNT_ID";
    ConfigRuleComplianceSummaryGroupKey["AWS_REGION"] = "AWS_REGION";
})(ConfigRuleComplianceSummaryGroupKey || (ConfigRuleComplianceSummaryGroupKey = {}));
export var RecorderStatus;
(function (RecorderStatus) {
    RecorderStatus["Failure"] = "Failure";
    RecorderStatus["Pending"] = "Pending";
    RecorderStatus["Success"] = "Success";
})(RecorderStatus || (RecorderStatus = {}));
export var ConformancePackState;
(function (ConformancePackState) {
    ConformancePackState["CREATE_COMPLETE"] = "CREATE_COMPLETE";
    ConformancePackState["CREATE_FAILED"] = "CREATE_FAILED";
    ConformancePackState["CREATE_IN_PROGRESS"] = "CREATE_IN_PROGRESS";
    ConformancePackState["DELETE_FAILED"] = "DELETE_FAILED";
    ConformancePackState["DELETE_IN_PROGRESS"] = "DELETE_IN_PROGRESS";
})(ConformancePackState || (ConformancePackState = {}));
export class ConformancePackTemplateValidationException extends __BaseException {
    constructor(opts) {
        super({
            name: "ConformancePackTemplateValidationException",
            $fault: "client",
            ...opts,
        });
        this.name = "ConformancePackTemplateValidationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ConformancePackTemplateValidationException.prototype);
    }
}
export class InvalidParameterValueException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidParameterValueException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidParameterValueException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidParameterValueException.prototype);
    }
}
export class NoSuchConfigRuleException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchConfigRuleException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchConfigRuleException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchConfigRuleException.prototype);
    }
}
export class ResourceInUseException extends __BaseException {
    constructor(opts) {
        super({
            name: "ResourceInUseException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceInUseException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceInUseException.prototype);
    }
}
export class NoSuchConfigurationRecorderException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchConfigurationRecorderException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchConfigurationRecorderException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchConfigurationRecorderException.prototype);
    }
}
export class NoSuchConformancePackException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchConformancePackException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchConformancePackException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchConformancePackException.prototype);
    }
}
export class LastDeliveryChannelDeleteFailedException extends __BaseException {
    constructor(opts) {
        super({
            name: "LastDeliveryChannelDeleteFailedException",
            $fault: "client",
            ...opts,
        });
        this.name = "LastDeliveryChannelDeleteFailedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, LastDeliveryChannelDeleteFailedException.prototype);
    }
}
export class NoSuchDeliveryChannelException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchDeliveryChannelException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchDeliveryChannelException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchDeliveryChannelException.prototype);
    }
}
export class NoSuchOrganizationConfigRuleException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchOrganizationConfigRuleException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchOrganizationConfigRuleException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchOrganizationConfigRuleException.prototype);
    }
}
export class OrganizationAccessDeniedException extends __BaseException {
    constructor(opts) {
        super({
            name: "OrganizationAccessDeniedException",
            $fault: "client",
            ...opts,
        });
        this.name = "OrganizationAccessDeniedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, OrganizationAccessDeniedException.prototype);
    }
}
export class NoSuchOrganizationConformancePackException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchOrganizationConformancePackException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchOrganizationConformancePackException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchOrganizationConformancePackException.prototype);
    }
}
export class InsufficientPermissionsException extends __BaseException {
    constructor(opts) {
        super({
            name: "InsufficientPermissionsException",
            $fault: "client",
            ...opts,
        });
        this.name = "InsufficientPermissionsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InsufficientPermissionsException.prototype);
    }
}
export class NoSuchRemediationConfigurationException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchRemediationConfigurationException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchRemediationConfigurationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchRemediationConfigurationException.prototype);
    }
}
export class RemediationInProgressException extends __BaseException {
    constructor(opts) {
        super({
            name: "RemediationInProgressException",
            $fault: "client",
            ...opts,
        });
        this.name = "RemediationInProgressException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RemediationInProgressException.prototype);
    }
}
export class NoSuchRemediationExceptionException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchRemediationExceptionException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchRemediationExceptionException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchRemediationExceptionException.prototype);
    }
}
export class NoRunningConfigurationRecorderException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoRunningConfigurationRecorderException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoRunningConfigurationRecorderException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoRunningConfigurationRecorderException.prototype);
    }
}
export class NoSuchRetentionConfigurationException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchRetentionConfigurationException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchRetentionConfigurationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchRetentionConfigurationException.prototype);
    }
}
export class ResourceNotFoundException extends __BaseException {
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceNotFoundException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
    }
}
export class InvalidLimitException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidLimitException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidLimitException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidLimitException.prototype);
    }
}
export class InvalidNextTokenException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidNextTokenException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidNextTokenException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidNextTokenException.prototype);
    }
}
export class NoSuchConfigRuleInConformancePackException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchConfigRuleInConformancePackException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchConfigRuleInConformancePackException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchConfigRuleInConformancePackException.prototype);
    }
}
export var OrganizationConfigRuleTriggerTypeNoSN;
(function (OrganizationConfigRuleTriggerTypeNoSN) {
    OrganizationConfigRuleTriggerTypeNoSN["CONFIGURATION_ITEM_CHANGE_NOTIFICATION"] = "ConfigurationItemChangeNotification";
    OrganizationConfigRuleTriggerTypeNoSN["OVERSIZED_CONFIGURATION_ITEM_CHANGE_NOTIFCATION"] = "OversizedConfigurationItemChangeNotification";
})(OrganizationConfigRuleTriggerTypeNoSN || (OrganizationConfigRuleTriggerTypeNoSN = {}));
export var OrganizationConfigRuleTriggerType;
(function (OrganizationConfigRuleTriggerType) {
    OrganizationConfigRuleTriggerType["CONFIGURATION_ITEM_CHANGE_NOTIFICATION"] = "ConfigurationItemChangeNotification";
    OrganizationConfigRuleTriggerType["OVERSIZED_CONFIGURATION_ITEM_CHANGE_NOTIFCATION"] = "OversizedConfigurationItemChangeNotification";
    OrganizationConfigRuleTriggerType["SCHEDULED_NOTIFICATION"] = "ScheduledNotification";
})(OrganizationConfigRuleTriggerType || (OrganizationConfigRuleTriggerType = {}));
export var OrganizationRuleStatus;
(function (OrganizationRuleStatus) {
    OrganizationRuleStatus["CREATE_FAILED"] = "CREATE_FAILED";
    OrganizationRuleStatus["CREATE_IN_PROGRESS"] = "CREATE_IN_PROGRESS";
    OrganizationRuleStatus["CREATE_SUCCESSFUL"] = "CREATE_SUCCESSFUL";
    OrganizationRuleStatus["DELETE_FAILED"] = "DELETE_FAILED";
    OrganizationRuleStatus["DELETE_IN_PROGRESS"] = "DELETE_IN_PROGRESS";
    OrganizationRuleStatus["DELETE_SUCCESSFUL"] = "DELETE_SUCCESSFUL";
    OrganizationRuleStatus["UPDATE_FAILED"] = "UPDATE_FAILED";
    OrganizationRuleStatus["UPDATE_IN_PROGRESS"] = "UPDATE_IN_PROGRESS";
    OrganizationRuleStatus["UPDATE_SUCCESSFUL"] = "UPDATE_SUCCESSFUL";
})(OrganizationRuleStatus || (OrganizationRuleStatus = {}));
export var OrganizationResourceStatus;
(function (OrganizationResourceStatus) {
    OrganizationResourceStatus["CREATE_FAILED"] = "CREATE_FAILED";
    OrganizationResourceStatus["CREATE_IN_PROGRESS"] = "CREATE_IN_PROGRESS";
    OrganizationResourceStatus["CREATE_SUCCESSFUL"] = "CREATE_SUCCESSFUL";
    OrganizationResourceStatus["DELETE_FAILED"] = "DELETE_FAILED";
    OrganizationResourceStatus["DELETE_IN_PROGRESS"] = "DELETE_IN_PROGRESS";
    OrganizationResourceStatus["DELETE_SUCCESSFUL"] = "DELETE_SUCCESSFUL";
    OrganizationResourceStatus["UPDATE_FAILED"] = "UPDATE_FAILED";
    OrganizationResourceStatus["UPDATE_IN_PROGRESS"] = "UPDATE_IN_PROGRESS";
    OrganizationResourceStatus["UPDATE_SUCCESSFUL"] = "UPDATE_SUCCESSFUL";
})(OrganizationResourceStatus || (OrganizationResourceStatus = {}));
export var ResourceValueType;
(function (ResourceValueType) {
    ResourceValueType["RESOURCE_ID"] = "RESOURCE_ID";
})(ResourceValueType || (ResourceValueType = {}));
export var RemediationTargetType;
(function (RemediationTargetType) {
    RemediationTargetType["SSM_DOCUMENT"] = "SSM_DOCUMENT";
})(RemediationTargetType || (RemediationTargetType = {}));
export var RemediationExecutionState;
(function (RemediationExecutionState) {
    RemediationExecutionState["FAILED"] = "FAILED";
    RemediationExecutionState["IN_PROGRESS"] = "IN_PROGRESS";
    RemediationExecutionState["QUEUED"] = "QUEUED";
    RemediationExecutionState["SUCCEEDED"] = "SUCCEEDED";
})(RemediationExecutionState || (RemediationExecutionState = {}));
export var RemediationExecutionStepState;
(function (RemediationExecutionStepState) {
    RemediationExecutionStepState["FAILED"] = "FAILED";
    RemediationExecutionStepState["PENDING"] = "PENDING";
    RemediationExecutionStepState["SUCCEEDED"] = "SUCCEEDED";
})(RemediationExecutionStepState || (RemediationExecutionStepState = {}));
export var ResourceCountGroupKey;
(function (ResourceCountGroupKey) {
    ResourceCountGroupKey["ACCOUNT_ID"] = "ACCOUNT_ID";
    ResourceCountGroupKey["AWS_REGION"] = "AWS_REGION";
    ResourceCountGroupKey["RESOURCE_TYPE"] = "RESOURCE_TYPE";
})(ResourceCountGroupKey || (ResourceCountGroupKey = {}));
export class OversizedConfigurationItemException extends __BaseException {
    constructor(opts) {
        super({
            name: "OversizedConfigurationItemException",
            $fault: "client",
            ...opts,
        });
        this.name = "OversizedConfigurationItemException";
        this.$fault = "client";
        Object.setPrototypeOf(this, OversizedConfigurationItemException.prototype);
    }
}
export class ResourceNotDiscoveredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ResourceNotDiscoveredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceNotDiscoveredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceNotDiscoveredException.prototype);
    }
}
export var MemberAccountRuleStatus;
(function (MemberAccountRuleStatus) {
    MemberAccountRuleStatus["CREATE_FAILED"] = "CREATE_FAILED";
    MemberAccountRuleStatus["CREATE_IN_PROGRESS"] = "CREATE_IN_PROGRESS";
    MemberAccountRuleStatus["CREATE_SUCCESSFUL"] = "CREATE_SUCCESSFUL";
    MemberAccountRuleStatus["DELETE_FAILED"] = "DELETE_FAILED";
    MemberAccountRuleStatus["DELETE_IN_PROGRESS"] = "DELETE_IN_PROGRESS";
    MemberAccountRuleStatus["DELETE_SUCCESSFUL"] = "DELETE_SUCCESSFUL";
    MemberAccountRuleStatus["UPDATE_FAILED"] = "UPDATE_FAILED";
    MemberAccountRuleStatus["UPDATE_IN_PROGRESS"] = "UPDATE_IN_PROGRESS";
    MemberAccountRuleStatus["UPDATE_SUCCESSFUL"] = "UPDATE_SUCCESSFUL";
})(MemberAccountRuleStatus || (MemberAccountRuleStatus = {}));
export var OrganizationResourceDetailedStatus;
(function (OrganizationResourceDetailedStatus) {
    OrganizationResourceDetailedStatus["CREATE_FAILED"] = "CREATE_FAILED";
    OrganizationResourceDetailedStatus["CREATE_IN_PROGRESS"] = "CREATE_IN_PROGRESS";
    OrganizationResourceDetailedStatus["CREATE_SUCCESSFUL"] = "CREATE_SUCCESSFUL";
    OrganizationResourceDetailedStatus["DELETE_FAILED"] = "DELETE_FAILED";
    OrganizationResourceDetailedStatus["DELETE_IN_PROGRESS"] = "DELETE_IN_PROGRESS";
    OrganizationResourceDetailedStatus["DELETE_SUCCESSFUL"] = "DELETE_SUCCESSFUL";
    OrganizationResourceDetailedStatus["UPDATE_FAILED"] = "UPDATE_FAILED";
    OrganizationResourceDetailedStatus["UPDATE_IN_PROGRESS"] = "UPDATE_IN_PROGRESS";
    OrganizationResourceDetailedStatus["UPDATE_SUCCESSFUL"] = "UPDATE_SUCCESSFUL";
})(OrganizationResourceDetailedStatus || (OrganizationResourceDetailedStatus = {}));
export class InvalidTimeRangeException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidTimeRangeException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidTimeRangeException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidTimeRangeException.prototype);
    }
}
export class InsufficientDeliveryPolicyException extends __BaseException {
    constructor(opts) {
        super({
            name: "InsufficientDeliveryPolicyException",
            $fault: "client",
            ...opts,
        });
        this.name = "InsufficientDeliveryPolicyException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InsufficientDeliveryPolicyException.prototype);
    }
}
export class InvalidConfigurationRecorderNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidConfigurationRecorderNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidConfigurationRecorderNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidConfigurationRecorderNameException.prototype);
    }
}
export class InvalidDeliveryChannelNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidDeliveryChannelNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidDeliveryChannelNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidDeliveryChannelNameException.prototype);
    }
}
export class InvalidExpressionException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidExpressionException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidExpressionException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidExpressionException.prototype);
    }
}
export class InvalidRecordingGroupException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRecordingGroupException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRecordingGroupException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRecordingGroupException.prototype);
    }
}
export class InvalidResultTokenException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidResultTokenException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidResultTokenException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidResultTokenException.prototype);
    }
}
export class InvalidRoleException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRoleException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRoleException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRoleException.prototype);
    }
}
export class InvalidS3KeyPrefixException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidS3KeyPrefixException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidS3KeyPrefixException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidS3KeyPrefixException.prototype);
    }
}
export class InvalidS3KmsKeyArnException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidS3KmsKeyArnException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidS3KmsKeyArnException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidS3KmsKeyArnException.prototype);
    }
}
export class InvalidSNSTopicARNException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidSNSTopicARNException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidSNSTopicARNException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidSNSTopicARNException.prototype);
    }
}
export class LimitExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "LimitExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "LimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, LimitExceededException.prototype);
    }
}
export var SortBy;
(function (SortBy) {
    SortBy["SCORE"] = "SCORE";
})(SortBy || (SortBy = {}));
export var SortOrder;
(function (SortOrder) {
    SortOrder["ASCENDING"] = "ASCENDING";
    SortOrder["DESCENDING"] = "DESCENDING";
})(SortOrder || (SortOrder = {}));
export class MaxActiveResourcesExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaxActiveResourcesExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaxActiveResourcesExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaxActiveResourcesExceededException.prototype);
    }
}
export class MaxNumberOfConfigRulesExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaxNumberOfConfigRulesExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaxNumberOfConfigRulesExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaxNumberOfConfigRulesExceededException.prototype);
    }
}
export class MaxNumberOfConfigurationRecordersExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaxNumberOfConfigurationRecordersExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaxNumberOfConfigurationRecordersExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaxNumberOfConfigurationRecordersExceededException.prototype);
    }
}
export class MaxNumberOfConformancePacksExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaxNumberOfConformancePacksExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaxNumberOfConformancePacksExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaxNumberOfConformancePacksExceededException.prototype);
    }
}
export class MaxNumberOfDeliveryChannelsExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaxNumberOfDeliveryChannelsExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaxNumberOfDeliveryChannelsExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaxNumberOfDeliveryChannelsExceededException.prototype);
    }
}
export class MaxNumberOfOrganizationConfigRulesExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaxNumberOfOrganizationConfigRulesExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaxNumberOfOrganizationConfigRulesExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaxNumberOfOrganizationConfigRulesExceededException.prototype);
    }
}
export class MaxNumberOfOrganizationConformancePacksExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaxNumberOfOrganizationConformancePacksExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaxNumberOfOrganizationConformancePacksExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaxNumberOfOrganizationConformancePacksExceededException.prototype);
    }
}
export class MaxNumberOfRetentionConfigurationsExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaxNumberOfRetentionConfigurationsExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaxNumberOfRetentionConfigurationsExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaxNumberOfRetentionConfigurationsExceededException.prototype);
    }
}
export class NoAvailableDeliveryChannelException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoAvailableDeliveryChannelException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoAvailableDeliveryChannelException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoAvailableDeliveryChannelException.prototype);
    }
}
export class NoAvailableOrganizationException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoAvailableOrganizationException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoAvailableOrganizationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoAvailableOrganizationException.prototype);
    }
}
export class NoSuchBucketException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoSuchBucketException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoSuchBucketException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoSuchBucketException.prototype);
    }
}
export class OrganizationAllFeaturesNotEnabledException extends __BaseException {
    constructor(opts) {
        super({
            name: "OrganizationAllFeaturesNotEnabledException",
            $fault: "client",
            ...opts,
        });
        this.name = "OrganizationAllFeaturesNotEnabledException";
        this.$fault = "client";
        Object.setPrototypeOf(this, OrganizationAllFeaturesNotEnabledException.prototype);
    }
}
export class OrganizationConformancePackTemplateValidationException extends __BaseException {
    constructor(opts) {
        super({
            name: "OrganizationConformancePackTemplateValidationException",
            $fault: "client",
            ...opts,
        });
        this.name = "OrganizationConformancePackTemplateValidationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, OrganizationConformancePackTemplateValidationException.prototype);
    }
}
export const AccountAggregationSourceFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ComplianceContributorCountFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ComplianceFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregateComplianceByConfigRuleFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregateConformancePackComplianceFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregateComplianceByConformancePackFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ComplianceSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregateComplianceCountFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregateConformancePackComplianceCountFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregateConformancePackComplianceFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregateConformancePackComplianceSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregateConformancePackComplianceSummaryFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregatedSourceStatusFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const EvaluationResultQualifierFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const EvaluationResultIdentifierFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregateEvaluationResultFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregateResourceIdentifierFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const AggregationAuthorizationFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const BaseConfigurationItemFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const BatchGetAggregateResourceConfigRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const BatchGetAggregateResourceConfigResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ResourceKeyFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const BatchGetResourceConfigRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const BatchGetResourceConfigResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ComplianceByConfigRuleFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ComplianceByResourceFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ComplianceSummaryByResourceTypeFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigExportDeliveryInfoFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ScopeFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const CustomPolicyDetailsFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const SourceDetailFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const SourceFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigRuleFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigRuleComplianceFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigRuleComplianceSummaryFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigRuleEvaluationStatusFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigSnapshotDeliveryPropertiesFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigStreamDeliveryInfoFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationAggregationSourceFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigurationAggregatorFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const RelationshipFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigurationItemFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const RecordingGroupFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigurationRecorderFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConfigurationRecorderStatusFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConformancePackComplianceFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConformancePackComplianceScoreFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConformancePackComplianceScoresFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConformancePackComplianceSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConformancePackInputParameterFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const TemplateSSMDocumentDetailsFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConformancePackDetailFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConformancePackEvaluationFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConformancePackEvaluationResultFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConformancePackRuleComplianceFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ConformancePackStatusDetailFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteAggregationAuthorizationRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteConfigRuleRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteConfigurationAggregatorRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteConfigurationRecorderRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteConformancePackRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteDeliveryChannelRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteEvaluationResultsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteEvaluationResultsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteOrganizationConfigRuleRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteOrganizationConformancePackRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeletePendingAggregationRequestRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteRemediationConfigurationRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteRemediationConfigurationResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const RemediationExceptionResourceKeyFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteRemediationExceptionsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const FailedDeleteRemediationExceptionsBatchFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteRemediationExceptionsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteResourceConfigRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteRetentionConfigurationRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteStoredQueryRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeleteStoredQueryResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeliverConfigSnapshotRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeliverConfigSnapshotResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeliveryChannelFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DeliveryChannelStatusFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeAggregateComplianceByConfigRulesRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeAggregateComplianceByConfigRulesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeAggregateComplianceByConformancePacksRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeAggregateComplianceByConformancePacksResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeAggregationAuthorizationsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeAggregationAuthorizationsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeComplianceByConfigRuleRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeComplianceByConfigRuleResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeComplianceByResourceRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeComplianceByResourceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigRuleEvaluationStatusRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigRuleEvaluationStatusResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigRulesRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigRulesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigurationAggregatorsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigurationAggregatorsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigurationAggregatorSourcesStatusRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigurationAggregatorSourcesStatusResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigurationRecordersRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigurationRecordersResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigurationRecorderStatusRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConfigurationRecorderStatusResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConformancePackComplianceRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConformancePackComplianceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConformancePacksRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConformancePacksResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConformancePackStatusRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeConformancePackStatusResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeDeliveryChannelsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeDeliveryChannelsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeDeliveryChannelStatusRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeDeliveryChannelStatusResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeOrganizationConfigRulesRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationCustomPolicyRuleMetadataNoPolicyFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationCustomRuleMetadataFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationManagedRuleMetadataFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationConfigRuleFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeOrganizationConfigRulesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeOrganizationConfigRuleStatusesRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationConfigRuleStatusFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeOrganizationConfigRuleStatusesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeOrganizationConformancePacksRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationConformancePackFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeOrganizationConformancePacksResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeOrganizationConformancePackStatusesRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationConformancePackStatusFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeOrganizationConformancePackStatusesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribePendingAggregationRequestsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PendingAggregationRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribePendingAggregationRequestsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeRemediationConfigurationsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const SsmControlsFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ExecutionControlsFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ResourceValueFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const StaticValueFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const RemediationParameterValueFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const RemediationConfigurationFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeRemediationConfigurationsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeRemediationExceptionsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const RemediationExceptionFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeRemediationExceptionsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeRemediationExecutionStatusRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const RemediationExecutionStepFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const RemediationExecutionStatusFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeRemediationExecutionStatusResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeRetentionConfigurationsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const RetentionConfigurationFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeRetentionConfigurationsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const EvaluationFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const EvaluationResultFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ExternalEvaluationFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const FailedRemediationBatchFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const FailedRemediationExceptionBatchFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const FieldInfoFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetAggregateComplianceDetailsByConfigRuleRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetAggregateComplianceDetailsByConfigRuleResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetAggregateConfigRuleComplianceSummaryRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetAggregateConfigRuleComplianceSummaryResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetAggregateConformancePackComplianceSummaryRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetAggregateConformancePackComplianceSummaryResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ResourceCountFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetAggregateDiscoveredResourceCountsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GroupedResourceCountFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetAggregateDiscoveredResourceCountsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetAggregateResourceConfigRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetAggregateResourceConfigResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetComplianceDetailsByConfigRuleRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetComplianceDetailsByConfigRuleResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetComplianceDetailsByResourceRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetComplianceDetailsByResourceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetComplianceSummaryByConfigRuleResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetComplianceSummaryByResourceTypeRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetComplianceSummaryByResourceTypeResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetConformancePackComplianceDetailsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetConformancePackComplianceDetailsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetConformancePackComplianceSummaryRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetConformancePackComplianceSummaryResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetCustomRulePolicyRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetCustomRulePolicyResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetDiscoveredResourceCountsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ResourceCountFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetDiscoveredResourceCountsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const StatusDetailFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetOrganizationConfigRuleDetailedStatusRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const MemberAccountStatusFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetOrganizationConfigRuleDetailedStatusResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationResourceDetailedStatusFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetOrganizationConformancePackDetailedStatusRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationConformancePackDetailedStatusFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetOrganizationConformancePackDetailedStatusResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetOrganizationCustomRulePolicyRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetOrganizationCustomRulePolicyResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetResourceConfigHistoryRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetResourceConfigHistoryResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetStoredQueryRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const StoredQueryFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const GetStoredQueryResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ResourceFiltersFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ListAggregateDiscoveredResourcesRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ListAggregateDiscoveredResourcesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ListConformancePackComplianceScoresRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ListConformancePackComplianceScoresResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ListDiscoveredResourcesRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ResourceIdentifierFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ListDiscoveredResourcesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ListStoredQueriesRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const StoredQueryMetadataFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ListStoredQueriesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ListTagsForResourceRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const TagFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const ListTagsForResourceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const OrganizationCustomPolicyRuleMetadataFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutAggregationAuthorizationRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutAggregationAuthorizationResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutConfigRuleRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutConfigurationAggregatorRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutConfigurationAggregatorResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const PutConfigurationRecorderRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
