// @mitre/inspec-objects/lib/utilities/diffMarkdown.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDiffMarkdown = void 0;
const tslib_1 = require("tslib");
const mustache_1 = tslib_1.__importDefault(require("mustache"));
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const automatticUpdateTemplate_json_1 = tslib_1.__importDefault(require("../resources/automatticUpdateTemplate.json"));
function createDiffMarkdown(diff) {
    const renderableDiffData = {
        addedControls: Object.values(diff.ignoreFormattingDiff.addedControls),
        hasRenamedControls: false,
        renamedControls: [],
        updatedChecks: [],
        updatedFixes: [],
        updatedImpacts: [],
        updatedTitles: [],
        updatedDescriptions: [],
    };
    Object.entries(diff.ignoreFormattingDiff.renamedControlIDs).forEach(([oldId, newId]) => {
        renderableDiffData.hasRenamedControls = true;
        renderableDiffData.renamedControls.push({
            oldId: oldId,
            newId: newId,
        });
    });
    Object.entries(diff.rawDiff.changedControls).forEach(([id, controlDiff]) => {
        var _a, _b;
        if ((_a = controlDiff.descs) === null || _a === void 0 ? void 0 : _a.check) {
            const oldCheck = lodash_1.default.get(controlDiff.descs.check, '__old');
            const newCheck = lodash_1.default.get(controlDiff.descs.check, '__new');
            if (oldCheck.replace(/\n/g, '').replace(/\W/g, '') !==
                newCheck.replace(/\n/g, '').replace(/\W/g, '')) {
                renderableDiffData.updatedChecks.push({
                    id: id,
                    old: oldCheck,
                    new: newCheck,
                });
            }
        }
        if ((_b = controlDiff.descs) === null || _b === void 0 ? void 0 : _b.fix) {
            const oldFix = lodash_1.default.get(controlDiff.descs.fix, '__old');
            const newFix = lodash_1.default.get(controlDiff.descs.fix, '__new');
            if (oldFix.replace(/\n/g, '').replace(/\W/g, '') !==
                newFix.replace(/\n/g, '').replace(/\W/g, '')) {
                renderableDiffData.updatedFixes.push({
                    id: id,
                    old: oldFix,
                    new: newFix,
                });
            }
        }
        if (controlDiff.impact) {
            const oldImpact = lodash_1.default.get(controlDiff.impact, '__old');
            const newImpact = lodash_1.default.get(controlDiff.impact, '__new');
            if (oldImpact !== newImpact) {
                renderableDiffData.updatedImpacts.push({
                    id: id,
                    old: oldImpact,
                    new: newImpact,
                });
            }
        }
        if (controlDiff.title) {
            const oldTitle = lodash_1.default.get(controlDiff.title, '__old');
            const newTitle = lodash_1.default.get(controlDiff.title, '__new');
            if (oldTitle !== newTitle) {
                renderableDiffData.updatedTitles.push({
                    id: id,
                    old: oldTitle,
                    new: newTitle,
                });
            }
        }
        if (controlDiff.desc) {
            const oldDesc = lodash_1.default.get(controlDiff.desc, '__old');
            const newDesc = lodash_1.default.get(controlDiff.desc, '__new');
            if (oldDesc !== newDesc) {
                renderableDiffData.updatedDescriptions.push({
                    id: id,
                    old: oldDesc,
                    new: newDesc,
                });
            }
        }
    });
    // Render output
    return mustache_1.default.render(automatticUpdateTemplate_json_1.default.data, renderableDiffData);
}
exports.createDiffMarkdown = createDiffMarkdown;
