// @mitre/inspec-objects/lib/utilities/diff.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.diffProfile = exports.simplifyDiff = exports.removeNewlines = void 0;
const tslib_1 = require("tslib");
const json_diff_1 = require("json-diff");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const update_1 = require("./update");
function removeNewlines(control) {
    if (!control) {
        return {};
    }
    return lodash_1.default.mapValues(control, (value) => {
        if (typeof value === "string") {
            return value.replace(/\n/g, "{{{{newlineHERE}}}}").trim();
        }
        else if (typeof value === "object" && value !== null) {
            return removeNewlines(value);
        }
        return value;
    });
}
exports.removeNewlines = removeNewlines;
function simplifyDiff(diffData) {
    return lodash_1.default.transform(diffData, (result, diffValue, key) => {
        if (lodash_1.default.has(diffValue, "__new")) {
            // Remove any trailing space
            if (typeof lodash_1.default.get(diffValue, "__new") === "string" &&
                typeof lodash_1.default.get(diffValue, "__old") === "string") {
                if (lodash_1.default.get(diffValue, "__new").trim() !==
                    lodash_1.default.get(diffValue, "__old").trim()) {
                    lodash_1.default.set(result, key, lodash_1.default.get(diffValue, "__new"));
                }
            }
            else {
                result[key] = lodash_1.default.get(diffValue, "__new");
            }
        }
        else if (Array.isArray(diffValue)) {
            result[key] = diffValue
                .map((value) => value[0] === "+" && value[1])
                .filter((value) => value);
        }
        else if (typeof diffValue === "object") {
            result[key] = simplifyDiff(diffValue);
        }
        else if (key.endsWith("__deleted")) {
            return undefined;
        }
        else {
            result[key] = diffValue;
        }
    });
}
exports.simplifyDiff = simplifyDiff;
function diffProfile(fromProfile, toProfile, logger) {
    const profileDiff = {
        addedControlIDs: [],
        removedControlIDs: [],
        renamedControlIds: {},
        addedControls: {},
        changedControls: {},
    };
    const originalDiff = {
        addedControlIDs: [],
        removedControlIDs: [],
        renamedControlIds: {},
        addedControls: {},
        changedControls: {},
    };
    const fromControlIDs = fromProfile.controls
        .map((control) => control.id)
        .sort();
    const toControlIDs = toProfile.controls.map((control) => control.id).sort();
    // Find new controls
    const controlIDDiff = (0, json_diff_1.diff)(fromControlIDs, toControlIDs);
    // Contains the new IDs
    const changedControlIds = [];
    controlIDDiff === null || controlIDDiff === void 0 ? void 0 : controlIDDiff.forEach((diffValue) => {
        if (diffValue[0] === "-") {
            const existingControl = fromProfile.controls.find((control) => control.id === diffValue[1]);
            // Check if the control has been given a new ID
            if (existingControl) {
                const newControl = (0, update_1.findUpdatedControlByAllIdentifiers)(existingControl, toProfile.controls);
                if (newControl && newControl.id !== existingControl.id) {
                    profileDiff.renamedControlIds[existingControl.id] = newControl.id;
                    originalDiff.renamedControlIds[existingControl.id] = newControl.id;
                    changedControlIds.push(newControl.id.toLowerCase());
                    const controlDiff = lodash_1.default.omit((0, json_diff_1.diff)(existingControl, newControl), "code__deleted");
                    profileDiff.changedControls[newControl.id] = simplifyDiff(controlDiff);
                    originalDiff.changedControls[newControl.id] = controlDiff;
                    logger.verbose(`Control ${existingControl.id} has been upgraded to ${newControl.id}`);
                }
                else {
                    profileDiff.removedControlIDs.push(diffValue[1]);
                    originalDiff.removedControlIDs.push(diffValue[1]);
                }
            }
            else {
                logger.error(`Unable to find existing control ${diffValue[1]}`);
            }
        }
        else if (diffValue[0] === "+" && !changedControlIds.includes(diffValue[1].toLowerCase())) {
            profileDiff.addedControlIDs.push(diffValue[1]);
            originalDiff.addedControlIDs.push(diffValue[1]);
        }
    });
    // Add new controls to addedControls
    profileDiff.addedControlIDs.forEach((addedControl) => {
        const newControl = toProfile.controls.find((control) => addedControl === control.id);
        if (newControl && !profileDiff.changedControls[newControl.id]) {
            profileDiff.addedControls[addedControl] = newControl;
            originalDiff.addedControls[addedControl] = newControl;
        }
    });
    // Find changed controls
    for (const fromControl of fromProfile.controls) {
        const toControl = toProfile.controls.find((control) => control.id === fromControl.id);
        if (toControl) {
            const controlDiff = lodash_1.default.omit((0, json_diff_1.diff)(fromControl, toControl), "code__deleted");
            if (controlDiff) {
                profileDiff.changedControls[toControl.id] = simplifyDiff(controlDiff);
                originalDiff.changedControls[toControl.id] = controlDiff;
            }
        }
    }
    return { simplified: profileDiff, originalDiff: originalDiff };
}
exports.diffProfile = diffProfile;
