// inspecjs/lib/compat_wrappers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hdfWrapControl = exports.convertImpactToSeverity = exports.titleCasedSeverities = exports.severities = exports.lowercasedControlStatuses = exports.controlStatuses = void 0;
const compat_inspec_1_0_1 = require("./compat_impl/compat_inspec_1_0");
exports.controlStatuses = [
    'Not Applicable',
    'From Profile',
    'Profile Error',
    'Passed',
    'Failed',
    'Not Reviewed'
];
exports.lowercasedControlStatuses = [
    'not applicable',
    'from profile',
    'profile error',
    'passed',
    'failed',
    'not reviewed'
];
exports.severities = [
    'none',
    'low',
    'medium',
    'high',
    'critical'
];
exports.titleCasedSeverities = [
    'None',
    'Low',
    'Medium',
    'High',
    'Critical'
];
function convertImpactToSeverity(impact) {
    if (impact < 0.1) {
        return 'none';
    }
    else if (impact < 0.4) {
        return 'low';
    }
    else if (impact < 0.7) {
        return 'medium';
    }
    else if (impact < 0.9) {
        return 'high';
    }
    else {
        return 'critical';
    }
}
exports.convertImpactToSeverity = convertImpactToSeverity;
function hdfWrapControl(ctrl) {
    if (ctrl.results !== undefined) {
        const rctrl = ctrl;
        return new compat_inspec_1_0_1.ExecControl(rctrl);
    }
    else if (ctrl.results === undefined) {
        const rctrl = ctrl;
        return new compat_inspec_1_0_1.ProfileControl(rctrl);
    }
    else {
        throw new Error('Control did not match any expected schema');
    }
}
exports.hdfWrapControl = hdfWrapControl;
