// @oclif/core/lib/util/fs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readJsonSync = exports.readJson = exports.fileExists = exports.dirExists = exports.exists = exports.requireJson = void 0;
const promises_1 = require("node:fs/promises");
const node_path_1 = require("node:path");
const node_fs_1 = require("node:fs");
const debug = require('debug');
function requireJson(...pathParts) {
    return JSON.parse((0, node_fs_1.readFileSync)((0, node_path_1.join)(...pathParts), 'utf8'));
}
exports.requireJson = requireJson;
async function exists(path) {
    try {
        await (0, promises_1.access)(path);
        return true;
    }
    catch {
        return false;
    }
}
exports.exists = exists;
const dirExists = async (input) => {
    if (!(await exists(input))) {
        throw new Error(`No directory found at ${input}`);
    }
    const fileStat = await (0, promises_1.stat)(input);
    if (!fileStat.isDirectory()) {
        throw new Error(`${input} exists but is not a directory`);
    }
    return input;
};
exports.dirExists = dirExists;
const fileExists = async (input) => {
    if (!(await exists(input))) {
        throw new Error(`No file found at ${input}`);
    }
    const fileStat = await (0, promises_1.stat)(input);
    if (!fileStat.isFile()) {
        throw new Error(`${input} exists but is not a file`);
    }
    return input;
};
exports.fileExists = fileExists;
async function readJson(path) {
    debug('config')('readJson %s', path);
    const contents = await (0, promises_1.readFile)(path, 'utf8');
    return JSON.parse(contents);
}
exports.readJson = readJson;
function readJsonSync(path, parse = true) {
    const contents = (0, node_fs_1.readFileSync)(path, 'utf8');
    return parse ? JSON.parse(contents) : contents;
}
exports.readJsonSync = readJsonSync;
