// @aws-sdk/client-config-service/dist-es/commands/BatchGetAggregateResourceConfigCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { BatchGetAggregateResourceConfigRequest, BatchGetAggregateResourceConfigResponse } from "../models/models_0";
import { deserializeAws_json1_1BatchGetAggregateResourceConfigCommand, serializeAws_json1_1BatchGetAggregateResourceConfigCommand, } from "../protocols/Aws_json1_1";
var BatchGetAggregateResourceConfigCommand = (function (_super) {
    __extends(BatchGetAggregateResourceConfigCommand, _super);
    function BatchGetAggregateResourceConfigCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    BatchGetAggregateResourceConfigCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "BatchGetAggregateResourceConfigCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: BatchGetAggregateResourceConfigRequest.filterSensitiveLog,
            outputFilterSensitiveLog: BatchGetAggregateResourceConfigResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    BatchGetAggregateResourceConfigCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1BatchGetAggregateResourceConfigCommand(input, context);
    };
    BatchGetAggregateResourceConfigCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1BatchGetAggregateResourceConfigCommand(output, context);
    };
    return BatchGetAggregateResourceConfigCommand;
}($Command));
export { BatchGetAggregateResourceConfigCommand };
