// @aws-sdk/client-config-service/dist-es/commands/BatchGetResourceConfigCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { BatchGetResourceConfigRequest, BatchGetResourceConfigResponse } from "../models/models_0";
import { deserializeAws_json1_1BatchGetResourceConfigCommand, serializeAws_json1_1BatchGetResourceConfigCommand, } from "../protocols/Aws_json1_1";
var BatchGetResourceConfigCommand = (function (_super) {
    __extends(BatchGetResourceConfigCommand, _super);
    function BatchGetResourceConfigCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    BatchGetResourceConfigCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "BatchGetResourceConfigCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: BatchGetResourceConfigRequest.filterSensitiveLog,
            outputFilterSensitiveLog: BatchGetResourceConfigResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    BatchGetResourceConfigCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1BatchGetResourceConfigCommand(input, context);
    };
    BatchGetResourceConfigCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1BatchGetResourceConfigCommand(output, context);
    };
    return BatchGetResourceConfigCommand;
}($Command));
export { BatchGetResourceConfigCommand };
