// @aws-sdk/client-config-service/dist-es/commands/DeleteAggregationAuthorizationCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteAggregationAuthorizationRequest } from "../models/models_0";
import { deserializeAws_json1_1DeleteAggregationAuthorizationCommand, serializeAws_json1_1DeleteAggregationAuthorizationCommand, } from "../protocols/Aws_json1_1";
var DeleteAggregationAuthorizationCommand = (function (_super) {
    __extends(DeleteAggregationAuthorizationCommand, _super);
    function DeleteAggregationAuthorizationCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteAggregationAuthorizationCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DeleteAggregationAuthorizationCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteAggregationAuthorizationRequest.filterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteAggregationAuthorizationCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DeleteAggregationAuthorizationCommand(input, context);
    };
    DeleteAggregationAuthorizationCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DeleteAggregationAuthorizationCommand(output, context);
    };
    return DeleteAggregationAuthorizationCommand;
}($Command));
export { DeleteAggregationAuthorizationCommand };
