// @aws-sdk/client-config-service/dist-es/commands/DeleteConformancePackCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteConformancePackRequest } from "../models/models_0";
import { deserializeAws_json1_1DeleteConformancePackCommand, serializeAws_json1_1DeleteConformancePackCommand, } from "../protocols/Aws_json1_1";
var DeleteConformancePackCommand = (function (_super) {
    __extends(DeleteConformancePackCommand, _super);
    function DeleteConformancePackCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteConformancePackCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DeleteConformancePackCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteConformancePackRequest.filterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteConformancePackCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DeleteConformancePackCommand(input, context);
    };
    DeleteConformancePackCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DeleteConformancePackCommand(output, context);
    };
    return DeleteConformancePackCommand;
}($Command));
export { DeleteConformancePackCommand };
