// @aws-sdk/client-config-service/dist-es/commands/DeleteRetentionConfigurationCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteRetentionConfigurationRequest } from "../models/models_0";
import { deserializeAws_json1_1DeleteRetentionConfigurationCommand, serializeAws_json1_1DeleteRetentionConfigurationCommand, } from "../protocols/Aws_json1_1";
var DeleteRetentionConfigurationCommand = (function (_super) {
    __extends(DeleteRetentionConfigurationCommand, _super);
    function DeleteRetentionConfigurationCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteRetentionConfigurationCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DeleteRetentionConfigurationCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteRetentionConfigurationRequest.filterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteRetentionConfigurationCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DeleteRetentionConfigurationCommand(input, context);
    };
    DeleteRetentionConfigurationCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DeleteRetentionConfigurationCommand(output, context);
    };
    return DeleteRetentionConfigurationCommand;
}($Command));
export { DeleteRetentionConfigurationCommand };
