// @aws-sdk/client-config-service/dist-es/commands/DeleteResourceConfigCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteResourceConfigRequestFilterSensitiveLog } from "../models/models_0";
import { deserializeAws_json1_1DeleteResourceConfigCommand, serializeAws_json1_1DeleteResourceConfigCommand, } from "../protocols/Aws_json1_1";
var DeleteResourceConfigCommand = (function (_super) {
    __extends(DeleteResourceConfigCommand, _super);
    function DeleteResourceConfigCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteResourceConfigCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DeleteResourceConfigCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteResourceConfigRequestFilterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteResourceConfigCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DeleteResourceConfigCommand(input, context);
    };
    DeleteResourceConfigCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DeleteResourceConfigCommand(output, context);
    };
    return DeleteResourceConfigCommand;
}($Command));
export { DeleteResourceConfigCommand };
