// @aws-sdk/client-config-service/dist-es/commands/DeletePendingAggregationRequestCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeletePendingAggregationRequestRequestFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1DeletePendingAggregationRequestCommand, serializeAws_json1_1DeletePendingAggregationRequestCommand, } from "../protocols/Aws_json1_1";
var DeletePendingAggregationRequestCommand = (function (_super) {
    __extends(DeletePendingAggregationRequestCommand, _super);
    function DeletePendingAggregationRequestCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeletePendingAggregationRequestCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DeletePendingAggregationRequestCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeletePendingAggregationRequestRequestFilterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeletePendingAggregationRequestCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DeletePendingAggregationRequestCommand(input, context);
    };
    DeletePendingAggregationRequestCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DeletePendingAggregationRequestCommand(output, context);
    };
    return DeletePendingAggregationRequestCommand;
}($Command));
export { DeletePendingAggregationRequestCommand };
