// @aws-sdk/client-config-service/dist-es/commands/DeliverConfigSnapshotCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeliverConfigSnapshotRequestFilterSensitiveLog, DeliverConfigSnapshotResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1DeliverConfigSnapshotCommand, serializeAws_json1_1DeliverConfigSnapshotCommand, } from "../protocols/Aws_json1_1";
var DeliverConfigSnapshotCommand = (function (_super) {
    __extends(DeliverConfigSnapshotCommand, _super);
    function DeliverConfigSnapshotCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeliverConfigSnapshotCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DeliverConfigSnapshotCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeliverConfigSnapshotRequestFilterSensitiveLog,
            outputFilterSensitiveLog: DeliverConfigSnapshotResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeliverConfigSnapshotCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DeliverConfigSnapshotCommand(input, context);
    };
    DeliverConfigSnapshotCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DeliverConfigSnapshotCommand(output, context);
    };
    return DeliverConfigSnapshotCommand;
}($Command));
export { DeliverConfigSnapshotCommand };
