// @aws-sdk/client-config-service/dist-es/commands/DescribeOrganizationConformancePackStatusesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeOrganizationConformancePackStatusesRequestFilterSensitiveLog, DescribeOrganizationConformancePackStatusesResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1DescribeOrganizationConformancePackStatusesCommand, serializeAws_json1_1DescribeOrganizationConformancePackStatusesCommand, } from "../protocols/Aws_json1_1";
var DescribeOrganizationConformancePackStatusesCommand = (function (_super) {
    __extends(DescribeOrganizationConformancePackStatusesCommand, _super);
    function DescribeOrganizationConformancePackStatusesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeOrganizationConformancePackStatusesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DescribeOrganizationConformancePackStatusesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeOrganizationConformancePackStatusesRequestFilterSensitiveLog,
            outputFilterSensitiveLog: DescribeOrganizationConformancePackStatusesResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeOrganizationConformancePackStatusesCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeOrganizationConformancePackStatusesCommand(input, context);
    };
    DescribeOrganizationConformancePackStatusesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeOrganizationConformancePackStatusesCommand(output, context);
    };
    return DescribeOrganizationConformancePackStatusesCommand;
}($Command));
export { DescribeOrganizationConformancePackStatusesCommand };
