// @aws-sdk/client-config-service/dist-es/commands/DescribeRemediationExecutionStatusCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeRemediationExecutionStatusRequestFilterSensitiveLog, DescribeRemediationExecutionStatusResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1DescribeRemediationExecutionStatusCommand, serializeAws_json1_1DescribeRemediationExecutionStatusCommand, } from "../protocols/Aws_json1_1";
var DescribeRemediationExecutionStatusCommand = (function (_super) {
    __extends(DescribeRemediationExecutionStatusCommand, _super);
    function DescribeRemediationExecutionStatusCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeRemediationExecutionStatusCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DescribeRemediationExecutionStatusCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeRemediationExecutionStatusRequestFilterSensitiveLog,
            outputFilterSensitiveLog: DescribeRemediationExecutionStatusResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeRemediationExecutionStatusCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeRemediationExecutionStatusCommand(input, context);
    };
    DescribeRemediationExecutionStatusCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeRemediationExecutionStatusCommand(output, context);
    };
    return DescribeRemediationExecutionStatusCommand;
}($Command));
export { DescribeRemediationExecutionStatusCommand };
