// @aws-sdk/client-config-service/dist-es/commands/DescribeConfigurationRecordersCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeConfigurationRecordersRequest, DescribeConfigurationRecordersResponse } from "../models/models_0";
import { deserializeAws_json1_1DescribeConfigurationRecordersCommand, serializeAws_json1_1DescribeConfigurationRecordersCommand, } from "../protocols/Aws_json1_1";
var DescribeConfigurationRecordersCommand = (function (_super) {
    __extends(DescribeConfigurationRecordersCommand, _super);
    function DescribeConfigurationRecordersCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeConfigurationRecordersCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DescribeConfigurationRecordersCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeConfigurationRecordersRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DescribeConfigurationRecordersResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeConfigurationRecordersCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeConfigurationRecordersCommand(input, context);
    };
    DescribeConfigurationRecordersCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeConfigurationRecordersCommand(output, context);
    };
    return DescribeConfigurationRecordersCommand;
}($Command));
export { DescribeConfigurationRecordersCommand };
