// @aws-sdk/client-config-service/dist-es/commands/DescribeOrganizationConfigRuleStatusesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeOrganizationConfigRuleStatusesRequest, DescribeOrganizationConfigRuleStatusesResponse, } from "../models/models_0";
import { deserializeAws_json1_1DescribeOrganizationConfigRuleStatusesCommand, serializeAws_json1_1DescribeOrganizationConfigRuleStatusesCommand, } from "../protocols/Aws_json1_1";
var DescribeOrganizationConfigRuleStatusesCommand = (function (_super) {
    __extends(DescribeOrganizationConfigRuleStatusesCommand, _super);
    function DescribeOrganizationConfigRuleStatusesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeOrganizationConfigRuleStatusesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DescribeOrganizationConfigRuleStatusesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeOrganizationConfigRuleStatusesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DescribeOrganizationConfigRuleStatusesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeOrganizationConfigRuleStatusesCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeOrganizationConfigRuleStatusesCommand(input, context);
    };
    DescribeOrganizationConfigRuleStatusesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeOrganizationConfigRuleStatusesCommand(output, context);
    };
    return DescribeOrganizationConfigRuleStatusesCommand;
}($Command));
export { DescribeOrganizationConfigRuleStatusesCommand };
