// @aws-sdk/client-config-service/dist-es/commands/DescribeConfigRuleEvaluationStatusCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeConfigRuleEvaluationStatusRequestFilterSensitiveLog, DescribeConfigRuleEvaluationStatusResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1DescribeConfigRuleEvaluationStatusCommand, serializeAws_json1_1DescribeConfigRuleEvaluationStatusCommand, } from "../protocols/Aws_json1_1";
var DescribeConfigRuleEvaluationStatusCommand = (function (_super) {
    __extends(DescribeConfigRuleEvaluationStatusCommand, _super);
    function DescribeConfigRuleEvaluationStatusCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeConfigRuleEvaluationStatusCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DescribeConfigRuleEvaluationStatusCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeConfigRuleEvaluationStatusRequestFilterSensitiveLog,
            outputFilterSensitiveLog: DescribeConfigRuleEvaluationStatusResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeConfigRuleEvaluationStatusCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeConfigRuleEvaluationStatusCommand(input, context);
    };
    DescribeConfigRuleEvaluationStatusCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeConfigRuleEvaluationStatusCommand(output, context);
    };
    return DescribeConfigRuleEvaluationStatusCommand;
}($Command));
export { DescribeConfigRuleEvaluationStatusCommand };
