// @aws-sdk/client-config-service/dist-es/commands/DescribeComplianceByConfigRuleCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeComplianceByConfigRuleRequestFilterSensitiveLog, DescribeComplianceByConfigRuleResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1DescribeComplianceByConfigRuleCommand, serializeAws_json1_1DescribeComplianceByConfigRuleCommand, } from "../protocols/Aws_json1_1";
var DescribeComplianceByConfigRuleCommand = (function (_super) {
    __extends(DescribeComplianceByConfigRuleCommand, _super);
    function DescribeComplianceByConfigRuleCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeComplianceByConfigRuleCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DescribeComplianceByConfigRuleCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeComplianceByConfigRuleRequestFilterSensitiveLog,
            outputFilterSensitiveLog: DescribeComplianceByConfigRuleResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeComplianceByConfigRuleCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeComplianceByConfigRuleCommand(input, context);
    };
    DescribeComplianceByConfigRuleCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeComplianceByConfigRuleCommand(output, context);
    };
    return DescribeComplianceByConfigRuleCommand;
}($Command));
export { DescribeComplianceByConfigRuleCommand };
