// @aws-sdk/client-config-service/dist-es/commands/DescribeConformancePackStatusCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeConformancePackStatusRequest, DescribeConformancePackStatusResponse } from "../models/models_0";
import { deserializeAws_json1_1DescribeConformancePackStatusCommand, serializeAws_json1_1DescribeConformancePackStatusCommand, } from "../protocols/Aws_json1_1";
var DescribeConformancePackStatusCommand = (function (_super) {
    __extends(DescribeConformancePackStatusCommand, _super);
    function DescribeConformancePackStatusCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeConformancePackStatusCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DescribeConformancePackStatusCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeConformancePackStatusRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DescribeConformancePackStatusResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeConformancePackStatusCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeConformancePackStatusCommand(input, context);
    };
    DescribeConformancePackStatusCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeConformancePackStatusCommand(output, context);
    };
    return DescribeConformancePackStatusCommand;
}($Command));
export { DescribeConformancePackStatusCommand };
