// @aws-sdk/client-config-service/dist-es/commands/DescribeOrganizationConformancePacksCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeOrganizationConformancePacksRequest, DescribeOrganizationConformancePacksResponse, } from "../models/models_0";
import { deserializeAws_json1_1DescribeOrganizationConformancePacksCommand, serializeAws_json1_1DescribeOrganizationConformancePacksCommand, } from "../protocols/Aws_json1_1";
var DescribeOrganizationConformancePacksCommand = (function (_super) {
    __extends(DescribeOrganizationConformancePacksCommand, _super);
    function DescribeOrganizationConformancePacksCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeOrganizationConformancePacksCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "DescribeOrganizationConformancePacksCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeOrganizationConformancePacksRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DescribeOrganizationConformancePacksResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeOrganizationConformancePacksCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeOrganizationConformancePacksCommand(input, context);
    };
    DescribeOrganizationConformancePacksCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeOrganizationConformancePacksCommand(output, context);
    };
    return DescribeOrganizationConformancePacksCommand;
}($Command));
export { DescribeOrganizationConformancePacksCommand };
