// @aws-sdk/client-config-service/dist-es/commands/GetAggregateComplianceDetailsByConfigRuleCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetAggregateComplianceDetailsByConfigRuleRequestFilterSensitiveLog, GetAggregateComplianceDetailsByConfigRuleResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1GetAggregateComplianceDetailsByConfigRuleCommand, serializeAws_json1_1GetAggregateComplianceDetailsByConfigRuleCommand, } from "../protocols/Aws_json1_1";
var GetAggregateComplianceDetailsByConfigRuleCommand = (function (_super) {
    __extends(GetAggregateComplianceDetailsByConfigRuleCommand, _super);
    function GetAggregateComplianceDetailsByConfigRuleCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetAggregateComplianceDetailsByConfigRuleCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "GetAggregateComplianceDetailsByConfigRuleCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetAggregateComplianceDetailsByConfigRuleRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetAggregateComplianceDetailsByConfigRuleResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetAggregateComplianceDetailsByConfigRuleCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1GetAggregateComplianceDetailsByConfigRuleCommand(input, context);
    };
    GetAggregateComplianceDetailsByConfigRuleCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1GetAggregateComplianceDetailsByConfigRuleCommand(output, context);
    };
    return GetAggregateComplianceDetailsByConfigRuleCommand;
}($Command));
export { GetAggregateComplianceDetailsByConfigRuleCommand };
