// @aws-sdk/client-config-service/dist-es/commands/GetAggregateDiscoveredResourceCountsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetAggregateDiscoveredResourceCountsRequestFilterSensitiveLog, GetAggregateDiscoveredResourceCountsResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1GetAggregateDiscoveredResourceCountsCommand, serializeAws_json1_1GetAggregateDiscoveredResourceCountsCommand, } from "../protocols/Aws_json1_1";
var GetAggregateDiscoveredResourceCountsCommand = (function (_super) {
    __extends(GetAggregateDiscoveredResourceCountsCommand, _super);
    function GetAggregateDiscoveredResourceCountsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetAggregateDiscoveredResourceCountsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "GetAggregateDiscoveredResourceCountsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetAggregateDiscoveredResourceCountsRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetAggregateDiscoveredResourceCountsResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetAggregateDiscoveredResourceCountsCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1GetAggregateDiscoveredResourceCountsCommand(input, context);
    };
    GetAggregateDiscoveredResourceCountsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1GetAggregateDiscoveredResourceCountsCommand(output, context);
    };
    return GetAggregateDiscoveredResourceCountsCommand;
}($Command));
export { GetAggregateDiscoveredResourceCountsCommand };
