// @aws-sdk/client-config-service/dist-es/commands/GetCustomRulePolicyCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetCustomRulePolicyRequestFilterSensitiveLog, GetCustomRulePolicyResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1GetCustomRulePolicyCommand, serializeAws_json1_1GetCustomRulePolicyCommand, } from "../protocols/Aws_json1_1";
var GetCustomRulePolicyCommand = (function (_super) {
    __extends(GetCustomRulePolicyCommand, _super);
    function GetCustomRulePolicyCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetCustomRulePolicyCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "GetCustomRulePolicyCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetCustomRulePolicyRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetCustomRulePolicyResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetCustomRulePolicyCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1GetCustomRulePolicyCommand(input, context);
    };
    GetCustomRulePolicyCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1GetCustomRulePolicyCommand(output, context);
    };
    return GetCustomRulePolicyCommand;
}($Command));
export { GetCustomRulePolicyCommand };
