// @aws-sdk/client-config-service/dist-es/commands/GetComplianceDetailsByConfigRuleCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetComplianceDetailsByConfigRuleRequestFilterSensitiveLog, GetComplianceDetailsByConfigRuleResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1GetComplianceDetailsByConfigRuleCommand, serializeAws_json1_1GetComplianceDetailsByConfigRuleCommand, } from "../protocols/Aws_json1_1";
var GetComplianceDetailsByConfigRuleCommand = (function (_super) {
    __extends(GetComplianceDetailsByConfigRuleCommand, _super);
    function GetComplianceDetailsByConfigRuleCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetComplianceDetailsByConfigRuleCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "GetComplianceDetailsByConfigRuleCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetComplianceDetailsByConfigRuleRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetComplianceDetailsByConfigRuleResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetComplianceDetailsByConfigRuleCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1GetComplianceDetailsByConfigRuleCommand(input, context);
    };
    GetComplianceDetailsByConfigRuleCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1GetComplianceDetailsByConfigRuleCommand(output, context);
    };
    return GetComplianceDetailsByConfigRuleCommand;
}($Command));
export { GetComplianceDetailsByConfigRuleCommand };
