// @aws-sdk/client-config-service/dist-es/commands/GetOrganizationConfigRuleDetailedStatusCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetOrganizationConfigRuleDetailedStatusRequestFilterSensitiveLog, GetOrganizationConfigRuleDetailedStatusResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1GetOrganizationConfigRuleDetailedStatusCommand, serializeAws_json1_1GetOrganizationConfigRuleDetailedStatusCommand, } from "../protocols/Aws_json1_1";
var GetOrganizationConfigRuleDetailedStatusCommand = (function (_super) {
    __extends(GetOrganizationConfigRuleDetailedStatusCommand, _super);
    function GetOrganizationConfigRuleDetailedStatusCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetOrganizationConfigRuleDetailedStatusCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "GetOrganizationConfigRuleDetailedStatusCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetOrganizationConfigRuleDetailedStatusRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetOrganizationConfigRuleDetailedStatusResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetOrganizationConfigRuleDetailedStatusCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1GetOrganizationConfigRuleDetailedStatusCommand(input, context);
    };
    GetOrganizationConfigRuleDetailedStatusCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1GetOrganizationConfigRuleDetailedStatusCommand(output, context);
    };
    return GetOrganizationConfigRuleDetailedStatusCommand;
}($Command));
export { GetOrganizationConfigRuleDetailedStatusCommand };
