// @aws-sdk/client-config-service/dist-es/commands/GetOrganizationConformancePackDetailedStatusCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetOrganizationConformancePackDetailedStatusRequest, GetOrganizationConformancePackDetailedStatusResponse, } from "../models/models_0";
import { deserializeAws_json1_1GetOrganizationConformancePackDetailedStatusCommand, serializeAws_json1_1GetOrganizationConformancePackDetailedStatusCommand, } from "../protocols/Aws_json1_1";
var GetOrganizationConformancePackDetailedStatusCommand = (function (_super) {
    __extends(GetOrganizationConformancePackDetailedStatusCommand, _super);
    function GetOrganizationConformancePackDetailedStatusCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetOrganizationConformancePackDetailedStatusCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "GetOrganizationConformancePackDetailedStatusCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetOrganizationConformancePackDetailedStatusRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetOrganizationConformancePackDetailedStatusResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetOrganizationConformancePackDetailedStatusCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1GetOrganizationConformancePackDetailedStatusCommand(input, context);
    };
    GetOrganizationConformancePackDetailedStatusCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1GetOrganizationConformancePackDetailedStatusCommand(output, context);
    };
    return GetOrganizationConformancePackDetailedStatusCommand;
}($Command));
export { GetOrganizationConformancePackDetailedStatusCommand };
