// @aws-sdk/client-config-service/dist-es/commands/GetResourceConfigHistoryCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetResourceConfigHistoryRequestFilterSensitiveLog, GetResourceConfigHistoryResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1GetResourceConfigHistoryCommand, serializeAws_json1_1GetResourceConfigHistoryCommand, } from "../protocols/Aws_json1_1";
var GetResourceConfigHistoryCommand = (function (_super) {
    __extends(GetResourceConfigHistoryCommand, _super);
    function GetResourceConfigHistoryCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetResourceConfigHistoryCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "GetResourceConfigHistoryCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetResourceConfigHistoryRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetResourceConfigHistoryResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetResourceConfigHistoryCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1GetResourceConfigHistoryCommand(input, context);
    };
    GetResourceConfigHistoryCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1GetResourceConfigHistoryCommand(output, context);
    };
    return GetResourceConfigHistoryCommand;
}($Command));
export { GetResourceConfigHistoryCommand };
