// @aws-sdk/client-config-service/dist-es/commands/GetStoredQueryCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetStoredQueryRequest, GetStoredQueryResponse } from "../models/models_0";
import { deserializeAws_json1_1GetStoredQueryCommand, serializeAws_json1_1GetStoredQueryCommand, } from "../protocols/Aws_json1_1";
var GetStoredQueryCommand = (function (_super) {
    __extends(GetStoredQueryCommand, _super);
    function GetStoredQueryCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetStoredQueryCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "GetStoredQueryCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetStoredQueryRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetStoredQueryResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetStoredQueryCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1GetStoredQueryCommand(input, context);
    };
    GetStoredQueryCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1GetStoredQueryCommand(output, context);
    };
    return GetStoredQueryCommand;
}($Command));
export { GetStoredQueryCommand };
