// @aws-sdk/client-config-service/dist-es/commands/ListDiscoveredResourcesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListDiscoveredResourcesRequestFilterSensitiveLog, ListDiscoveredResourcesResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1ListDiscoveredResourcesCommand, serializeAws_json1_1ListDiscoveredResourcesCommand, } from "../protocols/Aws_json1_1";
var ListDiscoveredResourcesCommand = (function (_super) {
    __extends(ListDiscoveredResourcesCommand, _super);
    function ListDiscoveredResourcesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListDiscoveredResourcesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "ListDiscoveredResourcesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListDiscoveredResourcesRequestFilterSensitiveLog,
            outputFilterSensitiveLog: ListDiscoveredResourcesResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListDiscoveredResourcesCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1ListDiscoveredResourcesCommand(input, context);
    };
    ListDiscoveredResourcesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1ListDiscoveredResourcesCommand(output, context);
    };
    return ListDiscoveredResourcesCommand;
}($Command));
export { ListDiscoveredResourcesCommand };
