// @aws-sdk/client-config-service/dist-es/commands/ListStoredQueriesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListStoredQueriesRequest, ListStoredQueriesResponse } from "../models/models_0";
import { deserializeAws_json1_1ListStoredQueriesCommand, serializeAws_json1_1ListStoredQueriesCommand, } from "../protocols/Aws_json1_1";
var ListStoredQueriesCommand = (function (_super) {
    __extends(ListStoredQueriesCommand, _super);
    function ListStoredQueriesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListStoredQueriesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "ListStoredQueriesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListStoredQueriesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: ListStoredQueriesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListStoredQueriesCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1ListStoredQueriesCommand(input, context);
    };
    ListStoredQueriesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1ListStoredQueriesCommand(output, context);
    };
    return ListStoredQueriesCommand;
}($Command));
export { ListStoredQueriesCommand };
