// @aws-sdk/client-config-service/dist-es/commands/PutAggregationAuthorizationCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutAggregationAuthorizationRequest, PutAggregationAuthorizationResponse } from "../models/models_0";
import { deserializeAws_json1_1PutAggregationAuthorizationCommand, serializeAws_json1_1PutAggregationAuthorizationCommand, } from "../protocols/Aws_json1_1";
var PutAggregationAuthorizationCommand = (function (_super) {
    __extends(PutAggregationAuthorizationCommand, _super);
    function PutAggregationAuthorizationCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutAggregationAuthorizationCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "PutAggregationAuthorizationCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutAggregationAuthorizationRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutAggregationAuthorizationResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutAggregationAuthorizationCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1PutAggregationAuthorizationCommand(input, context);
    };
    PutAggregationAuthorizationCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1PutAggregationAuthorizationCommand(output, context);
    };
    return PutAggregationAuthorizationCommand;
}($Command));
export { PutAggregationAuthorizationCommand };
