// @aws-sdk/client-config-service/dist-es/commands/PutEvaluationsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutEvaluationsRequestFilterSensitiveLog, PutEvaluationsResponseFilterSensitiveLog, } from "../models/models_1";
import { deserializeAws_json1_1PutEvaluationsCommand, serializeAws_json1_1PutEvaluationsCommand, } from "../protocols/Aws_json1_1";
var PutEvaluationsCommand = (function (_super) {
    __extends(PutEvaluationsCommand, _super);
    function PutEvaluationsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutEvaluationsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "PutEvaluationsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutEvaluationsRequestFilterSensitiveLog,
            outputFilterSensitiveLog: PutEvaluationsResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutEvaluationsCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1PutEvaluationsCommand(input, context);
    };
    PutEvaluationsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1PutEvaluationsCommand(output, context);
    };
    return PutEvaluationsCommand;
}($Command));
export { PutEvaluationsCommand };
