// @aws-sdk/client-config-service/dist-es/commands/PutOrganizationConfigRuleCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutOrganizationConfigRuleRequest, PutOrganizationConfigRuleResponse } from "../models/models_0";
import { deserializeAws_json1_1PutOrganizationConfigRuleCommand, serializeAws_json1_1PutOrganizationConfigRuleCommand, } from "../protocols/Aws_json1_1";
var PutOrganizationConfigRuleCommand = (function (_super) {
    __extends(PutOrganizationConfigRuleCommand, _super);
    function PutOrganizationConfigRuleCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutOrganizationConfigRuleCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "PutOrganizationConfigRuleCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutOrganizationConfigRuleRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutOrganizationConfigRuleResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutOrganizationConfigRuleCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1PutOrganizationConfigRuleCommand(input, context);
    };
    PutOrganizationConfigRuleCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1PutOrganizationConfigRuleCommand(output, context);
    };
    return PutOrganizationConfigRuleCommand;
}($Command));
export { PutOrganizationConfigRuleCommand };
