// @aws-sdk/client-config-service/dist-es/commands/PutRemediationConfigurationsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutRemediationConfigurationsRequest, PutRemediationConfigurationsResponse } from "../models/models_1";
import { deserializeAws_json1_1PutRemediationConfigurationsCommand, serializeAws_json1_1PutRemediationConfigurationsCommand, } from "../protocols/Aws_json1_1";
var PutRemediationConfigurationsCommand = (function (_super) {
    __extends(PutRemediationConfigurationsCommand, _super);
    function PutRemediationConfigurationsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutRemediationConfigurationsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "PutRemediationConfigurationsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutRemediationConfigurationsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutRemediationConfigurationsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutRemediationConfigurationsCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1PutRemediationConfigurationsCommand(input, context);
    };
    PutRemediationConfigurationsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1PutRemediationConfigurationsCommand(output, context);
    };
    return PutRemediationConfigurationsCommand;
}($Command));
export { PutRemediationConfigurationsCommand };
