// @aws-sdk/client-config-service/dist-es/commands/PutStoredQueryCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutStoredQueryRequest, PutStoredQueryResponse } from "../models/models_1";
import { deserializeAws_json1_1PutStoredQueryCommand, serializeAws_json1_1PutStoredQueryCommand, } from "../protocols/Aws_json1_1";
var PutStoredQueryCommand = (function (_super) {
    __extends(PutStoredQueryCommand, _super);
    function PutStoredQueryCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutStoredQueryCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "PutStoredQueryCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutStoredQueryRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutStoredQueryResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutStoredQueryCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1PutStoredQueryCommand(input, context);
    };
    PutStoredQueryCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1PutStoredQueryCommand(output, context);
    };
    return PutStoredQueryCommand;
}($Command));
export { PutStoredQueryCommand };
