// @aws-sdk/client-config-service/dist-es/commands/StartRemediationExecutionCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { StartRemediationExecutionRequestFilterSensitiveLog, StartRemediationExecutionResponseFilterSensitiveLog, } from "../models/models_1";
import { deserializeAws_json1_1StartRemediationExecutionCommand, serializeAws_json1_1StartRemediationExecutionCommand, } from "../protocols/Aws_json1_1";
var StartRemediationExecutionCommand = (function (_super) {
    __extends(StartRemediationExecutionCommand, _super);
    function StartRemediationExecutionCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    StartRemediationExecutionCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ConfigServiceClient";
        var commandName = "StartRemediationExecutionCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: StartRemediationExecutionRequestFilterSensitiveLog,
            outputFilterSensitiveLog: StartRemediationExecutionResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    StartRemediationExecutionCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1StartRemediationExecutionCommand(input, context);
    };
    StartRemediationExecutionCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1StartRemediationExecutionCommand(output, context);
    };
    return StartRemediationExecutionCommand;
}($Command));
export { StartRemediationExecutionCommand };
