// oclif/lib/commands/upload/win.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const fs = __importStar(require("node:fs"));
const aws_1 = __importDefault(require("../../aws"));
const log_1 = require("../../log");
const Tarballs = __importStar(require("../../tarballs"));
const upload_util_1 = require("../../upload-util");
class UploadWin extends core_1.Command {
    static description = 'Upload windows installers built with `pack win`.';
    static flags = {
        root: core_1.Flags.string({ char: 'r', default: '.', description: 'Path to oclif CLI root.', required: true }),
        targets: core_1.Flags.string({ description: 'Comma-separated targets to pack (e.g.: win32-x64,win32-x86).' }),
    };
    async run() {
        const { flags } = await this.parse(UploadWin);
        const buildConfig = await Tarballs.buildConfig(flags.root, { targets: flags?.targets?.split(',') });
        const { config, dist, s3Config } = buildConfig;
        const S3Options = {
            ACL: s3Config.acl || 'public-read',
            Bucket: s3Config.bucket,
        };
        const archs = buildConfig.targets.filter((t) => t.platform === 'win32').map((t) => t.arch);
        for (const arch of archs) {
            const templateKey = (0, upload_util_1.templateShortKey)('win32', {
                arch,
                bin: config.bin,
                sha: buildConfig.gitSha,
                version: config.version,
            });
            const localKey = dist(`win32/${templateKey}`);
            if (!fs.existsSync(localKey))
                this.error(`Cannot find Windows exe for ${arch}`, {
                    suggestions: ['Run "oclif pack win" before uploading'],
                });
        }
        const cloudKeyBase = (0, upload_util_1.commitAWSDir)(config.pjson.version, buildConfig.gitSha, s3Config);
        const uploadWin = async (arch) => {
            const templateKey = (0, upload_util_1.templateShortKey)('win32', {
                arch,
                bin: config.bin,
                sha: buildConfig.gitSha,
                version: config.version,
            });
            const localExe = dist(`win32/${templateKey}`);
            const cloudKey = `${cloudKeyBase}/${templateKey}`;
            if (fs.existsSync(localExe))
                await aws_1.default.s3.uploadFile(localExe, { ...S3Options, CacheControl: 'max-age=86400', Key: cloudKey });
        };
        await Promise.all([uploadWin('x64'), uploadWin('x86')]);
        (0, log_1.log)(`done uploading windows executables for v${config.version}-${buildConfig.gitSha}`);
    }
}
exports.default = UploadWin;
