// inspecjs/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertImpactToSeverity = exports.titleCasedSeverities = exports.severities = exports.lowercasedControlStatuses = exports.hdfWrapControl = exports.controlStatuses = void 0;
var compat_wrappers_1 = require("./compat_wrappers");
Object.defineProperty(exports, "controlStatuses", { enumerable: true, get: function () { return compat_wrappers_1.controlStatuses; } });
Object.defineProperty(exports, "hdfWrapControl", { enumerable: true, get: function () { return compat_wrappers_1.hdfWrapControl; } });
Object.defineProperty(exports, "lowercasedControlStatuses", { enumerable: true, get: function () { return compat_wrappers_1.lowercasedControlStatuses; } });
Object.defineProperty(exports, "severities", { enumerable: true, get: function () { return compat_wrappers_1.severities; } });
Object.defineProperty(exports, "titleCasedSeverities", { enumerable: true, get: function () { return compat_wrappers_1.titleCasedSeverities; } });
Object.defineProperty(exports, "convertImpactToSeverity", { enumerable: true, get: function () { return compat_wrappers_1.convertImpactToSeverity; } });
__exportStar(require("./context"), exports);
__exportStar(require("./fileparse"), exports);
__exportStar(require("./nist"), exports);
__exportStar(require("./versions/v_1_0"), exports);
