// @aws-sdk/client-securityhub/dist-cjs/SecurityHub.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityHub = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const AcceptAdministratorInvitationCommand_1 = require("./commands/AcceptAdministratorInvitationCommand");
const AcceptInvitationCommand_1 = require("./commands/AcceptInvitationCommand");
const BatchDeleteAutomationRulesCommand_1 = require("./commands/BatchDeleteAutomationRulesCommand");
const BatchDisableStandardsCommand_1 = require("./commands/BatchDisableStandardsCommand");
const BatchEnableStandardsCommand_1 = require("./commands/BatchEnableStandardsCommand");
const BatchGetAutomationRulesCommand_1 = require("./commands/BatchGetAutomationRulesCommand");
const BatchGetConfigurationPolicyAssociationsCommand_1 = require("./commands/BatchGetConfigurationPolicyAssociationsCommand");
const BatchGetSecurityControlsCommand_1 = require("./commands/BatchGetSecurityControlsCommand");
const BatchGetStandardsControlAssociationsCommand_1 = require("./commands/BatchGetStandardsControlAssociationsCommand");
const BatchImportFindingsCommand_1 = require("./commands/BatchImportFindingsCommand");
const BatchUpdateAutomationRulesCommand_1 = require("./commands/BatchUpdateAutomationRulesCommand");
const BatchUpdateFindingsCommand_1 = require("./commands/BatchUpdateFindingsCommand");
const BatchUpdateStandardsControlAssociationsCommand_1 = require("./commands/BatchUpdateStandardsControlAssociationsCommand");
const CreateActionTargetCommand_1 = require("./commands/CreateActionTargetCommand");
const CreateAutomationRuleCommand_1 = require("./commands/CreateAutomationRuleCommand");
const CreateConfigurationPolicyCommand_1 = require("./commands/CreateConfigurationPolicyCommand");
const CreateFindingAggregatorCommand_1 = require("./commands/CreateFindingAggregatorCommand");
const CreateInsightCommand_1 = require("./commands/CreateInsightCommand");
const CreateMembersCommand_1 = require("./commands/CreateMembersCommand");
const DeclineInvitationsCommand_1 = require("./commands/DeclineInvitationsCommand");
const DeleteActionTargetCommand_1 = require("./commands/DeleteActionTargetCommand");
const DeleteConfigurationPolicyCommand_1 = require("./commands/DeleteConfigurationPolicyCommand");
const DeleteFindingAggregatorCommand_1 = require("./commands/DeleteFindingAggregatorCommand");
const DeleteInsightCommand_1 = require("./commands/DeleteInsightCommand");
const DeleteInvitationsCommand_1 = require("./commands/DeleteInvitationsCommand");
const DeleteMembersCommand_1 = require("./commands/DeleteMembersCommand");
const DescribeActionTargetsCommand_1 = require("./commands/DescribeActionTargetsCommand");
const DescribeHubCommand_1 = require("./commands/DescribeHubCommand");
const DescribeOrganizationConfigurationCommand_1 = require("./commands/DescribeOrganizationConfigurationCommand");
const DescribeProductsCommand_1 = require("./commands/DescribeProductsCommand");
const DescribeStandardsCommand_1 = require("./commands/DescribeStandardsCommand");
const DescribeStandardsControlsCommand_1 = require("./commands/DescribeStandardsControlsCommand");
const DisableImportFindingsForProductCommand_1 = require("./commands/DisableImportFindingsForProductCommand");
const DisableOrganizationAdminAccountCommand_1 = require("./commands/DisableOrganizationAdminAccountCommand");
const DisableSecurityHubCommand_1 = require("./commands/DisableSecurityHubCommand");
const DisassociateFromAdministratorAccountCommand_1 = require("./commands/DisassociateFromAdministratorAccountCommand");
const DisassociateFromMasterAccountCommand_1 = require("./commands/DisassociateFromMasterAccountCommand");
const DisassociateMembersCommand_1 = require("./commands/DisassociateMembersCommand");
const EnableImportFindingsForProductCommand_1 = require("./commands/EnableImportFindingsForProductCommand");
const EnableOrganizationAdminAccountCommand_1 = require("./commands/EnableOrganizationAdminAccountCommand");
const EnableSecurityHubCommand_1 = require("./commands/EnableSecurityHubCommand");
const GetAdministratorAccountCommand_1 = require("./commands/GetAdministratorAccountCommand");
const GetConfigurationPolicyAssociationCommand_1 = require("./commands/GetConfigurationPolicyAssociationCommand");
const GetConfigurationPolicyCommand_1 = require("./commands/GetConfigurationPolicyCommand");
const GetEnabledStandardsCommand_1 = require("./commands/GetEnabledStandardsCommand");
const GetFindingAggregatorCommand_1 = require("./commands/GetFindingAggregatorCommand");
const GetFindingHistoryCommand_1 = require("./commands/GetFindingHistoryCommand");
const GetFindingsCommand_1 = require("./commands/GetFindingsCommand");
const GetInsightResultsCommand_1 = require("./commands/GetInsightResultsCommand");
const GetInsightsCommand_1 = require("./commands/GetInsightsCommand");
const GetInvitationsCountCommand_1 = require("./commands/GetInvitationsCountCommand");
const GetMasterAccountCommand_1 = require("./commands/GetMasterAccountCommand");
const GetMembersCommand_1 = require("./commands/GetMembersCommand");
const GetSecurityControlDefinitionCommand_1 = require("./commands/GetSecurityControlDefinitionCommand");
const InviteMembersCommand_1 = require("./commands/InviteMembersCommand");
const ListAutomationRulesCommand_1 = require("./commands/ListAutomationRulesCommand");
const ListConfigurationPoliciesCommand_1 = require("./commands/ListConfigurationPoliciesCommand");
const ListConfigurationPolicyAssociationsCommand_1 = require("./commands/ListConfigurationPolicyAssociationsCommand");
const ListEnabledProductsForImportCommand_1 = require("./commands/ListEnabledProductsForImportCommand");
const ListFindingAggregatorsCommand_1 = require("./commands/ListFindingAggregatorsCommand");
const ListInvitationsCommand_1 = require("./commands/ListInvitationsCommand");
const ListMembersCommand_1 = require("./commands/ListMembersCommand");
const ListOrganizationAdminAccountsCommand_1 = require("./commands/ListOrganizationAdminAccountsCommand");
const ListSecurityControlDefinitionsCommand_1 = require("./commands/ListSecurityControlDefinitionsCommand");
const ListStandardsControlAssociationsCommand_1 = require("./commands/ListStandardsControlAssociationsCommand");
const ListTagsForResourceCommand_1 = require("./commands/ListTagsForResourceCommand");
const StartConfigurationPolicyAssociationCommand_1 = require("./commands/StartConfigurationPolicyAssociationCommand");
const StartConfigurationPolicyDisassociationCommand_1 = require("./commands/StartConfigurationPolicyDisassociationCommand");
const TagResourceCommand_1 = require("./commands/TagResourceCommand");
const UntagResourceCommand_1 = require("./commands/UntagResourceCommand");
const UpdateActionTargetCommand_1 = require("./commands/UpdateActionTargetCommand");
const UpdateConfigurationPolicyCommand_1 = require("./commands/UpdateConfigurationPolicyCommand");
const UpdateFindingAggregatorCommand_1 = require("./commands/UpdateFindingAggregatorCommand");
const UpdateFindingsCommand_1 = require("./commands/UpdateFindingsCommand");
const UpdateInsightCommand_1 = require("./commands/UpdateInsightCommand");
const UpdateOrganizationConfigurationCommand_1 = require("./commands/UpdateOrganizationConfigurationCommand");
const UpdateSecurityControlCommand_1 = require("./commands/UpdateSecurityControlCommand");
const UpdateSecurityHubConfigurationCommand_1 = require("./commands/UpdateSecurityHubConfigurationCommand");
const UpdateStandardsControlCommand_1 = require("./commands/UpdateStandardsControlCommand");
const SecurityHubClient_1 = require("./SecurityHubClient");
const commands = {
    AcceptAdministratorInvitationCommand: AcceptAdministratorInvitationCommand_1.AcceptAdministratorInvitationCommand,
    AcceptInvitationCommand: AcceptInvitationCommand_1.AcceptInvitationCommand,
    BatchDeleteAutomationRulesCommand: BatchDeleteAutomationRulesCommand_1.BatchDeleteAutomationRulesCommand,
    BatchDisableStandardsCommand: BatchDisableStandardsCommand_1.BatchDisableStandardsCommand,
    BatchEnableStandardsCommand: BatchEnableStandardsCommand_1.BatchEnableStandardsCommand,
    BatchGetAutomationRulesCommand: BatchGetAutomationRulesCommand_1.BatchGetAutomationRulesCommand,
    BatchGetConfigurationPolicyAssociationsCommand: BatchGetConfigurationPolicyAssociationsCommand_1.BatchGetConfigurationPolicyAssociationsCommand,
    BatchGetSecurityControlsCommand: BatchGetSecurityControlsCommand_1.BatchGetSecurityControlsCommand,
    BatchGetStandardsControlAssociationsCommand: BatchGetStandardsControlAssociationsCommand_1.BatchGetStandardsControlAssociationsCommand,
    BatchImportFindingsCommand: BatchImportFindingsCommand_1.BatchImportFindingsCommand,
    BatchUpdateAutomationRulesCommand: BatchUpdateAutomationRulesCommand_1.BatchUpdateAutomationRulesCommand,
    BatchUpdateFindingsCommand: BatchUpdateFindingsCommand_1.BatchUpdateFindingsCommand,
    BatchUpdateStandardsControlAssociationsCommand: BatchUpdateStandardsControlAssociationsCommand_1.BatchUpdateStandardsControlAssociationsCommand,
    CreateActionTargetCommand: CreateActionTargetCommand_1.CreateActionTargetCommand,
    CreateAutomationRuleCommand: CreateAutomationRuleCommand_1.CreateAutomationRuleCommand,
    CreateConfigurationPolicyCommand: CreateConfigurationPolicyCommand_1.CreateConfigurationPolicyCommand,
    CreateFindingAggregatorCommand: CreateFindingAggregatorCommand_1.CreateFindingAggregatorCommand,
    CreateInsightCommand: CreateInsightCommand_1.CreateInsightCommand,
    CreateMembersCommand: CreateMembersCommand_1.CreateMembersCommand,
    DeclineInvitationsCommand: DeclineInvitationsCommand_1.DeclineInvitationsCommand,
    DeleteActionTargetCommand: DeleteActionTargetCommand_1.DeleteActionTargetCommand,
    DeleteConfigurationPolicyCommand: DeleteConfigurationPolicyCommand_1.DeleteConfigurationPolicyCommand,
    DeleteFindingAggregatorCommand: DeleteFindingAggregatorCommand_1.DeleteFindingAggregatorCommand,
    DeleteInsightCommand: DeleteInsightCommand_1.DeleteInsightCommand,
    DeleteInvitationsCommand: DeleteInvitationsCommand_1.DeleteInvitationsCommand,
    DeleteMembersCommand: DeleteMembersCommand_1.DeleteMembersCommand,
    DescribeActionTargetsCommand: DescribeActionTargetsCommand_1.DescribeActionTargetsCommand,
    DescribeHubCommand: DescribeHubCommand_1.DescribeHubCommand,
    DescribeOrganizationConfigurationCommand: DescribeOrganizationConfigurationCommand_1.DescribeOrganizationConfigurationCommand,
    DescribeProductsCommand: DescribeProductsCommand_1.DescribeProductsCommand,
    DescribeStandardsCommand: DescribeStandardsCommand_1.DescribeStandardsCommand,
    DescribeStandardsControlsCommand: DescribeStandardsControlsCommand_1.DescribeStandardsControlsCommand,
    DisableImportFindingsForProductCommand: DisableImportFindingsForProductCommand_1.DisableImportFindingsForProductCommand,
    DisableOrganizationAdminAccountCommand: DisableOrganizationAdminAccountCommand_1.DisableOrganizationAdminAccountCommand,
    DisableSecurityHubCommand: DisableSecurityHubCommand_1.DisableSecurityHubCommand,
    DisassociateFromAdministratorAccountCommand: DisassociateFromAdministratorAccountCommand_1.DisassociateFromAdministratorAccountCommand,
    DisassociateFromMasterAccountCommand: DisassociateFromMasterAccountCommand_1.DisassociateFromMasterAccountCommand,
    DisassociateMembersCommand: DisassociateMembersCommand_1.DisassociateMembersCommand,
    EnableImportFindingsForProductCommand: EnableImportFindingsForProductCommand_1.EnableImportFindingsForProductCommand,
    EnableOrganizationAdminAccountCommand: EnableOrganizationAdminAccountCommand_1.EnableOrganizationAdminAccountCommand,
    EnableSecurityHubCommand: EnableSecurityHubCommand_1.EnableSecurityHubCommand,
    GetAdministratorAccountCommand: GetAdministratorAccountCommand_1.GetAdministratorAccountCommand,
    GetConfigurationPolicyCommand: GetConfigurationPolicyCommand_1.GetConfigurationPolicyCommand,
    GetConfigurationPolicyAssociationCommand: GetConfigurationPolicyAssociationCommand_1.GetConfigurationPolicyAssociationCommand,
    GetEnabledStandardsCommand: GetEnabledStandardsCommand_1.GetEnabledStandardsCommand,
    GetFindingAggregatorCommand: GetFindingAggregatorCommand_1.GetFindingAggregatorCommand,
    GetFindingHistoryCommand: GetFindingHistoryCommand_1.GetFindingHistoryCommand,
    GetFindingsCommand: GetFindingsCommand_1.GetFindingsCommand,
    GetInsightResultsCommand: GetInsightResultsCommand_1.GetInsightResultsCommand,
    GetInsightsCommand: GetInsightsCommand_1.GetInsightsCommand,
    GetInvitationsCountCommand: GetInvitationsCountCommand_1.GetInvitationsCountCommand,
    GetMasterAccountCommand: GetMasterAccountCommand_1.GetMasterAccountCommand,
    GetMembersCommand: GetMembersCommand_1.GetMembersCommand,
    GetSecurityControlDefinitionCommand: GetSecurityControlDefinitionCommand_1.GetSecurityControlDefinitionCommand,
    InviteMembersCommand: InviteMembersCommand_1.InviteMembersCommand,
    ListAutomationRulesCommand: ListAutomationRulesCommand_1.ListAutomationRulesCommand,
    ListConfigurationPoliciesCommand: ListConfigurationPoliciesCommand_1.ListConfigurationPoliciesCommand,
    ListConfigurationPolicyAssociationsCommand: ListConfigurationPolicyAssociationsCommand_1.ListConfigurationPolicyAssociationsCommand,
    ListEnabledProductsForImportCommand: ListEnabledProductsForImportCommand_1.ListEnabledProductsForImportCommand,
    ListFindingAggregatorsCommand: ListFindingAggregatorsCommand_1.ListFindingAggregatorsCommand,
    ListInvitationsCommand: ListInvitationsCommand_1.ListInvitationsCommand,
    ListMembersCommand: ListMembersCommand_1.ListMembersCommand,
    ListOrganizationAdminAccountsCommand: ListOrganizationAdminAccountsCommand_1.ListOrganizationAdminAccountsCommand,
    ListSecurityControlDefinitionsCommand: ListSecurityControlDefinitionsCommand_1.ListSecurityControlDefinitionsCommand,
    ListStandardsControlAssociationsCommand: ListStandardsControlAssociationsCommand_1.ListStandardsControlAssociationsCommand,
    ListTagsForResourceCommand: ListTagsForResourceCommand_1.ListTagsForResourceCommand,
    StartConfigurationPolicyAssociationCommand: StartConfigurationPolicyAssociationCommand_1.StartConfigurationPolicyAssociationCommand,
    StartConfigurationPolicyDisassociationCommand: StartConfigurationPolicyDisassociationCommand_1.StartConfigurationPolicyDisassociationCommand,
    TagResourceCommand: TagResourceCommand_1.TagResourceCommand,
    UntagResourceCommand: UntagResourceCommand_1.UntagResourceCommand,
    UpdateActionTargetCommand: UpdateActionTargetCommand_1.UpdateActionTargetCommand,
    UpdateConfigurationPolicyCommand: UpdateConfigurationPolicyCommand_1.UpdateConfigurationPolicyCommand,
    UpdateFindingAggregatorCommand: UpdateFindingAggregatorCommand_1.UpdateFindingAggregatorCommand,
    UpdateFindingsCommand: UpdateFindingsCommand_1.UpdateFindingsCommand,
    UpdateInsightCommand: UpdateInsightCommand_1.UpdateInsightCommand,
    UpdateOrganizationConfigurationCommand: UpdateOrganizationConfigurationCommand_1.UpdateOrganizationConfigurationCommand,
    UpdateSecurityControlCommand: UpdateSecurityControlCommand_1.UpdateSecurityControlCommand,
    UpdateSecurityHubConfigurationCommand: UpdateSecurityHubConfigurationCommand_1.UpdateSecurityHubConfigurationCommand,
    UpdateStandardsControlCommand: UpdateStandardsControlCommand_1.UpdateStandardsControlCommand,
};
class SecurityHub extends SecurityHubClient_1.SecurityHubClient {
}
exports.SecurityHub = SecurityHub;
(0, smithy_client_1.createAggregatedClient)(commands, SecurityHub);
