// @aws-sdk/client-securityhub/dist-es/models/enums.js
export const SeverityLabel = {
    CRITICAL: "CRITICAL",
    HIGH: "HIGH",
    INFORMATIONAL: "INFORMATIONAL",
    LOW: "LOW",
    MEDIUM: "MEDIUM",
};
export const VerificationState = {
    BENIGN_POSITIVE: "BENIGN_POSITIVE",
    FALSE_POSITIVE: "FALSE_POSITIVE",
    TRUE_POSITIVE: "TRUE_POSITIVE",
    UNKNOWN: "UNKNOWN",
};
export const WorkflowStatus = {
    NEW: "NEW",
    NOTIFIED: "NOTIFIED",
    RESOLVED: "RESOLVED",
    SUPPRESSED: "SUPPRESSED",
};
export const AutomationRulesActionType = {
    FINDING_FIELDS_UPDATE: "FINDING_FIELDS_UPDATE",
};
export const ActorSessionMfaStatus = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
export const AdminStatus = {
    DISABLE_IN_PROGRESS: "DISABLE_IN_PROGRESS",
    ENABLED: "ENABLED",
};
export const AllowedOperators = {
    AND: "AND",
    OR: "OR",
};
export const ConfigurationPolicyAssociationStatus = {
    FAILED: "FAILED",
    PENDING: "PENDING",
    SUCCESS: "SUCCESS",
};
export const AssociationType = {
    APPLIED: "APPLIED",
    INHERITED: "INHERITED",
};
export const AssociationStatus = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
export const AutoEnableStandards = {
    DEFAULT: "DEFAULT",
    NONE: "NONE",
};
export const AutomationRulesActionTypeV2 = {
    EXTERNAL_INTEGRATION: "EXTERNAL_INTEGRATION",
    FINDING_FIELDS_UPDATE: "FINDING_FIELDS_UPDATE",
};
export const StringFilterComparison = {
    CONTAINS: "CONTAINS",
    CONTAINS_WORD: "CONTAINS_WORD",
    EQUALS: "EQUALS",
    NOT_CONTAINS: "NOT_CONTAINS",
    NOT_EQUALS: "NOT_EQUALS",
    PREFIX: "PREFIX",
    PREFIX_NOT_EQUALS: "PREFIX_NOT_EQUALS",
};
export const DateRangeUnit = {
    DAYS: "DAYS",
};
export const MapFilterComparison = {
    CONTAINS: "CONTAINS",
    EQUALS: "EQUALS",
    NOT_CONTAINS: "NOT_CONTAINS",
    NOT_EQUALS: "NOT_EQUALS",
};
export const RuleStatus = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
export const RuleStatusV2 = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
export const AwsIamAccessKeyStatus = {
    ACTIVE: "Active",
    INACTIVE: "Inactive",
};
export const AwsS3BucketNotificationConfigurationS3KeyFilterRuleName = {
    PREFIX: "Prefix",
    SUFFIX: "Suffix",
};
export const ComplianceStatus = {
    FAILED: "FAILED",
    NOT_AVAILABLE: "NOT_AVAILABLE",
    PASSED: "PASSED",
    WARNING: "WARNING",
};
export const ConnectionDirection = {
    INBOUND: "INBOUND",
    OUTBOUND: "OUTBOUND",
};
export const MalwareState = {
    OBSERVED: "OBSERVED",
    REMOVAL_FAILED: "REMOVAL_FAILED",
    REMOVED: "REMOVED",
};
export const MalwareType = {
    ADWARE: "ADWARE",
    BLENDED_THREAT: "BLENDED_THREAT",
    BOTNET_AGENT: "BOTNET_AGENT",
    COIN_MINER: "COIN_MINER",
    EXPLOIT_KIT: "EXPLOIT_KIT",
    KEYLOGGER: "KEYLOGGER",
    MACRO: "MACRO",
    POTENTIALLY_UNWANTED: "POTENTIALLY_UNWANTED",
    RANSOMWARE: "RANSOMWARE",
    REMOTE_ACCESS: "REMOTE_ACCESS",
    ROOTKIT: "ROOTKIT",
    SPYWARE: "SPYWARE",
    TROJAN: "TROJAN",
    VIRUS: "VIRUS",
    WORM: "WORM",
};
export const NetworkDirection = {
    IN: "IN",
    OUT: "OUT",
};
export const RecordState = {
    ACTIVE: "ACTIVE",
    ARCHIVED: "ARCHIVED",
};
export const Partition = {
    AWS: "aws",
    AWS_CN: "aws-cn",
    AWS_US_GOV: "aws-us-gov",
};
export const ThreatIntelIndicatorCategory = {
    BACKDOOR: "BACKDOOR",
    CARD_STEALER: "CARD_STEALER",
    COMMAND_AND_CONTROL: "COMMAND_AND_CONTROL",
    DROP_SITE: "DROP_SITE",
    EXPLOIT_SITE: "EXPLOIT_SITE",
    KEYLOGGER: "KEYLOGGER",
};
export const ThreatIntelIndicatorType = {
    DOMAIN: "DOMAIN",
    EMAIL_ADDRESS: "EMAIL_ADDRESS",
    HASH_MD5: "HASH_MD5",
    HASH_SHA1: "HASH_SHA1",
    HASH_SHA256: "HASH_SHA256",
    HASH_SHA512: "HASH_SHA512",
    IPV4_ADDRESS: "IPV4_ADDRESS",
    IPV6_ADDRESS: "IPV6_ADDRESS",
    MUTEX: "MUTEX",
    PROCESS: "PROCESS",
    URL: "URL",
};
export const VulnerabilityExploitAvailable = {
    NO: "NO",
    YES: "YES",
};
export const VulnerabilityFixAvailable = {
    NO: "NO",
    PARTIAL: "PARTIAL",
    YES: "YES",
};
export const WorkflowState = {
    ASSIGNED: "ASSIGNED",
    DEFERRED: "DEFERRED",
    IN_PROGRESS: "IN_PROGRESS",
    NEW: "NEW",
    RESOLVED: "RESOLVED",
};
export const StandardsControlsUpdatable = {
    NOT_READY_FOR_UPDATES: "NOT_READY_FOR_UPDATES",
    READY_FOR_UPDATES: "READY_FOR_UPDATES",
};
export const StandardsStatus = {
    DELETING: "DELETING",
    FAILED: "FAILED",
    INCOMPLETE: "INCOMPLETE",
    PENDING: "PENDING",
    READY: "READY",
};
export const StatusReasonCode = {
    INTERNAL_ERROR: "INTERNAL_ERROR",
    MAXIMUM_NUMBER_OF_CONFIG_RULES_EXCEEDED: "MAXIMUM_NUMBER_OF_CONFIG_RULES_EXCEEDED",
    NO_AVAILABLE_CONFIGURATION_RECORDER: "NO_AVAILABLE_CONFIGURATION_RECORDER",
};
export const TargetType = {
    ACCOUNT: "ACCOUNT",
    ORGANIZATIONAL_UNIT: "ORGANIZATIONAL_UNIT",
    ROOT: "ROOT",
};
export const ParameterValueType = {
    CUSTOM: "CUSTOM",
    DEFAULT: "DEFAULT",
};
export const ControlStatus = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
export const SeverityRating = {
    CRITICAL: "CRITICAL",
    HIGH: "HIGH",
    LOW: "LOW",
    MEDIUM: "MEDIUM",
};
export const UpdateStatus = {
    READY: "READY",
    UPDATING: "UPDATING",
};
export const UnprocessedErrorCode = {
    ACCESS_DENIED: "ACCESS_DENIED",
    INVALID_INPUT: "INVALID_INPUT",
    LIMIT_EXCEEDED: "LIMIT_EXCEEDED",
    NOT_FOUND: "NOT_FOUND",
    RESOURCE_NOT_FOUND: "RESOURCE_NOT_FOUND",
};
export const BatchUpdateFindingsV2UnprocessedFindingErrorCode = {
    ConflictException: "ConflictException",
    InternalServerException: "InternalServerException",
    ResourceNotFoundException: "ResourceNotFoundException",
    ValidationException: "ValidationException",
};
export const OcsfBooleanField = {
    COMPLIANCE_ASSESSMENTS_MEETS_CRITERIA: "compliance.assessments.meets_criteria",
    VULNERABILITIES_IS_EXPLOIT_AVAILABLE: "vulnerabilities.is_exploit_available",
    VULNERABILITIES_IS_FIX_AVAILABLE: "vulnerabilities.is_fix_available",
};
export const OcsfDateField = {
    FINDING_INFO_CREATED_TIME_DT: "finding_info.created_time_dt",
    FINDING_INFO_FIRST_SEEN_TIME_DT: "finding_info.first_seen_time_dt",
    FINDING_INFO_LAST_SEEN_TIME_DT: "finding_info.last_seen_time_dt",
    FINDING_INFO_MODIFIED_TIME_DT: "finding_info.modified_time_dt",
    RESOURCES_IMAGE_CREATED_TIME_DT: "resources.image.created_time_dt",
    RESOURCES_IMAGE_LAST_USED_TIME_DT: "resources.image.last_used_time_dt",
    RESOURCES_MODIFIED_TIME_DT: "resources.modified_time_dt",
};
export const OcsfIpField = {
    EVIDENCES_DST_ENDPOINT_IP: "evidences.dst_endpoint.ip",
    EVIDENCES_SRC_ENDPOINT_IP: "evidences.src_endpoint.ip",
};
export const OcsfMapField = {
    COMPLIANCE_CONTROL_PARAMETERS: "compliance.control_parameters",
    DATABUCKET_TAGS: "databucket.tags",
    FINDING_INFO_TAGS: "finding_info.tags",
    RESOURCES_TAGS: "resources.tags",
};
export const OcsfNumberField = {
    ACTIVITY_ID: "activity_id",
    COMPLIANCE_STATUS_ID: "compliance.status_id",
    CONFIDENCE_SCORE: "confidence_score",
    EVIDENCES_API_RESPONSE_CODE: "evidences.api.response.code",
    EVIDENCES_DST_ENDPOINT_AUTONOMOUS_SYSTEM_NUMBER: "evidences.dst_endpoint.autonomous_system.number",
    EVIDENCES_DST_ENDPOINT_PORT: "evidences.dst_endpoint.port",
    EVIDENCES_SRC_ENDPOINT_AUTONOMOUS_SYSTEM_NUMBER: "evidences.src_endpoint.autonomous_system.number",
    EVIDENCES_SRC_ENDPOINT_PORT: "evidences.src_endpoint.port",
    FINDING_INFO_RELATED_EVENTS_COUNT: "finding_info.related_events_count",
    RESOURCES_IMAGE_IN_USE_COUNT: "resources.image.in_use_count",
    SEVERITY_ID: "severity_id",
    STATUS_ID: "status_id",
    VENDOR_ATTRIBUTES_SEVERITY_ID: "vendor_attributes.severity_id",
    VULNERABILITIES_CVE_CVSS_BASE_SCORE: "vulnerabilities.cve.cvss.base_score",
};
export const OcsfStringField = {
    ACTIVITY_NAME: "activity_name",
    CLASS_NAME: "class_name",
    CLOUD_ACCOUNT_NAME: "cloud.account.name",
    CLOUD_ACCOUNT_UID: "cloud.account.uid",
    CLOUD_PROVIDER: "cloud.provider",
    CLOUD_REGION: "cloud.region",
    COMMENT: "comment",
    COMPLIANCE_ASSESSMENTS_CATEGORY: "compliance.assessments.category",
    COMPLIANCE_ASSESSMENTS_NAME: "compliance.assessments.name",
    COMPLIANCE_CONTROL: "compliance.control",
    COMPLIANCE_STANDARDS: "compliance.standards",
    COMPLIANCE_STATUS: "compliance.status",
    DATABUCKET_ENCRYPTION_DETAILS_ALGORITHM: "databucket.encryption_details.algorithm",
    DATABUCKET_ENCRYPTION_DETAILS_KEY_UID: "databucket.encryption_details.key_uid",
    DATABUCKET_FILE_DATA_CLASSIFICATIONS_CLASSIFIER_DETAILS_TYPE: "databucket.file.data_classifications.classifier_details.type",
    EVIDENCES_ACTOR_USER_ACCOUNT_UID: "evidences.actor.user.account.uid",
    EVIDENCES_API_OPERATION: "evidences.api.operation",
    EVIDENCES_API_RESPONSE_ERROR_MESSAGE: "evidences.api.response.error_message",
    EVIDENCES_API_SERVICE_NAME: "evidences.api.service.name",
    EVIDENCES_CONNECTION_INFO_DIRECTION: "evidences.connection_info.direction",
    EVIDENCES_CONNECTION_INFO_PROTOCOL_NAME: "evidences.connection_info.protocol_name",
    EVIDENCES_DST_ENDPOINT_AUTONOMOUS_SYSTEM_NAME: "evidences.dst_endpoint.autonomous_system.name",
    EVIDENCES_DST_ENDPOINT_LOCATION_CITY: "evidences.dst_endpoint.location.city",
    EVIDENCES_DST_ENDPOINT_LOCATION_COUNTRY: "evidences.dst_endpoint.location.country",
    EVIDENCES_SRC_ENDPOINT_AUTONOMOUS_SYSTEM_NAME: "evidences.src_endpoint.autonomous_system.name",
    EVIDENCES_SRC_ENDPOINT_HOSTNAME: "evidences.src_endpoint.hostname",
    EVIDENCES_SRC_ENDPOINT_LOCATION_CITY: "evidences.src_endpoint.location.city",
    EVIDENCES_SRC_ENDPOINT_LOCATION_COUNTRY: "evidences.src_endpoint.location.country",
    FINDING_INFO_ANALYTIC_NAME: "finding_info.analytic.name",
    FINDING_INFO_DESC: "finding_info.desc",
    FINDING_INFO_RELATED_EVENTS_PRODUCT_UID: "finding_info.related_events.product.uid",
    FINDING_INFO_RELATED_EVENTS_TITLE: "finding_info.related_events.title",
    FINDING_INFO_RELATED_EVENTS_TRAITS_CATEGORY: "finding_info.related_events.traits.category",
    FINDING_INFO_RELATED_EVENTS_UID: "finding_info.related_events.uid",
    FINDING_INFO_SRC_URL: "finding_info.src_url",
    FINDING_INFO_TITLE: "finding_info.title",
    FINDING_INFO_TYPES: "finding_info.types",
    FINDING_INFO_UID: "finding_info.uid",
    MALWARE_NAME: "malware.name",
    MALWARE_SCAN_INFO_UID: "malware_scan_info.uid",
    MALWARE_SEVERITY: "malware.severity",
    METADATA_PRODUCT_NAME: "metadata.product.name",
    METADATA_PRODUCT_UID: "metadata.product.uid",
    METADATA_PRODUCT_VENDOR_NAME: "metadata.product.vendor_name",
    METADATA_UID: "metadata.uid",
    REMEDIATION_DESC: "remediation.desc",
    REMEDIATION_REFERENCES: "remediation.references",
    RESOURCES_CLOUD_FUNCTION_LAYERS_UID_ALT: "resources.cloud_function.layers.uid_alt",
    RESOURCES_CLOUD_FUNCTION_RUNTIME: "resources.cloud_function.runtime",
    RESOURCES_CLOUD_FUNCTION_USER_UID: "resources.cloud_function.user.uid",
    RESOURCES_CLOUD_PARTITION: "resources.cloud_partition",
    RESOURCES_DEVICE_ENCRYPTION_DETAILS_KEY_UID: "resources.device.encryption_details.key_uid",
    RESOURCES_DEVICE_IMAGE_UID: "resources.device.image.uid",
    RESOURCES_IMAGE_ARCHITECTURE: "resources.image.architecture",
    RESOURCES_IMAGE_REGISTRY_UID: "resources.image.registry_uid",
    RESOURCES_IMAGE_REPOSITORY_NAME: "resources.image.repository_name",
    RESOURCES_IMAGE_UID: "resources.image.uid",
    RESOURCES_REGION: "resources.region",
    RESOURCES_SUBNET_INFO_UID: "resources.subnet_info.uid",
    RESOURCES_TYPE: "resources.type",
    RESOURCES_UID: "resources.uid",
    RESOURCES_VPC_UID: "resources.vpc_uid",
    SEVERITY: "severity",
    STATUS: "status",
    VENDOR_ATTRIBUTES_SEVERITY: "vendor_attributes.severity",
    VULNERABILITIES_AFFECTED_CODE_FILE_PATH: "vulnerabilities.affected_code.file.path",
    VULNERABILITIES_AFFECTED_PACKAGES_NAME: "vulnerabilities.affected_packages.name",
    VULNERABILITIES_CVE_EPSS_SCORE: "vulnerabilities.cve.epss.score",
    VULNERABILITIES_CVE_UID: "vulnerabilities.cve.uid",
    VULNERABILITIES_FIX_COVERAGE: "vulnerabilities.fix_coverage",
    VULNERABILITIES_RELATED_VULNERABILITIES: "vulnerabilities.related_vulnerabilities",
};
export const ConnectorAuthStatus = {
    ACTIVE: "ACTIVE",
    FAILED: "FAILED",
};
export const ConnectorProviderName = {
    JIRA_CLOUD: "JIRA_CLOUD",
    SERVICENOW: "SERVICENOW",
};
export const ConnectorStatus = {
    CONNECTED: "CONNECTED",
    FAILED_TO_CONNECT: "FAILED_TO_CONNECT",
    PENDING_AUTHORIZATION: "PENDING_AUTHORIZATION",
    PENDING_CONFIGURATION: "PENDING_CONFIGURATION",
};
export const ControlFindingGenerator = {
    SECURITY_CONTROL: "SECURITY_CONTROL",
    STANDARD_CONTROL: "STANDARD_CONTROL",
};
export const TicketCreationMode = {
    DRYRUN: "DRYRUN",
};
export const SecurityControlProperty = {
    Parameters: "Parameters",
};
export const OrganizationConfigurationConfigurationType = {
    CENTRAL: "CENTRAL",
    LOCAL: "LOCAL",
};
export const OrganizationConfigurationStatus = {
    ENABLED: "ENABLED",
    FAILED: "FAILED",
    PENDING: "PENDING",
};
export const IntegrationType = {
    RECEIVE_FINDINGS_FROM_SECURITY_HUB: "RECEIVE_FINDINGS_FROM_SECURITY_HUB",
    SEND_FINDINGS_TO_SECURITY_HUB: "SEND_FINDINGS_TO_SECURITY_HUB",
    UPDATE_FINDINGS_IN_SECURITY_HUB: "UPDATE_FINDINGS_IN_SECURITY_HUB",
};
export const IntegrationV2Type = {
    RECEIVE_FINDINGS_FROM_SECURITY_HUB: "RECEIVE_FINDINGS_FROM_SECURITY_HUB",
    SEND_FINDINGS_TO_SECURITY_HUB: "SEND_FINDINGS_TO_SECURITY_HUB",
    UPDATE_FINDINGS_IN_SECURITY_HUB: "UPDATE_FINDINGS_IN_SECURITY_HUB",
};
export const SecurityHubFeature = {
    SECURITY_HUB: "SecurityHub",
    SECURITY_HUB_V2: "SecurityHubV2",
};
export const FindingHistoryUpdateSourceType = {
    BATCH_IMPORT_FINDINGS: "BATCH_IMPORT_FINDINGS",
    BATCH_UPDATE_FINDINGS: "BATCH_UPDATE_FINDINGS",
};
export const FindingsTrendsStringField = {
    ACCOUNT_ID: "account_id",
    COMPLIANCE_CONTROL: "finding_control_id",
    COMPLIANCE_STATUS: "finding_compliance_status",
    CVE_ID: "finding_cve_ids",
    FINDING_ACTIVITY_NAME: "finding_activity_name",
    FINDING_CLASS: "finding_class_name",
    FINDING_STATUS: "finding_status",
    FINDING_TYPE: "finding_types",
    PROVIDER_NAME: "finding_provider",
    REGION: "region",
};
export const SortOrder = {
    ASCENDING: "asc",
    DESCENDING: "desc",
};
export const GroupByField = {
    ACTIVITY_NAME: "activity_name",
    CLASS_NAME: "class_name",
    CLOUD_ACCOUNT_NAME: "cloud.account.name",
    CLOUD_ACCOUNT_UID: "cloud.account.uid",
    CLOUD_PROVIDER: "cloud.provider",
    CLOUD_REGION: "cloud.region",
    COMPLIANCE_ASSESSMENTS_NAME: "compliance.assessments.name",
    COMPLIANCE_CONTROL: "compliance.control",
    COMPLIANCE_STANDARDS: "compliance.standards",
    COMPLIANCE_STATUS: "compliance.status",
    FINDING_INFO_ANALYTIC_NAME: "finding_info.analytic.name",
    FINDING_INFO_RELATED_EVENTS_TRAITS_CATEGORY: "finding_info.related_events.traits.category",
    FINDING_INFO_TITLE: "finding_info.title",
    FINDING_INFO_TYPES: "finding_info.types",
    METADATA_PRODUCT_NAME: "metadata.product.name",
    METADATA_PRODUCT_UID: "metadata.product.uid",
    RESOURCES_TYPE: "resources.type",
    RESOURCES_UID: "resources.uid",
    SEVERITY: "severity",
    STATUS: "status",
    VENDOR_ATTRIBUTES_SEVERITY: "vendor_attributes.severity",
    VULNERABILITIES_AFFECTED_PACKAGES_NAME: "vulnerabilities.affected_packages.name",
    VULNERABILITIES_FIX_COVERAGE: "vulnerabilities.fix_coverage",
};
export const GranularityField = {
    DAILY: "Daily",
    MONTHLY: "Monthly",
    WEEKLY: "Weekly",
};
export const ResourcesDateField = {
    RESOURCE_CREATION_TIME_DT: "ResourceCreationTime",
    RESOURCE_DETAIL_CAPTURE_TIME_DT: "ResourceDetailCaptureTime",
};
export const ResourcesMapField = {
    TAG: "ResourceTags",
};
export const ResourcesNumberField = {
    SEVERITY_CRITICAL: "FindingsSummary.Severities.Critical",
    SEVERITY_FATAL: "FindingsSummary.Severities.Fatal",
    SEVERITY_HIGH: "FindingsSummary.Severities.High",
    SEVERITY_INFORMATIONAL: "FindingsSummary.Severities.Informational",
    SEVERITY_LOW: "FindingsSummary.Severities.Low",
    SEVERITY_MEDIUM: "FindingsSummary.Severities.Medium",
    SEVERITY_OTHER: "FindingsSummary.Severities.Other",
    SEVERITY_UNKNOWN: "FindingsSummary.Severities.Unknown",
    TOTAL_FINDINGS: "FindingsSummary.TotalFindings",
};
export const ResourcesStringField = {
    ACCOUNT_ID: "AccountId",
    FINDING_TYPE: "FindingsSummary.FindingType",
    PRODUCT_NAME: "FindingsSummary.ProductName",
    REGION: "Region",
    RESOURCE_CATEGORY: "ResourceCategory",
    RESOURCE_GUID: "ResourceGuid",
    RESOURCE_ID: "ResourceId",
    RESOURCE_NAME: "ResourceName",
    RESOURCE_TYPE: "ResourceType",
};
export const ResourceGroupByField = {
    ACCOUNT_ID: "AccountId",
    FINDING_TYPE: "FindingsSummary.FindingType",
    REGION: "Region",
    RESOURCE_CATEGORY: "ResourceCategory",
    RESOURCE_NAME: "ResourceName",
    RESOURCE_TYPE: "ResourceType",
};
export const ResourcesTrendsStringField = {
    ACCOUNT_ID: "account_id",
    REGION: "region",
    RESOURCE_CATEGORY: "resource_category",
    RESOURCE_TYPE: "resource_type",
};
export const ResourceCategory = {
    AI_ML: "AI/ML",
    CODE: "Code",
    COMPUTE: "Compute",
    DATABASE: "Database",
    IDENTITY: "Identity",
    NETWORK: "Network",
    OTHER: "Other",
    STORAGE: "Storage",
};
export const RegionAvailabilityStatus = {
    AVAILABLE: "AVAILABLE",
    UNAVAILABLE: "UNAVAILABLE",
};
