// @aws-sdk/client-securityhub/dist-es/models/models_3.js
import { SENSITIVE_STRING } from "@smithy/smithy-client";
import { SecurityHubServiceException as __BaseException } from "./SecurityHubServiceException";
export const GroupByField = {
    ACTIVITY_NAME: "activity_name",
    CLASS_NAME: "class_name",
    CLOUD_ACCOUNT_UID: "cloud.account.uid",
    CLOUD_PROVIDER: "cloud.provider",
    CLOUD_REGION: "cloud.region",
    COMPLIANCE_ASSESSMENTS_NAME: "compliance.assessments.name",
    COMPLIANCE_CONTROL: "compliance.control",
    COMPLIANCE_STATUS: "compliance.status",
    FINDING_INFO_TITLE: "finding_info.title",
    FINDING_INFO_TYPES: "finding_info.types",
    METADATA_PRODUCT_NAME: "metadata.product.name",
    METADATA_PRODUCT_UID: "metadata.product.uid",
    RESOURCES_TYPE: "resources.type",
    RESOURCES_UID: "resources.uid",
    SEVERITY: "severity",
    STATUS: "status",
    VULNERABILITIES_FIX_COVERAGE: "vulnerabilities.fix_coverage",
};
export const ResourcesDateField = {
    RESOURCE_CREATION_TIME_DT: "resource_creation_time_dt",
    RESOURCE_DETAIL_CAPTURE_TIME_DT: "resource_detail_capture_time_dt",
};
export const ResourcesMapField = {
    TAG: "tags",
};
export const ResourcesNumberField = {
    SEVERITY_CRITICAL: "findings_summary.severities.critical",
    SEVERITY_FATAL: "findings_summary.severities.fatal",
    SEVERITY_HIGH: "findings_summary.severities.high",
    SEVERITY_INFORMATIONAL: "findings_summary.severities.informational",
    SEVERITY_LOW: "findings_summary.severities.low",
    SEVERITY_MEDIUM: "findings_summary.severities.medium",
    SEVERITY_OTHER: "findings_summary.severities.other",
    SEVERITY_UNKNOWN: "findings_summary.severities.unknown",
    TOTAL_FINDINGS: "findings_summary.total_findings",
};
export const ResourcesStringField = {
    ACCOUNT_ID: "account_id",
    FINDING_TYPE: "findings_summary.finding_type",
    PRODUCT_NAME: "findings_summary.product_name",
    REGION: "region",
    RESOURCE_ARN: "resource_arn",
    RESOURCE_CATEGORY: "resource_category",
    RESOURCE_ID: "resource_id",
    RESOURCE_NAME: "resource_name",
    RESOURCE_TYPE: "resource_type",
};
export const ResourceGroupByField = {
    ACCOUNT_ID: "account_id",
    FINDING_TYPE: "findings_summary.finding_type",
    REGION: "region",
    RESOURCE_CATEGORY: "resource_category",
    RESOURCE_NAME: "resource_name",
    RESOURCE_TYPE: "resource_type",
};
export const ResourceCategory = {
    AI_ML: "AI/ML",
    CODE: "Code",
    COMPUTE: "Compute",
    DATABASE: "Database",
    IDENTITY: "Identity",
    NETWORK: "Network",
    OTHER: "Other",
    STORAGE: "Storage",
};
export const RegionAvailabilityStatus = {
    AVAILABLE: "AVAILABLE",
    UNAVAILABLE: "UNAVAILABLE",
};
export var ProviderUpdateConfiguration;
(function (ProviderUpdateConfiguration) {
    ProviderUpdateConfiguration.visit = (value, visitor) => {
        if (value.JiraCloud !== undefined)
            return visitor.JiraCloud(value.JiraCloud);
        return visitor._(value.$unknown[0], value.$unknown[1]);
    };
})(ProviderUpdateConfiguration || (ProviderUpdateConfiguration = {}));
export class ResourceInUseException extends __BaseException {
    name = "ResourceInUseException";
    $fault = "client";
    Message;
    Code;
    constructor(opts) {
        super({
            name: "ResourceInUseException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceInUseException.prototype);
        this.Message = opts.Message;
        this.Code = opts.Code;
    }
}
export const UpdateConnectorV2RequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ClientSecret && { ClientSecret: SENSITIVE_STRING }),
    ...(obj.Provider && { Provider: obj.Provider }),
});
