// @aws-sdk/client-securityhub/dist-es/models/models_1.js
export const AwsIamAccessKeyStatus = {
    ACTIVE: "Active",
    INACTIVE: "Inactive",
};
export const AwsS3BucketNotificationConfigurationS3KeyFilterRuleName = {
    PREFIX: "Prefix",
    SUFFIX: "Suffix",
};
export const ComplianceStatus = {
    FAILED: "FAILED",
    NOT_AVAILABLE: "NOT_AVAILABLE",
    PASSED: "PASSED",
    WARNING: "WARNING",
};
export const MalwareState = {
    OBSERVED: "OBSERVED",
    REMOVAL_FAILED: "REMOVAL_FAILED",
    REMOVED: "REMOVED",
};
export const MalwareType = {
    ADWARE: "ADWARE",
    BLENDED_THREAT: "BLENDED_THREAT",
    BOTNET_AGENT: "BOTNET_AGENT",
    COIN_MINER: "COIN_MINER",
    EXPLOIT_KIT: "EXPLOIT_KIT",
    KEYLOGGER: "KEYLOGGER",
    MACRO: "MACRO",
    POTENTIALLY_UNWANTED: "POTENTIALLY_UNWANTED",
    RANSOMWARE: "RANSOMWARE",
    REMOTE_ACCESS: "REMOTE_ACCESS",
    ROOTKIT: "ROOTKIT",
    SPYWARE: "SPYWARE",
    TROJAN: "TROJAN",
    VIRUS: "VIRUS",
    WORM: "WORM",
};
export const NetworkDirection = {
    IN: "IN",
    OUT: "OUT",
};
export const RecordState = {
    ACTIVE: "ACTIVE",
    ARCHIVED: "ARCHIVED",
};
