// @aws-sdk/client-config-service/dist-cjs/commands/SelectAggregateResourceConfigCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectAggregateResourceConfigCommand = void 0;
const middleware_serde_1 = require("@aws-sdk/middleware-serde");
const smithy_client_1 = require("@aws-sdk/smithy-client");
const models_1_1 = require("../models/models_1");
const Aws_json1_1_1 = require("../protocols/Aws_json1_1");
class SelectAggregateResourceConfigCommand extends smithy_client_1.Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use((0, middleware_serde_1.getSerdePlugin)(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "ConfigServiceClient";
        const commandName = "SelectAggregateResourceConfigCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: models_1_1.SelectAggregateResourceConfigRequest.filterSensitiveLog,
            outputFilterSensitiveLog: models_1_1.SelectAggregateResourceConfigResponse.filterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return (0, Aws_json1_1_1.serializeAws_json1_1SelectAggregateResourceConfigCommand)(input, context);
    }
    deserialize(output, context) {
        return (0, Aws_json1_1_1.deserializeAws_json1_1SelectAggregateResourceConfigCommand)(output, context);
    }
}
exports.SelectAggregateResourceConfigCommand = SelectAggregateResourceConfigCommand;
