// @mitre/hdf-converters/lib/src/ckl-mapper/checklistJsonix.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Vulnattribute = exports.Status = exports.Severityoverride = exports.Sidname = exports.Techarea = exports.Role = exports.Assettype = exports.LocalPartEnum = void 0;
var LocalPartEnum;
(function (LocalPartEnum) {
    LocalPartEnum["Asset"] = "ASSET";
    LocalPartEnum["AssetType"] = "ASSET_TYPE";
    LocalPartEnum["AttributeData"] = "ATTRIBUTE_DATA";
    LocalPartEnum["Checklist"] = "CHECKLIST";
    LocalPartEnum["Comments"] = "COMMENTS";
    LocalPartEnum["FindingDetails"] = "FINDING_DETAILS";
    LocalPartEnum["HostFQDN"] = "HOST_FQDN";
    LocalPartEnum["HostGUID"] = "HOST_GUID";
    LocalPartEnum["HostIP"] = "HOST_IP";
    LocalPartEnum["HostMAC"] = "HOST_MAC";
    LocalPartEnum["HostName"] = "HOST_NAME";
    LocalPartEnum["ISTIG"] = "iSTIG";
    LocalPartEnum["Marking"] = "MARKING";
    LocalPartEnum["Role"] = "ROLE";
    LocalPartEnum["SeverityJustification"] = "SEVERITY_JUSTIFICATION";
    LocalPartEnum["SeverityOverride"] = "SEVERITY_OVERRIDE";
    LocalPartEnum["SiData"] = "SI_DATA";
    LocalPartEnum["SidData"] = "SID_DATA";
    LocalPartEnum["SidName"] = "SID_NAME";
    LocalPartEnum["Status"] = "STATUS";
    LocalPartEnum["StigData"] = "STIG_DATA";
    LocalPartEnum["StigGUID"] = "STIG_GUID";
    LocalPartEnum["StigInfo"] = "STIG_INFO";
    LocalPartEnum["Stigs"] = "STIGS";
    LocalPartEnum["TargetComment"] = "TARGET_COMMENT";
    LocalPartEnum["TargetKey"] = "TARGET_KEY";
    LocalPartEnum["TechArea"] = "TECH_AREA";
    LocalPartEnum["Vuln"] = "VULN";
    LocalPartEnum["VulnAttribute"] = "VULN_ATTRIBUTE";
    LocalPartEnum["WebDBInstance"] = "WEB_DB_INSTANCE";
    LocalPartEnum["WebDBSite"] = "WEB_DB_SITE";
    LocalPartEnum["WebOrDatabase"] = "WEB_OR_DATABASE";
})(LocalPartEnum = exports.LocalPartEnum || (exports.LocalPartEnum = {}));
var Assettype;
(function (Assettype) {
    Assettype["Computing"] = "Computing";
    Assettype["NonComputing"] = "Non-Computing";
})(Assettype = exports.Assettype || (exports.Assettype = {}));
var Role;
(function (Role) {
    Role["DomainController"] = "Domain Controller";
    Role["MemberServer"] = "Member Server";
    Role["None"] = "None";
    Role["Workstation"] = "Workstation";
})(Role = exports.Role || (exports.Role = {}));
var Techarea;
(function (Techarea) {
    Techarea["Empty"] = "";
    Techarea["ApplicationReview"] = "Application Review";
    Techarea["BoundarySecurity"] = "Boundary Security";
    Techarea["CDSAdminReview"] = "CDS Admin Review";
    Techarea["CDSTechnicalReview"] = "CDS Technical Review";
    Techarea["DatabaseReview"] = "Database Review";
    Techarea["DomainNameSystemDNS"] = "Domain Name System (DNS)";
    Techarea["ExchangeServer"] = "Exchange Server";
    Techarea["HostBasedSystemSecurityHBSS"] = "Host Based System Security (HBSS)";
    Techarea["InternalNetwork"] = "Internal Network";
    Techarea["Mobility"] = "Mobility";
    Techarea["OtherReview"] = "Other Review";
    Techarea["ReleasableNetworksREL"] = "Releasable Networks (REL)";
    Techarea["ReleaseableNetworksREL"] = "Releaseable Networks (REL)";
    Techarea["TraditionalSecurity"] = "Traditional Security";
    Techarea["UnixOS"] = "UNIX OS";
    Techarea["VVOIPReview"] = "VVOIP Review";
    Techarea["WebReview"] = "Web Review";
    Techarea["WindowsOS"] = "Windows OS";
})(Techarea = exports.Techarea || (exports.Techarea = {}));
var Sidname;
(function (Sidname) {
    Sidname["Classification"] = "classification";
    Sidname["Customname"] = "customname";
    Sidname["Description"] = "description";
    Sidname["Filename"] = "filename";
    Sidname["Notice"] = "notice";
    Sidname["Releaseinfo"] = "releaseinfo";
    Sidname["Source"] = "source";
    Sidname["Stigid"] = "stigid";
    Sidname["Title"] = "title";
    Sidname["UUID"] = "uuid";
    Sidname["Version"] = "version";
})(Sidname = exports.Sidname || (exports.Sidname = {}));
var Severityoverride;
(function (Severityoverride) {
    Severityoverride["Empty"] = "";
    Severityoverride["High"] = "high";
    Severityoverride["Low"] = "low";
    Severityoverride["Medium"] = "medium";
})(Severityoverride = exports.Severityoverride || (exports.Severityoverride = {}));
var Status;
(function (Status) {
    Status["NotAFinding"] = "NotAFinding";
    Status["NotApplicable"] = "Not_Applicable";
    Status["NotReviewed"] = "Not_Reviewed";
    Status["Open"] = "Open";
})(Status = exports.Status || (exports.Status = {}));
var Vulnattribute;
(function (Vulnattribute) {
    Vulnattribute["CciRef"] = "CCI_REF";
    Vulnattribute["CheckContent"] = "Check_Content";
    Vulnattribute["CheckContentRef"] = "Check_Content_Ref";
    Vulnattribute["Class"] = "Class";
    Vulnattribute["Documentable"] = "Documentable";
    Vulnattribute["FalseNegatives"] = "False_Negatives";
    Vulnattribute["FalsePositives"] = "False_Positives";
    Vulnattribute["FixText"] = "Fix_Text";
    Vulnattribute["GroupTitle"] = "Group_Title";
    Vulnattribute["IAControls"] = "IA_Controls";
    Vulnattribute["LegacyID"] = "LEGACY_ID";
    Vulnattribute["MitigationControl"] = "Mitigation_Control";
    Vulnattribute["Mitigations"] = "Mitigations";
    Vulnattribute["PotentialImpact"] = "Potential_Impact";
    Vulnattribute["Responsibility"] = "Responsibility";
    Vulnattribute["RuleID"] = "Rule_ID";
    Vulnattribute["RuleTitle"] = "Rule_Title";
    Vulnattribute["RuleVer"] = "Rule_Ver";
    Vulnattribute["STIGRef"] = "STIGRef";
    Vulnattribute["SecurityOverrideGuidance"] = "Security_Override_Guidance";
    Vulnattribute["Severity"] = "Severity";
    Vulnattribute["StigUUID"] = "STIG_UUID";
    Vulnattribute["TargetKey"] = "TargetKey";
    Vulnattribute["ThirdPartyTools"] = "Third_Party_Tools";
    Vulnattribute["VulnDiscuss"] = "Vuln_Discuss";
    Vulnattribute["VulnNum"] = "Vuln_Num";
    Vulnattribute["Weight"] = "Weight";
})(Vulnattribute = exports.Vulnattribute || (exports.Vulnattribute = {}));
