// @mitre/inspec-objects/lib/objects/profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const yaml_1 = tslib_1.__importDefault(require("yaml"));
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const global_1 = require("../utilities/global");
class Profile {
    constructor(data) {
        this.supports = [];
        this.depends = [];
        this.inputs = [];
        this.libraries = [];
        this.files = [];
        this.controls = [];
        if (data) {
            Object.entries(data).forEach(([key, value]) => {
                lodash_1.default.set(this, key, value);
            });
        }
    }
    createInspecYaml() {
        return yaml_1.default.stringify({
            name: this.name,
            title: this.title,
            maintainer: this.maintainer,
            copyright: this.copyright,
            copyright_email: this.copyright_email,
            license: this.license,
            summary: this.summary,
            description: this.description,
            version: this.version,
            supports: this.supports,
            depends: this.depends,
            inspec_version: this.inspec_version,
        });
    }
    toUnformattedObject() {
        const unformattedProfile = new Profile(this);
        Object.entries(this).forEach(([key, value]) => {
            if (typeof value === 'string') {
                lodash_1.default.set(unformattedProfile, key, (0, global_1.unformatText)(value));
            }
        });
        unformattedProfile.controls = this.controls.map((control) => control.toUnformattedObject());
        return unformattedProfile;
    }
}
exports.default = Profile;
