// @oclif/core/lib/cli-ux/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.config = exports.ActionBase = exports.Table = exports.ExitError = exports.warn = exports.wait = exports.url = exports.tree = exports.trace = exports.table = exports.styledObject = exports.styledJSON = exports.styledHeader = exports.prompt = exports.progress = exports.log = exports.info = exports.flush = exports.exit = exports.error = exports.done = exports.debug = exports.confirm = exports.anykey = exports.annotation = exports.action = exports.ux = void 0;
const tslib_1 = require("tslib");
const Errors = tslib_1.__importStar(require("../errors"));
const styled = tslib_1.__importStar(require("./styled"));
const uxPrompt = tslib_1.__importStar(require("./prompt"));
const config_1 = require("./config");
const flush_1 = require("./flush");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const stream_1 = require("./stream");
const node_util_1 = require("node:util");
const wait_1 = tslib_1.__importDefault(require("./wait"));
const hyperlinker = require('hyperlinker');
class ux {
    static config = config_1.config;
    static get prompt() {
        return uxPrompt.prompt;
    }
    /**
     * "press anykey to continue"
     */
    static get anykey() {
        return uxPrompt.anykey;
    }
    static get confirm() {
        return uxPrompt.confirm;
    }
    static get action() {
        return config_1.config.action;
    }
    static styledObject(obj, keys) {
        this.info(styled.styledObject(obj, keys));
    }
    static styledHeader(header) {
        this.info(chalk_1.default.dim('=== ') + chalk_1.default.bold(header) + '\n');
    }
    static styledJSON(obj) {
        const json = JSON.stringify(obj, null, 2);
        if (!chalk_1.default.level) {
            info(json);
            return;
        }
        const cardinal = require('cardinal');
        const theme = require('cardinal/themes/jq');
        this.info(cardinal.highlight(json, { json: true, theme }));
    }
    static get table() {
        return styled.Table.table;
    }
    static get tree() {
        return styled.tree;
    }
    static get wait() {
        return wait_1.default;
    }
    static get progress() {
        return styled.progress;
    }
    static async done() {
        config_1.config.action.stop();
    }
    static trace(format, ...args) {
        if (this.config.outputLevel === 'trace') {
            stream_1.stdout.write((0, node_util_1.format)(format, ...args) + '\n');
        }
    }
    static debug(format, ...args) {
        if (['trace', 'debug'].includes(this.config.outputLevel)) {
            stream_1.stdout.write((0, node_util_1.format)(format, ...args) + '\n');
        }
    }
    static info(format, ...args) {
        stream_1.stdout.write((0, node_util_1.format)(format, ...args) + '\n');
    }
    static log(format, ...args) {
        this.info(format || '', ...args);
    }
    static url(text, uri, params = {}) {
        const supports = require('supports-hyperlinks');
        if (supports.stdout) {
            this.log(hyperlinker(text, uri, params));
        }
        else {
            this.log(uri);
        }
    }
    static annotation(text, annotation) {
        const supports = require('supports-hyperlinks');
        if (supports.stdout) {
            // \u001b]8;;https://google.com\u0007sometext\u001b]8;;\u0007
            this.log(`\u001B]1337;AddAnnotation=${text.length}|${annotation}\u0007${text}`);
        }
        else {
            this.log(text);
        }
    }
    static async flush(ms = 10000) {
        await (0, flush_1.flush)(ms);
    }
}
exports.ux = ux;
const { action, annotation, anykey, confirm, debug, done, flush, info, log, progress, prompt, styledHeader, styledJSON, styledObject, table, trace, tree, url, wait, } = ux;
exports.action = action;
exports.annotation = annotation;
exports.anykey = anykey;
exports.confirm = confirm;
exports.debug = debug;
exports.done = done;
exports.flush = flush;
exports.info = info;
exports.log = log;
exports.progress = progress;
exports.prompt = prompt;
exports.styledHeader = styledHeader;
exports.styledJSON = styledJSON;
exports.styledObject = styledObject;
exports.table = table;
exports.trace = trace;
exports.tree = tree;
exports.url = url;
exports.wait = wait;
const { error, exit, warn } = Errors;
exports.error = error;
exports.exit = exit;
exports.warn = warn;
const uxProcessExitHandler = async () => {
    try {
        await ux.done();
    }
    catch (error) {
        console.error(error);
        process.exitCode = 1;
    }
};
// to avoid MaxListenersExceededWarning
// only attach named listener once
const uxListener = process.listeners('exit').find((fn) => fn.name === uxProcessExitHandler.name);
if (!uxListener) {
    process.once('exit', uxProcessExitHandler);
}
var exit_1 = require("./exit");
Object.defineProperty(exports, "ExitError", { enumerable: true, get: function () { return exit_1.ExitError; } });
var styled_1 = require("./styled");
Object.defineProperty(exports, "Table", { enumerable: true, get: function () { return styled_1.Table; } });
var base_1 = require("./action/base");
Object.defineProperty(exports, "ActionBase", { enumerable: true, get: function () { return base_1.ActionBase; } });
var config_2 = require("./config");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return config_2.config; } });
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return config_2.Config; } });
