// @mitre/saf/lib/commands/generate/inspec_metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const fs_1 = tslib_1.__importDefault(require("fs"));
const prompt_sync_1 = tslib_1.__importDefault(require("prompt-sync"));
const prompt = (0, prompt_sync_1.default)();
class GenerateInSpecMetadata extends core_1.Command {
    static usage = 'generate inspec_metadata -o <json-file>';
    static description = 'Generate an InSpec metadata template for "saf convert *2inspec_stub"';
    static examples = ['saf generate inspec_metadata -o ms_sql_baseline_metadata.json'];
    static flags = {
        help: core_1.Flags.help({ char: 'h' }),
        output: core_1.Flags.string({ char: 'o', required: true, description: 'Output JSON File' }),
    };
    async run() {
        const { flags } = await this.parse(GenerateInSpecMetadata);
        console.log('Please fill in the following fields to the best of your ability, if you do not have a value, please leave the field empty.');
        const inspecMetadata = {
            maintainer: prompt({ ask: 'Who is the maintainer? ' }) || null,
            copyright: prompt({ ask: 'Who is the copyright holder? ' }) || null,
            copyright_email: prompt({ ask: 'What is the email of the copyright holder? ' }) || null,
            license: prompt({ ask: 'What is the license of the profile? ' }),
            version: prompt({ ask: 'What is the version of the profile? ' }),
        };
        fs_1.default.writeFileSync(flags.output, JSON.stringify(inspecMetadata, null, 2));
    }
}
exports.default = GenerateInSpecMetadata;
