// @mitre/hdf-converters/lib/src/mappings/CweNistMappingItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CweNistMappingItem = void 0;
class CweNistMappingItem {
    constructor(values) {
        if (values['CWE-ID'] === undefined) {
            throw new Error('CWE Nist Mapping Data must contain an ID.');
        }
        else {
            this.id = values['CWE-ID'];
        }
        if (values['CWE Name'] === undefined) {
            throw new Error('CWE Nist Mapping Data must contain a name.');
        }
        else {
            this.name = values['CWE Name'].trim();
        }
        if (values['NIST-ID'] === undefined) {
            this.nistId = '';
        }
        else {
            this.nistId = values['NIST-ID'];
        }
        if (values.Rev === undefined) {
            throw new Error('CWE Nist Mapping Data must contain a rev.');
        }
        else {
            this.rev = values.Rev;
        }
        this.nistName = values['NIST Name'];
    }
}
exports.CweNistMappingItem = CweNistMappingItem;
