// @mitre/hdf-converters/lib/src/mappings/OwaspNistMappingItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OwaspNistMappingItem = void 0;
class OwaspNistMappingItem {
    constructor(values) {
        if (values['OWASP-ID'] === undefined) {
            throw new Error('OWASP Nist Mapping Data must contain an ID.');
        }
        else {
            this.id = values['OWASP-ID'];
        }
        if (values['OWASP Name'] === undefined) {
            throw new Error('OWASP Nist Mapping Data must contain a name.');
        }
        else {
            this.name = values['OWASP Name'];
        }
        if (values['NIST-ID'] === undefined) {
            this.nistId = '';
        }
        else {
            this.nistId = values['NIST-ID'];
        }
        this.rev = values['Rev'];
        if (values['NIST Name'] === undefined) {
            throw new Error('OWASP Nist Mapping Data must contain a nist name.');
        }
        else {
            this.nistName = values['NIST Name'];
        }
    }
}
exports.OwaspNistMappingItem = OwaspNistMappingItem;
