// @mitre/saf/lib/commands/emasser/post/cloud_resources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const json_colorizer_1 = tslib_1.__importDefault(require("json-colorizer"));
const core_1 = require("@oclif/core");
const emass_client_1 = require("@mitre/emass_client");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const utilities_1 = require("../../../utils/emasser/utilities");
const outputError_1 = require("../../../utils/emasser/outputError");
const promises_1 = require("fs/promises");
const fs_1 = tslib_1.__importDefault(require("fs"));
function printRedMsg(msg) {
    console.log('\x1B[91m', msg, '\x1B[0m');
}
function assertParamExists(object, value) {
    if (value === undefined) {
        printRedMsg(`Missing required parameter/field: ${object}`);
        throw new Error('Value not defined');
    }
}
function addRequiredFieldsToRequestBody(dataObj) {
    const bodyObj = {
        provider: '',
        resourceId: '',
        resourceName: '',
        resourceType: '',
        complianceResults: [],
    };
    const complianceResultsArray = [];
    try {
        assertParamExists('provider', dataObj.provider);
        assertParamExists('resourceId', dataObj.resourceId);
        assertParamExists('resourceName', dataObj.resourceName);
        assertParamExists('resourceType', dataObj.resourceType);
        let i = 0;
        dataObj.complianceResults.forEach((entryObject) => {
            assertParamExists(`dataObj.complianceResults[${i}].cspPolicyDefinitionId`, entryObject.cspPolicyDefinitionId);
            assertParamExists(`dataObj.complianceResults[${i}].isCompliant`, entryObject.isCompliant);
            assertParamExists(`dataObj.complianceResults[${i}].policyDefinitionTitle`, entryObject.policyDefinitionTitle);
            i++;
            const complianceResultsObj = {
                cspPolicyDefinitionId: '',
                isCompliant: false,
                policyDefinitionTitle: '',
            };
            complianceResultsObj.cspPolicyDefinitionId = entryObject.cspPolicyDefinitionId;
            complianceResultsObj.isCompliant = entryObject.isCompliant;
            complianceResultsObj.policyDefinitionTitle = entryObject.policyDefinitionTitle;
            complianceResultsArray.push(complianceResultsObj);
        });
    }
    catch (error) {
        console.log('Required JSON fields are:');
        console.log((0, json_colorizer_1.default)(JSON.stringify((0, utilities_1.getJsonExamples)('cloud_resources-required'), null, 2)));
        throw error;
    }
    bodyObj.provider = dataObj.provider;
    bodyObj.resourceId = dataObj.resourceId;
    bodyObj.resourceName = dataObj.resourceName;
    bodyObj.resourceType = dataObj.resourceType;
    bodyObj.complianceResults = complianceResultsArray;
    return bodyObj;
}
function addOptionalFields(bodyObject, dataObj) {
    // Add object optional entries
    if (Object.prototype.hasOwnProperty.call(dataObj, 'cspAccountId')) {
        bodyObject.cspAccountId = dataObj.cspAccountId;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'cspRegion')) {
        bodyObject.cspRegion = dataObj.cspRegion;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'initiatedBy')) {
        bodyObject.initiatedBy = dataObj.initiatedBy;
    }
    if (Object.prototype.hasOwnProperty.call(dataObj, 'isBaseline')) {
        bodyObject.isBaseline = dataObj.isBaseline;
    }
    // Add optional tags objects if available
    if (Object.prototype.hasOwnProperty.call(dataObj, 'tags')) {
        const tagsObj = {};
        Object.keys(dataObj.tags).forEach(key => {
            tagsObj[key.toString()] = dataObj.tags[key];
        });
        bodyObject.tags = tagsObj;
    }
    // Add optional compliance results fields
    const complianceResultsArray = [];
    dataObj.complianceResults.forEach((entryObject) => {
        const complianceResultsObj = {
            cspPolicyDefinitionId: '',
            isCompliant: false,
            policyDefinitionTitle: '',
        };
        complianceResultsObj.cspPolicyDefinitionId = entryObject.cspPolicyDefinitionId;
        complianceResultsObj.isCompliant = entryObject.isCompliant;
        complianceResultsObj.policyDefinitionTitle = entryObject.policyDefinitionTitle;
        if (Object.prototype.hasOwnProperty.call(entryObject, 'assessmentProcedure')) {
            complianceResultsObj.assessmentProcedure = entryObject.assessmentProcedure;
        }
        if (Object.prototype.hasOwnProperty.call(entryObject, 'complianceCheckTimestamp')) {
            complianceResultsObj.complianceCheckTimestamp = entryObject.complianceCheckTimestamp;
        }
        if (Object.prototype.hasOwnProperty.call(entryObject, 'complianceReason')) {
            complianceResultsObj.complianceReason = entryObject.complianceReason;
        }
        if (Object.prototype.hasOwnProperty.call(entryObject, 'control')) {
            complianceResultsObj.control = entryObject.control;
        }
        if (Object.prototype.hasOwnProperty.call(entryObject, 'policyDeploymentName')) {
            complianceResultsObj.policyDeploymentName = entryObject.policyDeploymentName;
        }
        if (Object.prototype.hasOwnProperty.call(entryObject, 'policyDeploymentVersion')) {
            complianceResultsObj.policyDeploymentVersion = entryObject.policyDeploymentVersion;
        }
        if (Object.prototype.hasOwnProperty.call(entryObject, 'severity')) {
            complianceResultsObj.severity = entryObject.severity;
        }
        complianceResultsArray.push(complianceResultsObj);
    });
    bodyObject.complianceResults = complianceResultsArray;
}
class EmasserPostCloudResources extends core_1.Command {
    static usage = '<%= command.id %> [ARGUMENTS]';
    static description = 'Add a cloud resource and their scan results in the assets module for a system';
    static examples = ['<%= config.bin %> <%= command.id %> [-s,--systemId] [-f,--cloudResourceFile]',
        'The input file should be a well formed JSON containing the cloud resources and their scan results information.',
        'Required JSON parameter/fields are: ',
        (0, json_colorizer_1.default)(JSON.stringify((0, utilities_1.getJsonExamples)('cloud_resources-required'), null, 2)),
        'Optional JSON parameters/fields are:',
        (0, json_colorizer_1.default)(JSON.stringify((0, utilities_1.getJsonExamples)('cloud_resources-optional'), null, 2))];
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Post (add) cloud resources and their scan results in the assets module for a system' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    async run() {
        const { flags } = await this.parse(EmasserPostCloudResources);
        const apiCxn = new apiConnection_1.ApiConnection();
        const addCloudResource = new emass_client_1.CloudResourcesApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        const requestBodyArray = [];
        // Check if a Cloud Resource json file was provided
        if (fs_1.default.existsSync(flags.cloudResourceFile)) {
            let data;
            try {
                data = JSON.parse(await (0, promises_1.readFile)(flags.cloudResourceFile, 'utf8'));
            }
            catch (error) {
                if (error.code === 'ENOENT') {
                    console.log('Cloud Resource JSON file not found!');
                    process.exit(1);
                }
                else {
                    console.log('Error reading Cloud Resource file, possible malformed json. Please use the -h flag for help.');
                    console.log('Error message was:', error.message);
                    process.exit(1);
                }
            }
            // Create request body based on key/pair values provide in the input file
            if (Array.isArray(data)) {
                data.forEach((dataObject) => {
                    let bodyObj = {
                        provider: '',
                        resourceId: '',
                        resourceName: '',
                        resourceType: '',
                        complianceResults: [],
                    };
                    // Add required fields to request array object based on business logic
                    try {
                        bodyObj = addRequiredFieldsToRequestBody(dataObject);
                        addOptionalFields(bodyObj, dataObject);
                        requestBodyArray.push(bodyObj);
                    }
                    catch {
                        process.exit(1);
                    }
                });
            }
            else if (typeof data === 'object') {
                const dataObject = data;
                let bodyObj = {
                    provider: '',
                    resourceId: '',
                    resourceName: '',
                    resourceType: '',
                    complianceResults: [],
                };
                // Add required fields to request array object based on business logic
                try {
                    bodyObj = addRequiredFieldsToRequestBody(dataObject);
                    addOptionalFields(bodyObj, dataObject);
                    requestBodyArray.push(bodyObj);
                }
                catch {
                    process.exit(1);
                }
            }
        }
        else {
            console.error('Invalid or Cloud Resource JSON file not found on the provided directory:', flags.cloudResourceFile);
            process.exit(1);
        }
        addCloudResource.addCloudResourcesBySystemId(flags.systemId, requestBodyArray).then((response) => {
            console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response, false)));
        }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
    }
}
exports.default = EmasserPostCloudResources;
