// @mitre/saf/lib/commands/emasser/post/pac.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const json_colorizer_1 = tslib_1.__importDefault(require("json-colorizer"));
const core_1 = require("@oclif/core");
const emass_client_1 = require("@mitre/emass_client");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const utilities_1 = require("../../../utils/emasser/utilities");
const outputError_1 = require("../../../utils/emasser/outputError");
class EmasserPostPac extends core_1.Command {
    static usage = '<%= command.id %> [ARGUMENTS]';
    static description = 'Add new Package Approval Chain (PAC) workflow(s) for a system';
    static examples = ['<%= config.bin %> <%= command.id %> [-s,--systemId] [-w,--workflow] [-n,--name] [-c,--comments]'];
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Post (add) a Package Approval Chain (PAC) item in a system' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    async run() {
        const { flags } = await this.parse(EmasserPostPac);
        const apiCxn = new apiConnection_1.ApiConnection();
        const addPac = new emass_client_1.PACApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        const requestBodyArray = [];
        requestBodyArray.push({
            workflow: flags.workflow,
            name: flags.name,
            comments: flags.comments,
        });
        addPac.addSystemPac(flags.systemId, requestBodyArray).then((response) => {
            console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response, false)));
        }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
    }
}
exports.default = EmasserPostPac;
