// @mitre/saf/lib/commands/emasser/post/test_results.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const json_colorizer_1 = tslib_1.__importDefault(require("json-colorizer"));
const core_1 = require("@oclif/core");
const emass_client_1 = require("@mitre/emass_client");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const utilities_1 = require("../../../utils/emasser/utilities");
const outputError_1 = require("../../../utils/emasser/outputError");
class EmasserPostTestResults extends core_1.Command {
    async run() {
        const { flags } = await this.parse(EmasserPostTestResults);
        const apiCxn = new apiConnection_1.ApiConnection();
        const addTestResults = new emass_client_1.TestResultsApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        const requestBodyArray = [];
        requestBodyArray.push({
            cci: flags.cci,
            testedBy: flags.testedBy,
            testDate: Number.parseFloat(flags.testDate),
            description: flags.description,
            complianceStatus: flags.complianceStatus,
        });
        addTestResults.addTestResultsBySystemId(flags.systemId, requestBodyArray).then((response) => {
            console.log((0, json_colorizer_1.default)((0, outputFormatter_1.outputFormat)(response, false)));
        }).catch((error) => console.error((0, json_colorizer_1.default)((0, outputError_1.outputError)(error))));
    }
}
exports.default = EmasserPostTestResults;
EmasserPostTestResults.usage = '<%= command.id %> [ARGUMENTS]';
EmasserPostTestResults.description = "Add test results for a system's Assessment Procedures (CCIs) which determine Security Control compliance";
EmasserPostTestResults.examples = ['<%= config.bin %> <%= command.id %> [-s,--systemId] [-c,--cci] [-b,--testedBy] [-t,--testDate] [-d,--description] [-S,--complianceStatus]'];
EmasserPostTestResults.flags = {
    help: core_1.Flags.help({ char: 'h', description: 'Post (add) test results to a system\'s Assessment Procedures (CCIs)' }),
    ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
};
