// @mitre/saf/lib/commands/emasser/post/test_results.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const json_colorizer_1 = require("json-colorizer");
const core_1 = require("@oclif/core");
const apiConnection_1 = require("../../../utils/emasser/apiConnection");
const outputFormatter_1 = require("../../../utils/emasser/outputFormatter");
const utilities_1 = require("../../../utils/emasser/utilities");
const emass_client_1 = require("@mitre/emass_client");
const CMD_HELP = 'saf emasser post test_results -h or --help';
class EmasserPostTestResults extends core_1.Command {
    static usage = '<%= command.id %> [FLAGS]';
    static description = "Add test results for a system's Assessment Procedures which determine Security Control compliance\n"
        + 'See the FLAGS section for required fields and acceptable values';
    static examples = ['<%= config.bin %> <%= command.id %> [-s,--systemId] [-a,--assessmentProcedure] [-b,--testedBy] [-t,--testDate] [-d,--description] [-S,--complianceStatus]'];
    static flags = {
        help: core_1.Flags.help({ char: 'h', description: 'Show eMASSer CLI help for the POST Test Results command' }),
        ...(0, utilities_1.getFlagsForEndpoint)(process.argv), // skipcq: JS-0349
    };
    async run() {
        const { flags } = await this.parse(EmasserPostTestResults);
        const apiCxn = new apiConnection_1.ApiConnection();
        const addTestResults = new emass_client_1.TestResultsApi(apiCxn.configuration, apiCxn.basePath, apiCxn.axiosInstances);
        const requestBodyArray = [
            {
                assessmentProcedure: flags.assessmentProcedure,
                testedBy: flags.testedBy,
                testDate: Number.parseFloat(flags.testDate),
                description: flags.description,
                complianceStatus: flags.complianceStatus,
            },
        ];
        addTestResults.addTestResultsBySystemId(flags.systemId, requestBodyArray).then((response) => {
            console.log((0, json_colorizer_1.colorize)((0, outputFormatter_1.outputFormat)(response, false)));
        }).catch((error) => (0, utilities_1.displayError)(error, 'Test Results'));
    }
    // skipcq: JS-0116 - Base class (CommandError) expects expected catch to return a Promise
    async catch(err) {
        // If error message is for missing flags, display
        // what fields are required, otherwise show the error
        if (err.message.includes('See more help with --help')) {
            this.warn(err.message.replace('with --help', `with: \x1B[93m${CMD_HELP}\x1B[0m`));
        }
        else {
            this.warn(err);
        }
    }
}
exports.default = EmasserPostTestResults;
