// @mitre/hdf-converters/lib/src/fortify-mapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FortifyMapper = void 0;
const inspecjs_1 = require("inspecjs");
const _ = __importStar(require("lodash"));
const package_json_1 = require("../package.json");
const base_converter_1 = require("./base-converter");
const global_1 = require("./utils/global");
const NIST_REFERENCE_NAME = 'Standards Mapping - NIST Special Publication 800-53 Revision 4';
const DEFAULT_NIST_TAG = [];
function impactMapping(input, id) {
    if (Array.isArray(input)) {
        const matches = input.find((element) => {
            return _.get(element, 'ClassInfo.ClassID') === id;
        });
        return parseFloat(_.get(matches, 'ClassInfo.DefaultSeverity')) / 5;
    }
    else {
        return parseFloat(_.get(input, 'ClassInfo.DefaultSeverity')) / 5;
    }
}
function nistTag(rule) {
    let references = _.get(rule, 'References.Reference');
    if (!Array.isArray(references)) {
        references = [references];
    }
    if (Array.isArray(references)) {
        const tag = references.find((element) => {
            return _.get(element, 'Author') === NIST_REFERENCE_NAME;
        });
        if (tag === null || tag === undefined) {
            return DEFAULT_NIST_TAG;
        }
        else {
            return _.get(tag, 'Title').match(/[a-zA-Z][a-zA-Z]-\d{1,2}/);
        }
    }
    return [];
}
function processEntry(input) {
    const output = [];
    output.push(`${_.get(input, 'id')}<=SNIPPET`);
    output.push(`\nPath: ${_.get(input, 'File')}\n`);
    output.push(`StartLine: ${_.get(input, 'StartLine')}, `);
    output.push(`EndLine: ${_.get(input, 'EndLine')}\n`);
    output.push(`Code:\n${_.get(input, 'Text').trim()}`);
    return output.join('');
}
function makeArray(input) {
    if (Array.isArray(input)) {
        return input;
    }
    else {
        return [input];
    }
}
function filterVuln(input, file) {
    input.forEach((element) => {
        if (element instanceof Object) {
            _.set(element, 'results', _.get(element, 'results').filter((result) => {
                const codedesc = _.get(result, 'code_desc').split('<=SNIPPET');
                const snippetid = codedesc[0];
                const classid = _.get(element, 'id');
                _.set(result, 'code_desc', codedesc[1]);
                let isMatch = false;
                const matches = _.get(file, 'FVDL.Vulnerabilities.Vulnerability').filter((subElement) => {
                    return _.get(subElement, 'ClassInfo.ClassID') === classid;
                });
                matches.forEach((match) => {
                    const traces = makeArray(_.get(match, 'AnalysisInfo.Unified.Trace'));
                    traces.forEach((trace) => {
                        const entries = makeArray(_.get(trace, 'Primary.Entry'));
                        const filteredEntries = entries.filter((entry) => {
                            return _.has(entry, 'Node.SourceLocation.snippet');
                        });
                        filteredEntries.forEach((entry) => {
                            if (_.get(entry, 'Node.SourceLocation.snippet') === snippetid) {
                                isMatch = true;
                            }
                        });
                    });
                });
                return isMatch;
            }));
            _.set(element, 'impact', impactMapping(_.get(element, 'impact'), _.get(element, 'id')));
        }
        return element;
    });
    return input;
}
class FortifyMapper extends base_converter_1.BaseConverter {
    constructor(fvdl, withRaw = false) {
        super((0, base_converter_1.parseXml)(fvdl, {
            stopNodes: ['FVDL.Description.Abstract', 'FVDL.Description.Explanation']
        }));
        this.mappings = {
            platform: {
                name: 'Heimdall Tools',
                release: package_json_1.version
            },
            version: package_json_1.version,
            statistics: {},
            profiles: [
                {
                    name: 'Fortify Static Analyzer Scan',
                    version: { path: 'FVDL.EngineData.EngineVersion' },
                    title: 'Fortify Static Analyzer Scan',
                    summary: {
                        path: 'FVDL.UUID',
                        transformer: (uuid) => {
                            return `Fortify Static Analyzer Scan of UUID: ${uuid}`;
                        }
                    },
                    supports: [],
                    attributes: [],
                    groups: [],
                    status: 'loaded',
                    controls: [
                        {
                            arrayTransformer: filterVuln,
                            path: 'FVDL.Description',
                            key: 'id',
                            tags: {
                                nist: { transformer: nistTag },
                                cci: {
                                    transformer: (data) => (0, global_1.getCCIsForNISTTags)(nistTag(data))
                                }
                            },
                            refs: [],
                            source_location: {},
                            title: { path: 'Abstract', transformer: base_converter_1.parseHtml },
                            id: { path: 'classID' },
                            desc: { path: 'Explanation', transformer: base_converter_1.parseHtml },
                            impact: { path: '$.FVDL.Vulnerabilities.Vulnerability' },
                            code: {
                                transformer: (vulnerability) => {
                                    return JSON.stringify(vulnerability, null, 2);
                                }
                            },
                            results: [
                                {
                                    path: '$.FVDL.Snippets.Snippet',
                                    status: inspecjs_1.ExecJSON.ControlResultStatus.Failed,
                                    code_desc: { transformer: processEntry },
                                    start_time: {
                                        path: '$.FVDL.CreatedTS',
                                        transformer: (input) => {
                                            return `${_.get(input, 'date')} ${_.get(input, 'time')}`;
                                        }
                                    }
                                }
                            ]
                        }
                    ],
                    sha256: ''
                }
            ],
            passthrough: {
                transformer: (data) => {
                    let auxData = _.get(data, 'FVDL');
                    if (_.isObject(auxData)) {
                        auxData = _.omit(auxData, [
                            'CreatedTS',
                            'UUID',
                            'Description',
                            'Snippets'
                        ]);
                    }
                    return {
                        auxiliary_data: [
                            {
                                name: 'Fortify',
                                data: { FVDL: auxData }
                            }
                        ],
                        ...(this.withRaw && { raw: data })
                    };
                }
            }
        };
        this.startTime = `${_.get(this.data, 'FVDL.CreatedTS.date')} ${_.get(this.data, 'FVDL.CreatedTS.time')}`;
        this.withRaw = withRaw;
    }
}
exports.FortifyMapper = FortifyMapper;
