// @mitre/hdf-converters/lib/src/gosec-mapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoSecMapper = void 0;
const inspecjs_1 = require("inspecjs");
const _ = __importStar(require("lodash"));
const package_json_1 = require("../package.json");
const base_converter_1 = require("./base-converter");
const CweNistMapping_1 = require("./mappings/CweNistMapping");
const CWE_NIST_MAPPING = new CweNistMapping_1.CweNistMapping();
const DEFAULT_NIST_TAG = ['SI-2', 'RA-5'];
function nistTag(input) {
    const cwe = [`${_.get(input, 'id')}`];
    return CWE_NIST_MAPPING.nistFilter(cwe, DEFAULT_NIST_TAG);
}
function formatMessage(input) {
    return `${_.get(input, 'file')}, line:${_.get(input, 'line')}, column:${_.get(input, 'column')}`;
}
class GoSecMapper extends base_converter_1.BaseConverter {
    constructor(gosecJson, withRaw = false) {
        super(JSON.parse(gosecJson));
        this.mappings = {
            passthrough: {
                transformer: (data) => {
                    return {
                        auxiliary_data: [
                            {
                                name: 'Gosec',
                                data: {
                                    'Golang errors': _.get(data, 'Golang errors')
                                }
                            }
                        ],
                        ...(this.withRaw && { raw: data })
                    };
                }
            },
            platform: {
                name: 'Heimdall Tools',
                release: package_json_1.version
            },
            version: package_json_1.version,
            statistics: {},
            profiles: [
                {
                    name: 'Gosec scanner',
                    title: 'gosec',
                    version: { path: 'GosecVersion' },
                    supports: [],
                    attributes: [],
                    groups: [],
                    status: 'loaded',
                    controls: [
                        {
                            path: 'Issues',
                            key: 'id',
                            tags: {
                                nist: {
                                    path: 'cwe',
                                    transformer: nistTag
                                },
                                cwe: { path: 'cwe' },
                                nosec: { path: 'nosec' },
                                suppressions: { path: 'suppressions' },
                                severity: { path: 'severity' },
                                confidence: { path: 'confidence' }
                            },
                            refs: [],
                            source_location: {},
                            title: { path: 'details' },
                            id: { path: 'rule_id' },
                            desc: '',
                            impact: 0.5,
                            results: [
                                {
                                    status: inspecjs_1.ExecJSON.ControlResultStatus.Failed,
                                    code_desc: { path: 'code' },
                                    message: { transformer: formatMessage },
                                    start_time: ''
                                }
                            ]
                        }
                    ],
                    sha256: ''
                }
            ]
        };
        this.withRaw = withRaw;
    }
}
exports.GoSecMapper = GoSecMapper;
