// @mitre/hdf-converters/lib/src/jfrog-xray-mapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JfrogXrayMapper = void 0;
const inspecjs_1 = require("inspecjs");
const lodash_1 = __importDefault(require("lodash"));
const package_json_1 = require("../package.json");
const base_converter_1 = require("./base-converter");
const CweNistMapping_1 = require("./mappings/CweNistMapping");
const global_1 = require("./utils/global");
const IMPACT_MAPPING = new Map([
    ['high', 0.7],
    ['medium', 0.5],
    ['low', 0.3]
]);
const CWE_PATH = 'component_versions.more_details.cves[0].cwe';
const CWE_NIST_MAPPING = new CweNistMapping_1.CweNistMapping();
function hashId(vulnerability) {
    if (lodash_1.default.get(vulnerability, 'id') === '') {
        return (0, base_converter_1.generateHash)(lodash_1.default.get(vulnerability, 'summary').toString(), 'md5');
    }
    else {
        return lodash_1.default.get(vulnerability, 'id');
    }
}
function formatDesc(vulnerability) {
    const text = [];
    if (lodash_1.default.has(vulnerability, 'description')) {
        text.push(lodash_1.default.get(vulnerability, 'description').toString());
    }
    if (lodash_1.default.has(vulnerability, 'cves')) {
        const re1 = /":/gi;
        const re2 = /,/gi;
        text.push(`cves: ${JSON.stringify(lodash_1.default.get(vulnerability, 'cves'))
            .replace(re1, '"=>')
            .replace(re2, ', ')}`);
    }
    return text.join('<br>');
}
function formatCodeDesc(vulnerability) {
    const codeDescArray = [];
    const re = /,/gi;
    if (lodash_1.default.has(vulnerability, 'source_comp_id')) {
        codeDescArray.push(`source_comp_id : ${lodash_1.default.get(vulnerability, 'source_comp_id')}`);
    }
    else {
        codeDescArray.push('source_comp_id : ');
    }
    if (lodash_1.default.has(vulnerability, 'component_versions.vulnerable_versions')) {
        codeDescArray.push(`vulnerable_versions : ${JSON.stringify(lodash_1.default.get(vulnerability, 'component_versions.vulnerable_versions'))}`);
    }
    else {
        codeDescArray.push('vulnerable_versions : ');
    }
    if (lodash_1.default.has(vulnerability, 'component_versions.fixed_versions')) {
        codeDescArray.push(`fixed_versions : ${JSON.stringify(lodash_1.default.get(vulnerability, 'component_versions.fixed_versions'))}`);
    }
    else {
        codeDescArray.push('fixed_versions : ');
    }
    if (lodash_1.default.has(vulnerability, 'issue_type')) {
        codeDescArray.push(`issue_type : ${lodash_1.default.get(vulnerability, 'issue_type')}`);
    }
    else {
        codeDescArray.push('issue_type : ');
    }
    if (lodash_1.default.has(vulnerability, 'provider')) {
        codeDescArray.push(`provider : ${lodash_1.default.get(vulnerability, 'provider')}`);
    }
    else {
        codeDescArray.push('provider : ');
    }
    return codeDescArray.join('\n').replace(re, ', ');
}
function nistTag(identifier) {
    const identifiers = [];
    if (Array.isArray(identifier)) {
        identifier.forEach((element) => {
            if (element.split('CWE-')[1]) {
                identifiers.push(element.split('CWE-')[1]);
            }
        });
    }
    return CWE_NIST_MAPPING.nistFilter(identifiers, global_1.DEFAULT_STATIC_CODE_ANALYSIS_NIST_TAGS);
}
class JfrogXrayMapper extends base_converter_1.BaseConverter {
    constructor(xrayJson, withRaw = false) {
        super(JSON.parse(xrayJson), true);
        this.mappings = {
            platform: {
                name: 'Heimdall Tools',
                release: package_json_1.version
            },
            version: package_json_1.version,
            statistics: {},
            profiles: [
                {
                    name: 'JFrog Xray Scan',
                    title: 'JFrog Xray Scan',
                    summary: 'Continuous Security and Universal Artifact Analysis',
                    supports: [],
                    attributes: [],
                    groups: [],
                    status: 'loaded',
                    controls: [
                        {
                            path: 'data',
                            key: 'id',
                            tags: {
                                cci: {
                                    path: CWE_PATH,
                                    transformer: (identifier) => (0, global_1.getCCIsForNISTTags)(nistTag(identifier))
                                },
                                nist: {
                                    path: CWE_PATH,
                                    transformer: nistTag
                                },
                                cweid: { path: CWE_PATH }
                            },
                            refs: [],
                            source_location: {},
                            id: { transformer: hashId },
                            title: { path: 'summary' },
                            desc: {
                                path: 'component_versions.more_details',
                                transformer: formatDesc
                            },
                            impact: {
                                path: 'severity',
                                transformer: (0, base_converter_1.impactMapping)(IMPACT_MAPPING)
                            },
                            code: {
                                transformer: (vulnerability) => {
                                    return JSON.stringify(vulnerability, null, 2);
                                }
                            },
                            results: [
                                {
                                    status: inspecjs_1.ExecJSON.ControlResultStatus.Failed,
                                    code_desc: { transformer: formatCodeDesc },
                                    start_time: ''
                                }
                            ]
                        }
                    ],
                    sha256: ''
                }
            ],
            passthrough: {
                transformer: (data) => {
                    return {
                        auxiliary_data: [
                            {
                                name: 'JFrog Xray',
                                data: lodash_1.default.pick(data, ['total_count'])
                            }
                        ],
                        ...(this.withRaw && { raw: data })
                    };
                }
            }
        };
        this.withRaw = withRaw;
    }
}
exports.JfrogXrayMapper = JfrogXrayMapper;
