// @mitre/hdf-converters/lib/src/nessus-mapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NessusMapper = exports.NessusResults = void 0;
const inspecjs_1 = require("inspecjs");
const lodash_1 = __importDefault(require("lodash"));
const package_json_1 = require("../package.json");
const base_converter_1 = require("./base-converter");
const CciNistMapping_1 = require("./mappings/CciNistMapping");
const NessusPluginsNistMapping_1 = require("./mappings/NessusPluginsNistMapping");
const IMPACT_MAPPING = new Map([
    ['4', 0.9],
    ['3', 0.7],
    ['i', 0.7],
    ['2', 0.5],
    ['ii', 0.5],
    ['1', 0.3],
    ['iii', 0.3],
    ['0', 0.0]
]);
const COMPLIANCE_PATH = 'cm:compliance-reference';
const NA_PLUGIN_OUTPUT = 'This Nessus Plugin does not provide output message.';
const NESSUS_PLUGINS_NIST_MAPPING = new NessusPluginsNistMapping_1.NessusPluginsNistMapping();
const CCI_NIST_MAPPING = new CciNistMapping_1.CciNistMapping();
const DEFAULT_NIST_TAG = [];
let policyName;
let version;
function getPolicyName() {
    return 'Nessus ' + policyName;
}
function getVersion() {
    return version;
}
function getId(item) {
    if (lodash_1.default.has(item, COMPLIANCE_PATH)) {
        return parseRef(lodash_1.default.get(item, COMPLIANCE_PATH), 'Vuln-ID')[0];
    }
    else {
        return lodash_1.default.get(item, 'pluginID');
    }
}
function getTitle(item) {
    if (lodash_1.default.has(item, 'cm:compliance-check-name')) {
        return lodash_1.default.get(item, 'cm:compliance-check-name');
    }
    else {
        return lodash_1.default.get(item, 'pluginName');
    }
}
function getDesc(item) {
    if (lodash_1.default.has(item, 'cm:compliance-info')) {
        return (0, base_converter_1.parseHtml)(lodash_1.default.get(item, 'cm:compliance-info'));
    }
    else {
        return (0, base_converter_1.parseHtml)(formatDesc(item));
    }
}
function formatDesc(issue) {
    const desc = [];
    desc.push(`Plugin Family: ${lodash_1.default.get(issue, 'pluginFamily')}`);
    desc.push(`Port: ${lodash_1.default.get(issue, 'port')}`);
    desc.push(`Protocol: ${lodash_1.default.get(issue, 'protocol')}`);
    return desc.join('; ') + ';';
}
function pluginNistTag(item) {
    const family = lodash_1.default.get(item, 'pluginFamily');
    const id = lodash_1.default.get(item, 'pluginID');
    return NESSUS_PLUGINS_NIST_MAPPING.nistFilter(family, id, DEFAULT_NIST_TAG);
}
function cciNistTag(input) {
    const identifiers = parseRef(input, 'CCI');
    return CCI_NIST_MAPPING.nistFilter(identifiers, DEFAULT_NIST_TAG, false);
}
function parseRef(input, key) {
    const matches = input.split(',').filter((element) => element.startsWith(key));
    return matches.map((element) => element.split('|')[1]);
}
function getImpact(item) {
    if (lodash_1.default.has(item, COMPLIANCE_PATH)) {
        return (0, base_converter_1.impactMapping)(IMPACT_MAPPING)(parseRef(lodash_1.default.get(item, COMPLIANCE_PATH), 'CAT').join(''));
    }
    else {
        return (0, base_converter_1.impactMapping)(IMPACT_MAPPING)(lodash_1.default.get(item, 'severity'));
    }
}
function getCheck(item) {
    if (lodash_1.default.has(item, 'cm:compliance-solution')) {
        return (0, base_converter_1.parseHtml)(lodash_1.default.get(item, 'cm:compliance-solution'));
    }
    else {
        return '';
    }
}
function getFix(item) {
    const fix = lodash_1.default.get(item, 'solution');
    if (fix && fix !== 'n/a') {
        return fix;
    }
    return '';
}
function getNist(item) {
    if (lodash_1.default.has(item, COMPLIANCE_PATH)) {
        return cciNistTag(lodash_1.default.get(item, COMPLIANCE_PATH));
    }
    else {
        return pluginNistTag(item);
    }
}
function getCci(item) {
    if (lodash_1.default.has(item, COMPLIANCE_PATH)) {
        return parseRef(lodash_1.default.get(item, COMPLIANCE_PATH), 'CCI');
    }
    else {
        return [];
    }
}
function getRid(item) {
    if (lodash_1.default.has(item, COMPLIANCE_PATH)) {
        return parseRef(lodash_1.default.get(item, COMPLIANCE_PATH), 'Rule-ID').join(',');
    }
    else {
        return lodash_1.default.get(item, 'pluginID');
    }
}
function getStig(item) {
    if (lodash_1.default.has(item, COMPLIANCE_PATH)) {
        return parseRef(lodash_1.default.get(item, COMPLIANCE_PATH), 'STIG-ID').join(',');
    }
    else {
        return '';
    }
}
function getStatus(item) {
    const result = lodash_1.default.get(item, 'cm:compliance-result');
    switch (result) {
        case 'PASSED':
            return inspecjs_1.ExecJSON.ControlResultStatus.Passed;
        case 'WARNING':
            return inspecjs_1.ExecJSON.ControlResultStatus.Skipped;
        case 'ERROR':
            return inspecjs_1.ExecJSON.ControlResultStatus.Error;
        default:
            return inspecjs_1.ExecJSON.ControlResultStatus.Failed;
    }
}
function formatCodeDesc(item) {
    if (lodash_1.default.has(item, 'description')) {
        return (0, base_converter_1.parseHtml)(lodash_1.default.get(item, 'description') || NA_PLUGIN_OUTPUT);
    }
    else {
        return (0, base_converter_1.parseHtml)(lodash_1.default.get(item, 'plugin_output') || NA_PLUGIN_OUTPUT);
    }
}
function getStartTime(tag) {
    if (Array.isArray(tag)) {
        return lodash_1.default.get(tag.find((element) => {
            return lodash_1.default.get(element, 'name') === 'HOST_START';
        }), 'text');
    }
    else {
        return lodash_1.default.get(tag, 'text');
    }
}
function cleanData(control) {
    const filteredControl = control;
    filteredControl.forEach((element) => {
        if (element instanceof Object) {
            if (lodash_1.default.get(element.tags, 'cci').length === 0) {
                element.tags = lodash_1.default.omit(element.tags, 'cci');
            }
            if (lodash_1.default.get(element.tags, 'rid') === '') {
                element.tags = lodash_1.default.omit(element.tags, 'rid');
            }
            if (lodash_1.default.get(element.tags, 'stig_id') === '') {
                element.tags = lodash_1.default.omit(element.tags, 'stig_id');
            }
            element.refs = element.refs.filter((ref) => ref.url);
            if (element.descriptions !== undefined && element.descriptions !== null) {
                element.descriptions = element.descriptions.filter((description) => description && description.data);
            }
        }
    });
    return filteredControl;
}
class NessusResults {
    constructor(nessusXml) {
        this.data = (0, base_converter_1.parseXml)(nessusXml);
    }
    toHdf() {
        const results = [];
        policyName = lodash_1.default.get(this.data, 'NessusClientData_v2.Policy.policyName');
        const preference = lodash_1.default.get(this.data, 'NessusClientData_v2.Policy.Preferences.ServerPreferences.preference');
        console.log(this.data);
        if (Array.isArray(preference)) {
            version =
                lodash_1.default.get(preference.find((element) => {
                    return lodash_1.default.get(element, 'name') === 'sc_version';
                }), 'value') || '';
        }
        const reportHost = lodash_1.default.get(this.data, 'NessusClientData_v2.Report.ReportHost');
        if (Array.isArray(reportHost)) {
            reportHost.forEach((element) => {
                const entry = new NessusMapper(element);
                if (this.customMapping !== undefined) {
                    entry.setMappings(this.customMapping);
                }
                results.push(entry.toHdf());
            });
            return results;
        }
        else {
            const result = new NessusMapper(reportHost);
            if (this.customMapping !== undefined) {
                result.setMappings(this.customMapping);
            }
            return result.toHdf();
        }
    }
    setMappings(customMapping) {
        this.customMapping = customMapping;
    }
}
exports.NessusResults = NessusResults;
class NessusMapper extends base_converter_1.BaseConverter {
    constructor(nessusJson) {
        super(nessusJson);
        this.mappings = {
            platform: {
                name: 'Heimdall Tools',
                release: package_json_1.version,
                target_id: { path: 'name' }
            },
            version: package_json_1.version,
            statistics: {
                duration: null
            },
            profiles: [
                {
                    name: { transformer: getPolicyName },
                    version: { transformer: getVersion },
                    title: { transformer: getPolicyName },
                    maintainer: null,
                    summary: { transformer: getPolicyName },
                    license: null,
                    copyright: null,
                    copyright_email: null,
                    supports: [],
                    attributes: [],
                    depends: [],
                    groups: [],
                    status: 'loaded',
                    controls: [
                        {
                            arrayTransformer: cleanData,
                            path: 'ReportItem',
                            key: 'id',
                            tags: {
                                nist: { transformer: getNist },
                                cci: { transformer: getCci },
                                rid: { transformer: getRid },
                                stig_id: { transformer: getStig },
                                risk_factor: { path: 'risk_factor' },
                                plugin_type: { path: 'plugin_type' },
                                plugin_publication_date: { path: 'plugin_publication_date' },
                                fname: { path: 'fname' },
                                cvss3_base_score: { path: 'cvss3_base_score' },
                                cvss_base_score: { path: 'cvss_base_score' }
                            },
                            descriptions: [
                                {
                                    data: { transformer: getCheck },
                                    label: 'check'
                                },
                                {
                                    data: { transformer: getFix },
                                    label: 'fix'
                                }
                            ],
                            refs: [
                                {
                                    url: {
                                        path: 'see_also'
                                    }
                                }
                            ],
                            source_location: {},
                            id: { transformer: getId },
                            title: { transformer: getTitle },
                            desc: { transformer: getDesc },
                            impact: { transformer: getImpact },
                            code: {
                                transformer: (reportItem) => JSON.stringify(reportItem, null, 2)
                            },
                            results: [
                                {
                                    status: { transformer: getStatus },
                                    code_desc: { transformer: formatCodeDesc },
                                    message: {
                                        path: ['plugin_output', 'cm:compliance-actual-value']
                                    },
                                    run_time: 0,
                                    start_time: {
                                        path: '$.HostProperties.tag',
                                        transformer: getStartTime
                                    }
                                }
                            ]
                        }
                    ],
                    sha256: ''
                }
            ]
        };
    }
}
exports.NessusMapper = NessusMapper;
