// @mitre/hdf-converters/lib/src/prisma-mapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrismaMapper = exports.PrismaControlMapper = exports.nistTag = void 0;
const inspecjs_1 = require("inspecjs");
const _ = __importStar(require("lodash"));
const package_json_1 = require("../package.json");
const base_converter_1 = require("./base-converter");
const global_1 = require("./utils/global");
const SEVERITY_LOOKUP = {
    low: 0.3,
    moderate: 0.5,
    high: 0.7,
    important: 0.9,
    critical: 1
};
function nistTag(cveTag) {
    if (!cveTag) {
        return global_1.DEFAULT_STATIC_CODE_ANALYSIS_NIST_TAGS;
    }
    else {
        return global_1.DEFAULT_UPDATE_REMEDIATION_NIST_TAGS;
    }
}
exports.nistTag = nistTag;
class PrismaControlMapper extends base_converter_1.BaseConverter {
    constructor(prismaControls) {
        super({ records: prismaControls });
        this.mappings = {
            platform: {
                name: 'Heimdall Tools',
                release: package_json_1.version,
                target_id: 'Prisma Cloud Scan Report'
            },
            version: package_json_1.version,
            statistics: {
                duration: null
            },
            profiles: [
                {
                    name: 'Palo Alto Prisma Cloud Tool',
                    version: '',
                    title: 'Prisma Cloud Scan Report',
                    maintainer: null,
                    summary: '',
                    license: null,
                    copyright: null,
                    copyright_email: null,
                    supports: [],
                    attributes: [],
                    depends: [],
                    groups: [],
                    status: 'loaded',
                    controls: [
                        {
                            path: 'records',
                            key: 'id',
                            desc: { path: 'Description' },
                            tags: {
                                cci: {
                                    path: 'CVE ID',
                                    transformer: (cve) => (0, global_1.getCCIsForNISTTags)(nistTag(cve))
                                },
                                nist: {
                                    path: 'CVE ID',
                                    transformer: nistTag
                                },
                                cve: { path: 'CVE ID' },
                                cvss: { path: 'cssv' }
                            },
                            descriptions: [],
                            refs: [{ url: { path: 'Vulnerability Link' } }],
                            source_location: { path: 'Hostname' },
                            id: {
                                transformer: (item) => {
                                    if (item['CVE ID']) {
                                        return `${item['Compliance ID']}-${item['CVE ID']}`;
                                    }
                                    else {
                                        return `${item['Compliance ID']}-${item.Distro}-${item.Severity}`;
                                    }
                                }
                            },
                            title: {
                                transformer: (item) => `${item.Hostname}-${item.Distro}-${item.Type}`
                            },
                            impact: {
                                path: 'Severity',
                                transformer: (severity) => {
                                    if (severity) {
                                        return SEVERITY_LOOKUP[severity];
                                    }
                                    else {
                                        return 0.5;
                                    }
                                }
                            },
                            code: {
                                transformer: (obj) => JSON.stringify(obj, null, 2)
                            },
                            results: [
                                {
                                    status: inspecjs_1.ExecJSON.ControlResultStatus.Failed,
                                    code_desc: {
                                        transformer: (obj) => {
                                            let result = '';
                                            if (obj.Type === 'image') {
                                                if (obj['Packages'] !== '') {
                                                    result += `Version check of package: ${obj['Packages']}`;
                                                }
                                            }
                                            else if (obj.Type === 'linux') {
                                                if (obj.Distro !== '') {
                                                    result += `Configuration check for ${obj.Distro}`;
                                                }
                                                else {
                                                    result += ``;
                                                }
                                            }
                                            else {
                                                result += `${obj.Type} check for ${obj.Hostname}`;
                                            }
                                            result += `\n\n${obj.Description}`;
                                            return result;
                                        }
                                    },
                                    message: {
                                        transformer: (obj) => {
                                            let result = '';
                                            if (obj['Fix Status'] !== '' && obj.Cause !== '') {
                                                result += `Fix Status: ${obj['Fix Status']}\n\n${obj.Cause}`;
                                            }
                                            else if (obj['Fix Status'] !== '') {
                                                result += `Fix Status: ${obj['Fix Status']}`;
                                            }
                                            else if (obj.Cause !== '') {
                                                result += `Cause: ${obj.Cause}`;
                                            }
                                            else {
                                                result += 'Unknown';
                                            }
                                            return result;
                                        }
                                    },
                                    start_time: { path: 'Published' }
                                }
                            ]
                        }
                    ],
                    sha256: ''
                }
            ]
        };
    }
}
exports.PrismaControlMapper = PrismaControlMapper;
class PrismaMapper {
    toHdf() {
        const executions = [];
        const hostnameToControls = {};
        this.data.forEach((record) => {
            hostnameToControls[record['Hostname']] =
                hostnameToControls[record['Hostname']] || [];
            hostnameToControls[record['Hostname']].push(record);
        });
        Object.entries(hostnameToControls).forEach(([hostname, controls]) => {
            const converted = new PrismaControlMapper(controls).toHdf();
            _.set(converted, 'platform.target_id', hostname);
            executions.push(converted);
        });
        return executions;
    }
    constructor(prismaCsv) {
        this.data = [];
        this.data = (0, base_converter_1.parseCsv)(prismaCsv);
    }
}
exports.PrismaMapper = PrismaMapper;
