// @mitre/hdf-converters/lib/src/scoutsuite-mapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScoutsuiteMapper = void 0;
const inspecjs_1 = require("inspecjs");
const lodash_1 = __importDefault(require("lodash"));
const package_json_1 = require("../package.json");
const base_converter_1 = require("./base-converter");
const ScoutsuiteNistMapping_1 = require("./mappings/ScoutsuiteNistMapping");
const INSPEC_INPUTS_MAPPING = {
    string: 'String',
    numeric: 'Numeric',
    regexp: 'Regexp',
    array: 'Array',
    hash: 'Hash',
    boolean: 'Boolean',
    any: 'Any'
};
const IMPACT_MAPPING = new Map([
    ['danger', 0.7],
    ['warning', 0.5]
]);
const SCOUTSUITE_NIST_MAPPING = new ScoutsuiteNistMapping_1.ScoutsuiteNistMapping();
function getRulesetName(file) {
    return lodash_1.default.get(file, 'last_run.ruleset_name');
}
function formatTargetId(file) {
    return `${getRulesetName(file)} ruleset:${lodash_1.default.get(file, 'provider_name')}:${lodash_1.default.get(file, 'account_id')}`;
}
function formatTitle(file) {
    return `Scout Suite Report using ${getRulesetName(file)} ruleset on ${lodash_1.default.get(file, 'provider_name')} with account ${lodash_1.default.get(file, 'account_id')}`;
}
function compliance(input) {
    if (Array.isArray(input)) {
        return input
            .map((element) => `Compliant with ${lodash_1.default.get(element, 'name')}, reference ${lodash_1.default.get(element, 'reference')}, version ${lodash_1.default.get(element, 'version')}`)
            .join('\n');
    }
    else {
        return '';
    }
}
function getStatus(input) {
    if (lodash_1.default.get(input, 'checked_items') === 0) {
        return inspecjs_1.ExecJSON.ControlResultStatus.Skipped;
    }
    else if (lodash_1.default.get(input, 'flagged_items') === 0) {
        return inspecjs_1.ExecJSON.ControlResultStatus.Passed;
    }
    else {
        return inspecjs_1.ExecJSON.ControlResultStatus.Failed;
    }
}
function checkSkip(input) {
    if (lodash_1.default.get(input, 'checked_items') === 0) {
        return 'Skipped because no items were checked';
    }
    else {
        return '';
    }
}
function getMessage(input) {
    if (lodash_1.default.get(input, 'checked_items') === 0) {
        return '';
    }
    else if (lodash_1.default.get(input, 'flagged_items') === 0) {
        return `0 flagged items out of ${lodash_1.default.get(input, 'checked_items')} checked items`;
    }
    else {
        return `${lodash_1.default.get(input, 'flagged_items')} flagged items out of ${lodash_1.default.get(input, 'checked_items')} checked items:\n${lodash_1.default.get(input, 'items').join('\n')}`;
    }
}
function nistTag(rule) {
    return SCOUTSUITE_NIST_MAPPING.nistTag(rule);
}
function checkArray(input) {
    if (typeof input === 'string') {
        return input;
    }
    return input.join(', ');
}
function collapseServices(file) {
    const services = Object.values(lodash_1.default.get(file, 'services'));
    const findings = [];
    services.forEach((element) => {
        findings.push(lodash_1.default.get(element, 'findings'));
    });
    const entries = [];
    Object.values(findings).forEach((element) => {
        Object.entries(element).forEach((subElement) => {
            entries.push(subElement);
        });
    });
    lodash_1.default.set(file, 'services', entries);
    return file;
}
class ScoutsuiteMapper extends base_converter_1.BaseConverter {
    constructor(scoutsuiteJson) {
        super(collapseServices(JSON.parse(scoutsuiteJson.split('\n', 2)[1])));
        this.mappings = {
            platform: {
                name: 'Heimdall Tools',
                release: package_json_1.version,
                target_id: { transformer: formatTargetId }
            },
            version: package_json_1.version,
            statistics: {
                duration: null
            },
            profiles: [
                {
                    name: 'Scout Suite Multi-Cloud Security Auditing Tool',
                    version: { path: 'last_run.version' },
                    title: { transformer: formatTitle },
                    maintainer: null,
                    summary: { path: 'last_run.ruleset_about' },
                    license: null,
                    copyright: null,
                    copyright_email: null,
                    supports: [],
                    attributes: [
                        {
                            name: 'account_id',
                            options: {
                                value: { path: 'account_id' },
                                required: true,
                                sensitive: false,
                                type: INSPEC_INPUTS_MAPPING.string
                            }
                        },
                        {
                            name: 'environment',
                            options: {
                                value: { path: 'environment' }
                            }
                        },
                        {
                            name: 'ruleset',
                            options: {
                                value: { path: 'last_run.ruleset_name' }
                            }
                        },
                        {
                            name: 'run_parameters_excluded_regions',
                            options: {
                                value: {
                                    path: 'last_run.run_parameters.excluded_region',
                                    transformer: checkArray
                                }
                            }
                        },
                        {
                            name: 'run_parameters_regions',
                            options: {
                                value: {
                                    path: 'last_run.run_parameters.regions',
                                    transformer: checkArray
                                }
                            }
                        },
                        {
                            name: 'run_parameters_services',
                            options: {
                                value: {
                                    path: 'last_run.run_parameters.services',
                                    transformer: checkArray
                                }
                            }
                        },
                        {
                            name: 'run_parameters_skipped_services',
                            options: {
                                value: {
                                    path: 'last_run.run_parameters.skipped_services',
                                    transformer: checkArray
                                }
                            }
                        },
                        {
                            name: 'time',
                            options: {
                                value: { path: 'last_run.time' }
                            }
                        },
                        {
                            name: 'partition',
                            options: {
                                value: { path: 'partition' }
                            }
                        },
                        {
                            name: 'provider_code',
                            options: {
                                value: { path: 'provider_code' }
                            }
                        },
                        {
                            name: 'provider_name',
                            options: {
                                value: { path: 'provider_name' }
                            }
                        }
                    ],
                    depends: [],
                    groups: [],
                    status: 'loaded',
                    controls: [
                        {
                            path: 'services',
                            key: 'id',
                            id: { path: '[0]' },
                            title: { path: '[1].description' },
                            tags: {
                                nist: { path: '[0]', transformer: nistTag }
                            },
                            impact: {
                                path: '[1].level',
                                transformer: (0, base_converter_1.impactMapping)(IMPACT_MAPPING)
                            },
                            desc: { path: '[1].rationale' },
                            descriptions: [
                                { data: { path: '[1].remediation' }, label: 'fix' },
                                { data: { path: '[1].service' }, label: 'service' },
                                { data: { path: '[1].path' }, label: 'path' },
                                { data: { path: '[1].id_suffix' }, label: 'id_suffix' }
                            ],
                            refs: [
                                { url: { path: '[1].references[0]' } },
                                { ref: { path: '[1].compliance', transformer: compliance } }
                            ],
                            source_location: {},
                            code: '',
                            results: [
                                {
                                    status: { path: '[1]', transformer: getStatus },
                                    skip_message: { path: '[1]', transformer: checkSkip },
                                    message: { path: '[1]', transformer: getMessage },
                                    code_desc: { path: '[1].description' },
                                    start_time: { path: '$.last_run.time' }
                                }
                            ]
                        }
                    ],
                    sha256: ''
                }
            ]
        };
    }
    setMappings(customMappings) {
        super.setMappings(customMappings);
    }
}
exports.ScoutsuiteMapper = ScoutsuiteMapper;
