// @mitre/hdf-converters/lib/src/snyk-mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnykMapper = exports.SnykResults = void 0;
const inspecjs_1 = require("inspecjs");
const package_json_1 = require("../package.json");
const base_converter_1 = require("./base-converter");
const CweNistMapping_1 = require("./mappings/CweNistMapping");
const IMPACT_MAPPING = new Map([
    ['high', 0.7],
    ['medium', 0.5],
    ['low', 0.3]
]);
const CWE_NIST_MAPPING = new CweNistMapping_1.CweNistMapping();
const DEFAULT_NIST_TAG = ['SA-11', 'RA-5'];
function parseIdentifier(identifiers) {
    const output = [];
    if (identifiers !== undefined && Array.isArray(identifiers)) {
        identifiers.forEach((element) => {
            const numbers = element.split('-');
            numbers.shift();
            output.push(numbers.join('-'));
        });
        return output;
    }
    else {
        return [];
    }
}
function nistTag(identifiers) {
    return CWE_NIST_MAPPING.nistFilter(parseIdentifier(identifiers), DEFAULT_NIST_TAG);
}
class SnykResults {
    constructor(snykJson) {
        this.data = JSON.parse(snykJson);
    }
    toHdf() {
        const results = [];
        if (Array.isArray(this.data)) {
            this.data.forEach((element) => {
                const entry = new SnykMapper(element);
                if (this.customMapping !== undefined) {
                    entry.setMappings(this.customMapping);
                }
                results.push(entry.toHdf());
            });
            return results;
        }
        else {
            const result = new SnykMapper(this.data);
            if (this.customMapping !== undefined) {
                result.setMappings(this.customMapping);
            }
            return result.toHdf();
        }
    }
    setMappings(customMapping) {
        this.customMapping = customMapping;
    }
}
exports.SnykResults = SnykResults;
class SnykMapper extends base_converter_1.BaseConverter {
    constructor(snykJson) {
        super(snykJson);
        this.mappings = {
            platform: {
                name: 'Heimdall Tools',
                release: package_json_1.version,
                target_id: { path: 'projectName' }
            },
            version: package_json_1.version,
            statistics: {
                duration: null
            },
            profiles: [
                {
                    name: { path: 'policy' },
                    version: {
                        path: 'policy',
                        transformer: (policy) => {
                            if (typeof policy === 'string') {
                                return policy.split('version: ')[1].split('\n')[0];
                            }
                            else {
                                return '';
                            }
                        }
                    },
                    title: {
                        path: 'projectName',
                        transformer: (projectName) => {
                            return `Snyk Project: ${projectName}`;
                        }
                    },
                    maintainer: null,
                    summary: {
                        path: 'summary',
                        transformer: (summary) => {
                            return `Snyk Summary: ${summary}`;
                        }
                    },
                    license: null,
                    copyright: null,
                    copyright_email: null,
                    supports: [],
                    attributes: [],
                    depends: [],
                    groups: [],
                    status: 'loaded',
                    controls: [
                        {
                            path: 'vulnerabilities',
                            key: 'id',
                            tags: {
                                nist: { path: 'identifiers.CWE', transformer: nistTag },
                                cweid: { path: 'identifiers.CWE', transformer: parseIdentifier },
                                cveid: { path: 'identifiers.CVE', transformer: parseIdentifier },
                                ghsaid: { path: 'identifiers.GHSA', transformer: parseIdentifier }
                            },
                            descriptions: [],
                            refs: [],
                            source_location: {},
                            title: { path: 'title' },
                            id: { path: 'id' },
                            desc: { path: 'description' },
                            impact: {
                                path: 'severity',
                                transformer: (0, base_converter_1.impactMapping)(IMPACT_MAPPING)
                            },
                            code: '',
                            results: [
                                {
                                    status: inspecjs_1.ExecJSON.ControlResultStatus.Failed,
                                    code_desc: {
                                        path: 'from',
                                        transformer: (input) => {
                                            if (Array.isArray(input)) {
                                                return `From : [ ${input.join(' , ')} ]`;
                                            }
                                            else {
                                                return '';
                                            }
                                        }
                                    },
                                    run_time: 0,
                                    start_time: ''
                                }
                            ]
                        }
                    ],
                    sha256: ''
                }
            ]
        };
    }
    setMappings(customMappings) {
        super.setMappings(customMappings);
    }
}
exports.SnykMapper = SnykMapper;
