// @mitre/hdf-converters/lib/src/twistlock-mapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TwistlockMapper = exports.TwistlockResults = void 0;
const inspecjs_1 = require("inspecjs");
const _ = __importStar(require("lodash"));
const package_json_1 = require("../package.json");
const base_converter_1 = require("./base-converter");
const global_1 = require("./utils/global");
const IMPACT_MAPPING = new Map([
    ['critical', 0.9],
    ['important', 0.9],
    ['high', 0.7],
    ['medium', 0.5],
    ['moderate', 0.5],
    ['low', 0.3]
]);
class TwistlockResults {
    constructor(twistlockJson, withRaw = false) {
        this.data = JSON.parse(twistlockJson);
        this.withRaw = withRaw;
        if (!_.has(this.data, 'results')) {
            this.data = { results: [this.data] };
        }
    }
    toHdf() {
        return new TwistlockMapper(this.data, this.withRaw).toHdf();
    }
}
exports.TwistlockResults = TwistlockResults;
class TwistlockMapper extends base_converter_1.BaseConverter {
    constructor(twistlockJson, withRaw = false) {
        super(twistlockJson, true);
        this.mappings = {
            platform: {
                name: 'Heimdall Tools',
                release: package_json_1.version,
                target_id: { path: ['results[0].name', 'results[0].repository'] }
            },
            version: package_json_1.version,
            statistics: {},
            profiles: [
                {
                    path: 'results',
                    name: 'Twistlock Scan',
                    title: {
                        transformer: (data) => {
                            let projectArr = 'N/A';
                            if (_.has(data, 'collections')) {
                                projectArr = _.get(data, 'collections');
                            }
                            if (_.has(data, 'repository')) {
                                projectArr = _.get(data, 'repository');
                            }
                            const projectName = Array.isArray(projectArr)
                                ? projectArr.join(' / ')
                                : projectArr;
                            return `Twistlock Project: ${projectName}`;
                        }
                    },
                    summary: {
                        transformer: (data) => {
                            const vulnerabilityTotal = _.has(data, 'vulnerabilityDistribution')
                                ? `${JSON.stringify(_.get(data, 'vulnerabilityDistribution.total'))}`
                                : 'N/A';
                            const complianceTotal = _.has(data, 'complianceDistribution')
                                ? `${JSON.stringify(_.get(data, 'complianceDistribution.total'))}`
                                : 'N/A';
                            return `Package Vulnerability Summary: ${vulnerabilityTotal} Application Compliance Issue Total: ${complianceTotal}`;
                        }
                    },
                    supports: [],
                    attributes: [],
                    groups: [],
                    status: 'loaded',
                    controls: [
                        {
                            path: 'vulnerabilities',
                            key: 'id',
                            pathTransform: (value) => (Array.isArray(value) ? value : []),
                            tags: {
                                nist: global_1.DEFAULT_UPDATE_REMEDIATION_NIST_TAGS,
                                cci: (0, global_1.getCCIsForNISTTags)(global_1.DEFAULT_UPDATE_REMEDIATION_NIST_TAGS),
                                cveid: { path: 'id' }
                            },
                            refs: [],
                            source_location: {},
                            title: { path: 'id' },
                            id: { path: 'id' },
                            desc: { path: 'description' },
                            impact: {
                                path: 'severity',
                                transformer: (0, base_converter_1.impactMapping)(IMPACT_MAPPING)
                            },
                            code: {
                                transformer: (vulnerability) => JSON.stringify(vulnerability, null, 2)
                            },
                            results: [
                                {
                                    status: inspecjs_1.ExecJSON.ControlResultStatus.Failed,
                                    code_desc: {
                                        transformer: (data) => {
                                            const packageName = _.has(data, 'packageName')
                                                ? `${JSON.stringify(_.get(data, 'packageName'))}`
                                                : 'N/A';
                                            const impactedVersions = _.has(data, 'impactedVersions')
                                                ? `${JSON.stringify(_.get(data, 'impactedVersions'))}`
                                                : 'N/A';
                                            return `Package ${packageName} should be updated to latest version above impacted versions ${impactedVersions}`;
                                        }
                                    },
                                    message: {
                                        transformer: (data) => {
                                            const packageName = _.has(data, 'packageName')
                                                ? `${JSON.stringify(_.get(data, 'packageName'))}`
                                                : 'N/A';
                                            const packageVersion = _.has(data, 'packageVersion')
                                                ? `${JSON.stringify(_.get(data, 'packageVersion'))}`
                                                : 'N/A';
                                            return `Expected latest version of ${packageName}\nDetected vulnerable version ${packageVersion} of ${packageName}`;
                                        }
                                    },
                                    start_time: { path: 'discoveredDate' }
                                }
                            ]
                        }
                    ],
                    sha256: ''
                }
            ],
            passthrough: {
                transformer: (data) => {
                    let resultsData = _.get(data, 'results');
                    if (Array.isArray(resultsData)) {
                        resultsData = resultsData.map((result) => _.omit(result, [
                            'name',
                            'collections',
                            'complianceDistribution',
                            'vulnerabilities',
                            'vulnerabilityDistribution'
                        ]));
                    }
                    return {
                        auxiliary_data: [
                            {
                                name: 'Twistlock',
                                data: {
                                    results: resultsData,
                                    consoleURL: _.get(data, 'consoleURL')
                                }
                            }
                        ],
                        ...(this.withRaw && { raw: data })
                    };
                }
            }
        };
        this.withRaw = withRaw;
    }
}
exports.TwistlockMapper = TwistlockMapper;
