// @mitre/hdf-converters/lib/src/veracode-mapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VeracodeMapper = void 0;
const inspecjs_1 = require("inspecjs");
const _ = __importStar(require("lodash"));
const package_json_1 = require("../package.json");
const base_converter_1 = require("./base-converter");
const CweNistMapping_1 = require("./mappings/CweNistMapping");
const global_1 = require("./utils/global");
const STATIC_FLAWS = 'staticflaws.flaw';
const SEVERITY = 'detailedreport.severity';
const FILE_PATH_VALUE = 'file_paths.file_path.@_.value';
const CWE_NIST_MAPPING = new CweNistMapping_1.CweNistMapping();
const DEFAULT_NIST_TAG = ['SI-2', 'RA-5'];
const IMPACT_MAPPING = new Map([
    ['5', 0.9],
    ['4', 0.7],
    ['3', 0.5],
    ['2', 0.3],
    ['1', 0.1],
    ['0', 0.0]
]);
function impactMapping(severity) {
    return IMPACT_MAPPING.get(severity.toString()) || 0.1;
}
function nistTag(input) {
    const cwes = [];
    let cwe = _.get(input, 'cwe');
    if (!Array.isArray(cwe)) {
        cwe = [cwe];
    }
    cwes.push(...cwe.map((value) => _.get(value, '@_.cweid')));
    return CWE_NIST_MAPPING.nistFilter(cwes, DEFAULT_NIST_TAG);
}
function formatRecommendations(input) {
    const text = [];
    if (_.has(input, 'recommendations.para')) {
        if (_.has(input, 'recommendations.para.@_.text')) {
            text.push(`${_.get(input, 'recommendations.para.@_.text')}`);
        }
        else {
            text.push(..._.get(input, `recommendations.para`).map((value) => _.get(value, '@_.text')));
        }
    }
    if (_.has(input, 'recommendations.para.bulletitem')) {
        if (Array.isArray(_.get(input, `recommendations.para.bulletitem`))) {
            text.push(..._.get(input, `recommendations.para.bulletitem`).map((value) => _.get(value, '@_.text')));
        }
    }
    return text.join('\n');
}
function formatDesc(input) {
    const text = [];
    if (_.has(input, 'desc.para')) {
        if (_.has(input, 'desc.para.@_.text')) {
            text.push(`${_.get(input, 'desc.para.@_.text')}`);
        }
        else {
            text.push(..._.get(input, `desc.para`).map((value) => _.get(value, '@_.text')));
        }
    }
    return text.join('\n');
}
function formatCweData(input) {
    const text = [];
    const categories = [
        'pcrirelated',
        'owasp',
        'owasp2013',
        'sans',
        'certc',
        'certccp',
        'certjava',
        'owaspmobile'
    ];
    if (_.has(input, 'cwe')) {
        let cweInput = _.get(input, 'cwe');
        if (!Array.isArray(cweInput)) {
            cweInput = [cweInput];
        }
        text.push(...cweInput.map((cweinfo) => {
            let cwe = `CWE-${_.get(cweinfo, '@_.cweid')}: `;
            cwe += `${_.get(cweinfo, '@_.cwename')}`;
            cwe += categories
                .map((value) => {
                if (_.has(cweinfo, `@_.${value}`)) {
                    const val = _.get(cweinfo, `@_.${value}`);
                    return `${value}: ${val}\n`;
                }
                else {
                    return '';
                }
            })
                .join('');
            return cwe;
        }));
    }
    return text.join('\n');
}
function formatCweDesc(input) {
    const text = [];
    if (_.has(input, 'cwe')) {
        let cwe = _.get(input, 'cwe');
        if (!Array.isArray(cwe)) {
            cwe = [cwe];
        }
        text.push(...cwe.map((value) => `CWE-${_.get(value, '@_.cweid')}: ${_.get(value, '@_.cwename')} Description: ${_.get(value, 'description.text.@_.text')}; `));
    }
    return text.join('\n');
}
function getFlaws(input) {
    const flawArr = [];
    if (!Array.isArray(input)) {
        input = [input];
    }
    for (const value of input) {
        let staticFlaw = _.get(value, STATIC_FLAWS);
        if (!Array.isArray(staticFlaw)) {
            staticFlaw = [staticFlaw];
        }
        flawArr.push(...staticFlaw);
    }
    return flawArr;
}
function formatCodeDesc(input) {
    let flawDesc = '';
    const categories = [
        ['Line Number', 'line'],
        ['Affect Policy Compliance', 'affects_policy_compliance'],
        ['Remediation Effort', 'remediationeffort'],
        ['Exploit level', 'exploitLevel'],
        ['Issue ID', 'issueid'],
        ['Module', 'module'],
        ['Type', 'type'],
        ['CWE ID', 'cweid'],
        ['Date First Occurence', 'date_first_occurence'],
        ['CIA Impact', 'cia_impact'],
        ['Description', 'description'],
        ['Source File', 'sourcefile'],
        ['Scope', 'scope'],
        ['CIA Impact', 'cia_impact'],
        ['PCI Related', 'pcirelated'],
        ['Function Prototype', 'functionprototype'],
        ['Function Relative Location', 'functionrelativelocation']
    ];
    if (_.has(input, '@_.sourcefilepath')) {
        flawDesc = `Sourcefile Path: ${_.get(input, '@_.sourcefilepath')}\n`;
        flawDesc += categories
            .map(([title, name]) => {
            if (_.has(input, `@_.${name}`)) {
                const nameVal = _.get(input, `@_.${name}`);
                return `${title}: ${nameVal}\n`;
            }
            else {
                return '';
            }
        })
            .join('');
    }
    return flawDesc;
}
function formatSCACodeDesc(input) {
    let flawDesc = '';
    const categories = [
        'sha1',
        'file_name',
        'max_cvss_score',
        'version',
        'library',
        'library_id',
        'vendor',
        'description',
        'added_date',
        'component_affects_policy_compliance'
    ];
    if (_.has(input, '@_.component_id')) {
        flawDesc = `component_id: ${_.get(input, '@_.component_id')}\n`;
        flawDesc += _.compact(categories.map((value) => {
            if (_.has(input, `@_.${value}`)) {
                const val = _.get(input, `@_.${value}`);
                return `${value}: ${val}`;
            }
            else {
                return '';
            }
        })).join('\n');
        if (_.has(input, FILE_PATH_VALUE)) {
            flawDesc += `\nfile_path: ${_.get(input, FILE_PATH_VALUE)}\n`;
        }
    }
    return flawDesc;
}
function formatSourceLocation(input) {
    const flawArr = [];
    if (!Array.isArray(input)) {
        input = [input];
    }
    for (const value of input) {
        if (!Array.isArray(_.get(value, STATIC_FLAWS))) {
            flawArr.push(_.get(value, STATIC_FLAWS));
        }
        else {
            flawArr.push(..._.get(value, STATIC_FLAWS));
        }
    }
    return flawArr.map((value) => _.get(value, '@_.sourcefile')).join('\n');
}
function componentListCreate(input) {
    const componentList = [];
    let component = _.get(input, 'component');
    if (!Array.isArray(component)) {
        component = [component];
    }
    for (const value of component) {
        if (_.get(value, '@_.vulnerabilities') !== '0') {
            componentList.push(value);
        }
    }
    return componentList;
}
function componentTransform(input) {
    const componentList = componentListCreate(input);
    const vulns = componentList
        .map((component) => {
        let vulnerability = _.get(component, 'vulnerabilities.vulnerability');
        if (!Array.isArray(vulnerability)) {
            vulnerability = [vulnerability];
        }
        vulnerability = vulnerability.map((vuln) => ({
            ...vuln,
            components: [component]
        }));
        return vulnerability;
    })
        .flat()
        .reduce((acc, cur) => {
        const cveId = _.get(cur, '@_.cve_id');
        const index = acc.findIndex((vuln) => cveId === _.get(vuln, '@_.cve_id'));
        if (index === -1) {
            return [...acc, cur];
        }
        else {
            _.get(acc[index], 'components').push(..._.get(cur, 'components'));
            return acc;
        }
    }, []);
    return vulns;
}
function controlMappingCve() {
    return {
        id: { path: '@_.cve_id' },
        title: { path: '@_.cve_id' },
        desc: { path: '@_.cve_summary' },
        impact: { path: '@_.severity', transformer: impactMapping },
        refs: [],
        tags: {
            cwe: { path: '@_.cwe_id' },
            nist: {
                path: '@_.cwe_id',
                transformer: (value) => {
                    if (!Array.isArray(value)) {
                        value = [value];
                    }
                    return CWE_NIST_MAPPING.nistFilter(value.map((val) => val.substring(4)), DEFAULT_NIST_TAG);
                }
            }
        },
        source_location: {
            ref: {
                transformer: (vuln) => _.get(vuln, 'components')
                    .map((comp) => _.get(comp, FILE_PATH_VALUE))
                    .join('\n')
            }
        },
        results: [
            {
                path: 'components',
                status: inspecjs_1.ExecJSON.ControlResultStatus.Failed,
                code_desc: { transformer: formatSCACodeDesc },
                start_time: { path: '$.detailedreport.@_.first_build_submitted_date' }
            }
        ]
    };
}
function controlMappingCwe(severity) {
    return {
        id: { path: '@_.categoryid' },
        title: { path: '@_.categoryname' },
        desc: { transformer: formatDesc },
        descriptions: [
            {
                data: { transformer: formatRecommendations },
                label: 'fix'
            }
        ],
        impact: impactMapping(severity),
        refs: [],
        tags: {
            cweid: { transformer: formatCweData },
            cweDescription: { transformer: formatCweDesc },
            cci: {
                transformer: (data) => (0, global_1.getCCIsForNISTTags)(nistTag(data))
            },
            nist: { transformer: nistTag }
        },
        source_location: {
            ref: {
                path: 'cwe',
                transformer: formatSourceLocation
            }
        },
        results: [
            {
                path: 'cwe',
                pathTransform: getFlaws,
                status: inspecjs_1.ExecJSON.ControlResultStatus.Failed,
                code_desc: { transformer: formatCodeDesc },
                start_time: { path: '$.detailedreport.@_.first_build_submitted_date' },
                message: {
                    path: 'exploitability_adjustments.exploitability_adjustment.note'
                }
            }
        ]
    };
}
function componentPass(component) {
    const vulnList = [];
    _.set(component, 'control_ids', vulnList);
    if (_.get(component, 'vulnerabilities') !== '') {
        if (!Array.isArray(_.get(component, 'vulnerabilities.vulnerability'))) {
            vulnList.push(_.get(component, 'vulnerabilities.vulnerability.@_.cve_id'));
            _.set(component, 'control_ids', vulnList);
        }
        else {
            vulnList.push(..._.get(component, 'vulnerabilities.vulnerability').map((vuln) => _.get(vuln, '@_.cve_id')));
            _.set(component, 'control_ids', vulnList);
        }
    }
    return _.omit(component, 'vulnerabilities');
}
class VeracodeMapper extends base_converter_1.BaseConverter {
    constructor(xml, withRaw = false) {
        const parsedXML = (0, base_converter_1.parseXml)(xml, {
            attributesGroupName: '@_',
            textNodeName: 'text_'
        });
        if (_.has(parsedXML, 'summaryreport')) {
            throw new Error('Current mapper does not accept summary reports');
        }
        const arrayedControls = _.get(parsedXML, SEVERITY).map((control) => {
            if (Array.isArray(control.category)) {
                return { level: control.level, category: control.category };
            }
            else if (!control.category) {
                return { level: control.level };
            }
            else {
                return { level: control.level, category: [control.category] };
            }
        });
        _.set(parsedXML, SEVERITY, arrayedControls);
        super(parsedXML);
        this.originalData = xml;
        this.setMappings(this.defaultMapping(withRaw));
    }
    defaultMapping(withRaw = false) {
        return {
            platform: {
                name: 'Heimdall Tools',
                release: package_json_1.version
            },
            passthrough: {
                components: {
                    path: 'detailedreport.software_composition_analysis.vulnerable_components',
                    transformer: (value) => {
                        if (_.get(value, 'component')) {
                            return _.get(value, 'component').map((component) => componentPass(component));
                        }
                        else {
                            return '';
                        }
                    }
                },
                auxiliary_data: [
                    {
                        name: 'veracode',
                        data: {
                            transformer: (value) => _.omit(value, SEVERITY, 'detailedreport.software_composition_analysis')
                        }
                    }
                ],
                ...(withRaw && { raw: this.originalData })
            },
            version: package_json_1.version,
            statistics: {},
            profiles: [
                {
                    name: { path: 'detailedreport.@_.policy_name' },
                    version: { path: 'detailedreport.@_.policy_version' },
                    title: {
                        path: 'detailedreport.static-analysis.modules.module.@_.name'
                    },
                    supports: [],
                    attributes: [],
                    groups: [],
                    status: 'loaded',
                    controls: [
                        ..._.range(0, 6).map((value) => ({
                            path: `detailedreport.severity[${value}].category`,
                            ...controlMappingCwe(5 - value)
                        })),
                        {
                            path: 'detailedreport.software_composition_analysis.vulnerable_components',
                            pathTransform: componentTransform,
                            ...controlMappingCve()
                        }
                    ],
                    sha256: ''
                }
            ]
        };
    }
}
exports.VeracodeMapper = VeracodeMapper;
