// @mitre/hdf-converters/lib/src/zap-mapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZapMapper = void 0;
const inspecjs_1 = require("inspecjs");
const _ = __importStar(require("lodash"));
const package_json_1 = require("../package.json");
const base_converter_1 = require("./base-converter");
const CweNistMapping_1 = require("./mappings/CweNistMapping");
const global_1 = require("./utils/global");
const CWE_NIST_MAPPING = new CweNistMapping_1.CweNistMapping();
function filterSite(input, name) {
    if (name) {
        return input.find((element) => _.get(element, '@name') === name);
    }
    else {
        return input.reduce((a, b) => _.get(a, 'alerts').length >
            _.get(b, 'alerts').length
            ? a
            : b);
    }
}
function impactMapping(input) {
    if (typeof input === 'string') {
        const impact = parseInt(input);
        if (0 <= impact && impact <= 1) {
            return 0.3;
        }
        else if (impact === 2) {
            return 0.5;
        }
        else if (impact >= 3) {
            return 0.7;
        }
        else {
            return 0;
        }
    }
    else {
        return 0;
    }
}
function nistTag(cweid) {
    return CWE_NIST_MAPPING.nistFilter([cweid], global_1.DEFAULT_STATIC_CODE_ANALYSIS_NIST_TAGS);
}
function checkText(input) {
    const text = [];
    text.push(_.get(input, 'solution'));
    text.push(_.get(input, 'otherinfo'));
    text.push(_.get(input, 'otherinfo'));
    return text.join('\n');
}
function formatCodeDesc(input) {
    const text = [];
    if (input instanceof Object) {
        Object.keys(input).forEach((key) => {
            text.push(`${key.charAt(0).toUpperCase() + key.slice(1)}: ${_.get(input, key)}`);
        });
    }
    return text.join('\n') + '\n';
}
function deduplicateId(input) {
    const controlId = input.map((element) => {
        return _.get(element, 'id');
    });
    const dupId = _.chain(controlId)
        .groupBy()
        .pickBy((value) => value.length > 1)
        .keys()
        .value();
    dupId.forEach((id) => {
        let index = 1;
        input
            .filter((element) => _.get(element, 'id') === id)
            .forEach((element) => {
            if (element instanceof Object) {
                _.set(element, 'id', `${id}.${index.toString()}`);
            }
            index++;
        });
    });
    return input;
}
class ZapMapper extends base_converter_1.BaseConverter {
    constructor(zapJson, name, withRaw = false) {
        super(_.set(JSON.parse(zapJson), 'site', filterSite(_.get(JSON.parse(zapJson), 'site'), name)), false);
        this.mappings = {
            platform: {
                name: 'Heimdall Tools',
                release: package_json_1.version
            },
            version: package_json_1.version,
            statistics: {},
            profiles: [
                {
                    name: 'OWASP ZAP Scan',
                    version: { path: '@version' },
                    title: {
                        path: 'site.@host',
                        transformer: (input) => {
                            return `OWASP ZAP Scan of Host: ${input}`;
                        }
                    },
                    summary: {
                        path: 'site.@host',
                        transformer: (input) => {
                            return `OWASP ZAP Scan of Host: ${input}`;
                        }
                    },
                    supports: [],
                    attributes: [],
                    groups: [],
                    status: 'loaded',
                    controls: [
                        {
                            path: 'site.alerts',
                            arrayTransformer: deduplicateId,
                            tags: {
                                cci: {
                                    path: 'cweid',
                                    transformer: (cwe) => (0, global_1.getCCIsForNISTTags)(nistTag(cwe))
                                },
                                nist: { path: 'cweid', transformer: nistTag },
                                cweid: { path: 'cweid' },
                                wascid: { path: 'wascid' },
                                sourceid: { path: 'sourceid' },
                                confidence: { path: 'confidence' },
                                riskdesc: { path: 'riskdesc' }
                            },
                            refs: [],
                            source_location: {},
                            title: { path: 'name' },
                            id: { path: 'pluginid' },
                            desc: { path: 'desc', transformer: base_converter_1.parseHtml },
                            descriptions: [
                                {
                                    data: { transformer: checkText },
                                    label: 'check'
                                }
                            ],
                            impact: { path: 'riskcode', transformer: impactMapping },
                            code: {
                                transformer: (vulnerability) => JSON.stringify(vulnerability, null, 2)
                            },
                            results: [
                                {
                                    path: 'instances',
                                    status: inspecjs_1.ExecJSON.ControlResultStatus.Failed,
                                    code_desc: { transformer: formatCodeDesc },
                                    start_time: { path: '$.@generated' }
                                }
                            ]
                        }
                    ],
                    sha256: ''
                }
            ],
            passthrough: {
                transformer: (data) => {
                    return {
                        auxiliary_data: [
                            {
                                name: 'OWASP ZAP',
                                data: _.pick(data, ['site.@port', 'site.@ssl'])
                            }
                        ],
                        ...(this.withRaw && { raw: data })
                    };
                }
            }
        };
        this.withRaw = withRaw;
    }
    toHdf() {
        const original = super.toHdf();
        _.get(original, 'profiles').forEach((profile) => {
            _.get(profile, 'controls').forEach((control) => {
                _.set(control, 'results', _.get(control, 'results').filter(function (element, index, self) {
                    return index === self.indexOf(element);
                }));
            });
        });
        return original;
    }
}
exports.ZapMapper = ZapMapper;
